// teenybase/dist/sql/schema/tableQueries.js
import { randomString } from '../../security/random';
import { literalToQuery, logSQLQuery } from '../build/query';
// todo
function buildSql(q) {
    return (q || typeof q === 'string') ? logSQLQuery(literalToQuery(typeof q === 'string' ? { q } : q, false)) : undefined;
}
export function createFieldQuery(opts) {
    const def = buildSql(opts.default);
    if (opts.notNull) {
        const isNull = def?.toUpperCase() === 'NULL';
        if (isNull)
            throw new Error(`Field ${opts.name} cannot be notNull with default null`);
    }
    let suffix = opts.sqlType.toUpperCase() ?? 'TEXT';
    if (opts.primary)
        suffix += ' PRIMARY KEY';
    if (opts.autoIncrement)
        suffix += ' AUTOINCREMENT';
    if (opts.unique)
        suffix += ' UNIQUE'; // todo conflict-clause
    if (opts.notNull)
        suffix += ' NOT NULL'; // todo conflict-clause
    if (opts.collate)
        suffix += ` COLLATE ${opts.collate}`;
    if (def)
        suffix += ` DEFAULT ${def}`;
    if (opts.check) {
        const chk = buildSql(opts.check);
        suffix += ` CHECK(${chk})`;
    }
    if (opts.foreignKey) {
        const { table, column, onUpdate, onDelete } = opts.foreignKey;
        suffix += ` REFERENCES ${table}(${column})` + (onUpdate ? ` ON UPDATE ${onUpdate}` : '') + (onDelete ? ` ON DELETE ${onDelete}` : '');
        // todo match, defer
    }
    // todo generated
    return `${opts.name} ${suffix}`;
}
export function createFieldsQuery(Type) {
    return `(\n\t${Type.map(createFieldQuery).join(', \n\t')}\n)`;
}
export function indexName(index) {
    const fields = (Array.isArray(index.fields) ? index.fields : index.fields.split(','));
    const name = index.name ?? fields.map(f => f.trim().toLowerCase().split(' ')[0].replace(/[^a-z0-9]/g, '_')).join('_');
    return { name, fields };
}
export function createIndexQuery(table, index) {
    const where = buildSql(index.where);
    const { name, fields } = indexName(index);
    return `CREATE ${index.unique ? 'UNIQUE ' : ''}INDEX idx_${table.name}_${name} ON ${table.name} (${fields.join(', ')})${where ? ` WHERE ${where}` : ''}`;
}
export function dropIndexQuery(table, index) {
    const { name, fields } = indexName(index);
    return `DROP INDEX IF EXISTS idx_${table.name}_${name}`;
}
export function createTriggerQuery(table, trigger) {
    const body = (Array.isArray(trigger.body) ? trigger.body : [trigger.body]).map(b => buildSql(b)?.trim().replace(/;+$/g, '')).filter(b => b).join(';\n') + ';';
    const seq = trigger.seq ? (trigger.seq + ' ') : '';
    const when = trigger.when ? `WHEN ${buildSql(trigger.when)} ` : '';
    const updateOf = trigger.updateOf ? `OF ${Array.isArray(trigger.updateOf) ? trigger.updateOf.join(', ') : trigger.updateOf} ` : '';
    const forEach = trigger.forEach ? `FOR EACH ${trigger.forEach} ` : '';
    return `CREATE TRIGGER tgr_${table.name}_${trigger.name} ${seq}${trigger.event} ${updateOf}ON ${table.name} ${forEach}${when}BEGIN ${body} END`;
}
export function dropTriggerQuery(table, trigger) {
    return `DROP TRIGGER IF EXISTS tgr_${table.name}_${trigger.name}`;
}
export function fts5TableName(name) {
    return `fts5_${name}_idx`;
}
export function createFts5Query(data) {
    if (!data.fullTextSearch)
        return null;
    let options = [];
    if (data.fullTextSearch.tokenize)
        options.push(`tokenize=${data.fullTextSearch.tokenize}`);
    if (data.fullTextSearch.prefix)
        options.push(`prefix=${data.fullTextSearch.prefix}`);
    // if(data.fullTextSearch.content) options.push(`content=${data.fullTextSearch.content}`)
    if (data.fullTextSearch.contentless ?? true)
        options.push(`content="${data.name}"`);
    if (data.fullTextSearch.content_rowid)
        options.push(`content_rowid=${data.fullTextSearch.content_rowid}`);
    if (data.fullTextSearch.columnsize)
        options.push(`columnsize=${data.fullTextSearch.columnsize}`);
    if (data.fullTextSearch.detail)
        options.push(`detail=${data.fullTextSearch.detail}`);
    const tableName = fts5TableName(data.name);
    let sql = `CREATE VIRTUAL TABLE ${tableName} USING fts5(${[...data.fullTextSearch.fields, ...options].map(f => f).join(', ')});`;
    let sqlRevert = `DROP TABLE IF EXISTS ${tableName};`;
    const fields = data.fullTextSearch.fields.map(f => f.split(' ')[0]).filter(f => f);
    const invalidFields = fields.filter(f => !data.fields.find(f1 => f1.name === f));
    if (invalidFields.length)
        throw new Error(`Invalid fields for fts - ${invalidFields.join(', ')}`);
    const rowId = data.fullTextSearch.content_rowid || 'rowid';
    if (fields.includes(rowId))
        throw new Error(`Invalid fields for fts - rowid/content_rowid cannot be included in fields`);
    // insert any existing data. this is required because of migrateTableQuery
    // INSERT INTO fts_idx(rowid, b, c) VALUES (new.a, new.b, new.c);
    sql += '\n' + `INSERT INTO ${tableName}(rowid, ${fields.join(', ')}) SELECT ${rowId}, ${fields.join(', ')} FROM ${data.name};`;
    // triggers
    // -- Create a table. And an external content fts5 table to index it.
    // CREATE TABLE t1(a INTEGER PRIMARY KEY, b, c);
    // CREATE VIRTUAL TABLE fts_idx USING fts5(b, c, content='t1', content_rowid='a');
    //
    // -- Triggers to keep the FTS index up to date.
    // CREATE TRIGGER t1_ai AFTER INSERT ON t1 BEGIN
    //   INSERT INTO fts_idx(rowid, b, c) VALUES (new.a, new.b, new.c);
    // END;
    // CREATE TRIGGER t1_ad AFTER DELETE ON t1 BEGIN
    //   INSERT INTO fts_idx(fts_idx, rowid, b, c) VALUES('delete', old.a, old.b, old.c);
    // END;
    // CREATE TRIGGER t1_au AFTER UPDATE ON t1 BEGIN
    //   INSERT INTO fts_idx(fts_idx, rowid, b, c) VALUES('delete', old.a, old.b, old.c);
    //   INSERT INTO fts_idx(rowid, b, c) VALUES (new.a, new.b, new.c);
    // END;
    sql += `
CREATE TRIGGER ${tableName}_ai AFTER INSERT ON ${data.name} BEGIN
    INSERT INTO ${tableName}(rowid, ${fields.join(', ')}) VALUES (new.${rowId}, new.${fields.join(', new.')});
END;`;
    sql += `
CREATE TRIGGER ${tableName}_ad AFTER DELETE ON ${data.name} BEGIN
    INSERT INTO ${tableName}(${tableName}, rowid, ${fields.join(', ')}) VALUES ('delete', old.${rowId}, old.${fields.join(', old.')});
END;`;
    sql += `
CREATE TRIGGER ${tableName}_au AFTER UPDATE ON ${data.name} BEGIN
    INSERT INTO ${tableName}(${tableName}, rowid, ${fields.join(', ')}) VALUES ('delete', old.${rowId}, old.${fields.join(', old.')});
    INSERT INTO ${tableName}(rowid, ${fields.join(', ')}) VALUES (new.${rowId}, new.${fields.join(', new.')});
END;`;
    sqlRevert = `DROP TRIGGER IF EXISTS ${tableName}_au;\n` + sqlRevert;
    sqlRevert = `DROP TRIGGER IF EXISTS ${tableName}_ad;\n` + sqlRevert;
    sqlRevert = `DROP TRIGGER IF EXISTS ${tableName}_ai;\n` + sqlRevert;
    return { sql, sqlRevert };
}
export function createTableQuery(data, _rnd, tempTableName) {
    const fields = data.fields;
    // note - dont use WITHOUT ROWID, as it would break fts5
    let sql = `CREATE TABLE ${tempTableName ?? data.name} ${createFieldsQuery(fields)};`;
    let sqlRevert = `DROP TABLE IF EXISTS ${data.name};`;
    // for triggers, indexes etc
    let sql2 = '';
    if (data.indexes) {
        for (const index of data.indexes) {
            sql2 += '\n' + createIndexQuery(data, index) + ';';
        }
    }
    if (data.triggers) {
        for (const trigger of data.triggers) {
            sql2 += '\n' + createTriggerQuery(data, trigger) + ';';
        }
    }
    if (data.fullTextSearch) {
        const fts = createFts5Query(data);
        if (fts) {
            sql2 += fts.sql;
            sqlRevert += fts.sqlRevert;
        }
    }
    return { sql, sql2, sqlRevert };
}
export function migrateTableQuery(table, existingFields) {
    // If foreign key constraints are enabled, disable them using PRAGMA foreign_keys=OFF.
    // Start a transaction.
    // Remember the format of all indexes, triggers, and views associated with table X. This information will be needed in step 8 below. One way to do this is to run a query like the following: SELECT type, sql FROM sqlite_schema WHERE tbl_name='X'.
    // Use CREATE TABLE to construct a new table "new_X" that is in the desired revised format of table X. Make sure that the name "new_X" does not collide with any existing table name, of course.
    // Transfer content from X into new_X using a statement like: INSERT INTO new_X SELECT ... FROM X.
    // Drop the old table X: DROP TABLE X.
    // Change the name of new_X to X using: ALTER TABLE new_X RENAME TO X.
    // Use CREATE INDEX, CREATE TRIGGER, and CREATE VIEW to reconstruct indexes, triggers, and views associated with table X. Perhaps use the old format of the triggers, indexes, and views saved from step 3 above as a guide, making changes as appropriate for the alteration.
    // If any views refer to table X in a way that is affected by the schema change, then drop those views using DROP VIEW and recreate them with whatever changes are necessary to accommodate the schema change using CREATE VIEW.
    // If foreign key constraints were originally enabled then run PRAGMA foreign_key_check to verify that the schema change did not break any foreign key constraints.
    // Commit the transaction started in step 2.
    // If foreign keys constraints were originally enabled, reenable them now.
    const tempName = table.name + '_' + randomString(5);
    const createTable = createTableQuery(table, undefined, tempName);
    return `
        PRAGMA foreign_keys=OFF;
        ${createTable.sql}
        INSERT INTO ${tempName} (${existingFields})
         SELECT ${existingFields} FROM ${table.name};
        ${createTable.sqlRevert /* drop old table and old table_fts. it will be created in sql2 */}
        ALTER TABLE ${tempName} RENAME TO ${table.name};
        ${createTable.sql2}
        PRAGMA foreign_key_check;
        PRAGMA foreign_keys=ON;
    `.split('\n').map(s => s.trim()).join('\n');
}
export function sqliteSupportedColumnAdd(field) {
    // The ADD COLUMN syntax is used to add a new column to an existing table. The new column is always appended to the end of the list of existing columns. The column-def rule defines the characteristics of the new column. The new column may take any of the forms permissible in a CREATE TABLE statement, with the following restrictions:
    //
    // The column may not have a PRIMARY KEY or UNIQUE constraint.
    // The column may not have a default value of CURRENT_TIME, CURRENT_DATE, CURRENT_TIMESTAMP, or an expression in parentheses.
    // If a NOT NULL constraint is specified, then the column must have a default value other than NULL.
    // If foreign key constraints are enabled and a column with a REFERENCES clause is added, the column must have a default value of NULL.
    // The column may not be GENERATED ALWAYS ... STORED, though VIRTUAL columns are allowed.
    // When adding a column with a CHECK constraint, or a NOT NULL constraint on a generated column, the added constraints are tested against all preexisting rows in the table and the ADD COLUMN fails if any constraint fails. The testing of added constraints against preexisting rows is a new enhancement as of SQLite version 3.37.0 (2021-11-27).
    //
    // The ALTER TABLE command works by modifying the SQL text of the schema stored in the sqlite_schema table. No changes are made to table content for renames or column addition without constraints. Because of this, the execution time of such ALTER TABLE commands is independent of the amount of data in the table and such commands will run as quickly on a table with 10 million rows as on a table with 1 row. When adding new columns that have CHECK constraints, or adding generated columns with NOT NULL constraints, or when deleting columns, then all existing data in the table must be either read (to test new constraints against existing rows) or written (to remove deleted columns). In those cases, the ALTER TABLE command takes time that is proportional to the amount of content in the table being altered.
    //
    // After ADD COLUMN has been run on a database, that database will not be readable by SQLite version 3.1.3 (2005-02-20) and earlier.
    const def = buildSql(field.default);
    const unsupported = field.primary || field.unique ||
        (def && (def.startsWith('(') || def.includes('CURRENT_'))) ||
        // (field.notNull && def === 'NULL') ||
        (field.foreignKey && field.notNull) ||
        (field.foreignKey && def?.toUpperCase() !== 'NULL');
    return !unsupported;
}
