// teenybase/dist/security/JWTTokenHelper.js
import { decode, sign, verify } from '@tsndr/cloudflare-worker-jwt';
import { HTTPException } from 'hono/http-exception';
export class JWTTokenHelper {
    secret;
    issuer;
    algorithm;
    allowedIssuers = [];
    constructor(secret, issuer = '$db', algorithm = 'HS256', allowedIssuers) {
        this.secret = secret;
        this.issuer = issuer;
        this.algorithm = algorithm;
        this.allowedIssuers.push(issuer);
        if (allowedIssuers)
            this.allowedIssuers.push(...allowedIssuers);
    }
    async createJwtToken(payload, secret, secondsDuration) {
        const time = Math.floor(Date.now() / 1000);
        const claims = {
            ...payload,
            iat: time,
            exp: time + secondsDuration,
            // admin: false,
            iss: this.issuer,
        };
        return sign(claims, await this.secret() + secret, { algorithm: this.algorithm, header: { typ: 'JWT' } });
    }
    async decodeAuth(auth, secret, /*hasOtp: boolean,*/ onlyVerified = true, issuers, payload) {
        payload = payload || decode(auth).payload;
        if (!this.allowedIssuers.includes(payload.iss))
            throw new HTTPException(401, { message: 'Invalid app, ' + payload.iss });
        if (issuers && !issuers.includes(payload.iss))
            throw new HTTPException(401, { message: 'Invalid issuer' });
        if (payload.iss === 'https://accounts.google.com') {
            // todo add client id check in the config. also how long is the google token valid?
            // if(payload.azp !== globalConfig.AUTH_GOOGLE_OAUTH_CLIENT_ID) throw new HTTPException(400, {message: 'Invalid google token, unauthorized app'})
            // google - https://www.googleapis.com/oauth2/v3/tokeninfo?access_token=<access token>
            // or the proper way - https://developers.google.com/identity/gsi/web/guides/verify-google-id-token#using-a-google-api-client-library
            // ideally we should issue our own token from google's token and verify that easily.
            // doing first way. todo do it properly in DO or when we figure out public cert caching in cf workers, till then this is better
            const res = await fetch(`https://oauth2.googleapis.com/tokeninfo?id_token=${auth}`);
            const json = await res.json().catch(_ => ({}));
            // console.log(json)
            // todo use zod for validation
            if (!res.ok || !json || json.error || !json.email || !json.email_verified)
                throw new HTTPException(401, { message: 'Invalid google token, unable to find email' });
            if (payload.email !== json.email)
                throw new HTTPException(401, { message: 'Unknown error, Invalid email from google verification' });
            // todo mark this token as "used" so it cannot be used again to login and create a new session.
            return { issData: payload, verified: true, sub: json.email, iss: 'google' };
        }
        if (payload.iss !== this.issuer)
            throw new HTTPException(400, { message: 'Invalid issuer' });
        const valid = await verify(auth, await this.secret() + secret, {
            throwError: false,
            algorithm: this.algorithm,
        });
        // todo if the token is expired, dont throw error, instead ignore it for resend etc?
        if (!valid)
            throw new HTTPException(401, { message: 'Unauthorized - invalid token' });
        const email = payload.sub;
        // let otp = undefined
        if (!payload.verified) {
            if (onlyVerified)
                throw new HTTPException(403, { message: 'Not verified' });
            // const otpEncrypted = payload.key
            // if(!otpEncrypted) throw new HTTPException(400, {message: 'Invalid token'})
            // otp = await aesGcmDecrypt(atob(otpEncrypted), email + secret)
            // if (!hasOtp) if (!payload.iat || (Date.now() / 1000) - payload.iat < globalConfig.OTP_EMAIL_RESEND_THRESH)
            // 	throw new HTTPException(400, {message: 'Too soon to resend'})
        }
        // return {
        // 	verified: payload.verified || false, email,
        // 	// otp,
        // 	issuer: payload.iss,
        // 	payload,
        // }
        return payload;
    }
}
