// teenybase/dist/types/dataTypes.js
export var TableFieldDataType;
(function (TableFieldDataType) {
    TableFieldDataType["text"] = "text";
    TableFieldDataType["number"] = "number";
    TableFieldDataType["bool"] = "bool";
    TableFieldDataType["email"] = "email";
    TableFieldDataType["url"] = "url";
    TableFieldDataType["editor"] = "editor";
    TableFieldDataType["date"] = "date";
    TableFieldDataType["select"] = "select";
    TableFieldDataType["json"] = "json";
    TableFieldDataType["file"] = "file";
    TableFieldDataType["relation"] = "relation";
    TableFieldDataType["password"] = "password";
    // autodate = 'autodate',
    TableFieldDataType["integer"] = "integer";
    TableFieldDataType["blob"] = "blob";
})(TableFieldDataType || (TableFieldDataType = {}));
export var TableFieldSqlDataType0;
(function (TableFieldSqlDataType0) {
    TableFieldSqlDataType0["text"] = "text";
    TableFieldSqlDataType0["integer"] = "integer";
    TableFieldSqlDataType0["real"] = "real";
    TableFieldSqlDataType0["blob"] = "blob";
    TableFieldSqlDataType0["null"] = "null";
})(TableFieldSqlDataType0 || (TableFieldSqlDataType0 = {}));
export var TableFieldSqlDataType1;
(function (TableFieldSqlDataType1) {
    TableFieldSqlDataType1["json"] = "json";
    TableFieldSqlDataType1["date"] = "date";
    TableFieldSqlDataType1["datetime"] = "datetime";
    TableFieldSqlDataType1["time"] = "time";
    TableFieldSqlDataType1["timestamp"] = "timestamp";
    TableFieldSqlDataType1["float"] = "float";
    TableFieldSqlDataType1["int"] = "int";
    TableFieldSqlDataType1["boolean"] = "boolean";
    TableFieldSqlDataType1["numeric"] = "numeric";
})(TableFieldSqlDataType1 || (TableFieldSqlDataType1 = {}));
