// teenybase/dist/scaffolds/fields.js
import { TableFieldUsageAuth, TableFieldUsageRecord } from '../worker/usages';
import { tableField } from '../types/config/tableField';
import { sql, sqlValue } from '../types/config/sqlUtils';
export const fields = {
    id: tableField("id", 'text', 'text', { usage: TableFieldUsageRecord.record_uid, primary: true, notNull: true, noUpdate: true }),
    created: tableField("created", 'date', 'timestamp', { usage: TableFieldUsageRecord.record_created, default: sql `CURRENT_TIMESTAMP`, notNull: true, noInsert: true, noUpdate: true }),
    updated: tableField("updated", 'date', 'timestamp', { usage: TableFieldUsageRecord.record_updated, default: sql `CURRENT_TIMESTAMP`, notNull: true, noInsert: true, noUpdate: true }),
    username: tableField("username", 'text', 'text', { notNull: true, unique: true, usage: TableFieldUsageAuth.auth_username }),
    email: tableField("email", 'text', 'text', { notNull: true, unique: true, usage: TableFieldUsageAuth.auth_email, noUpdate: true }),
    email_verified: tableField("email_verified", 'bool', 'boolean', { notNull: true, default: sqlValue(false), usage: TableFieldUsageAuth.auth_email_verified, noInsert: true, noUpdate: true }),
    password: tableField("password", 'text', 'text', { notNull: true, usage: TableFieldUsageAuth.auth_password, noSelect: true }),
    password_salt: tableField("password_salt", 'text', 'text', { notNull: true, usage: TableFieldUsageAuth.auth_password_salt, noSelect: true, noInsert: true, noUpdate: true }),
    name: tableField("name", 'text', 'text', { notNull: true, usage: TableFieldUsageAuth.auth_name }),
    avatar: tableField("avatar", 'file', 'text', { usage: TableFieldUsageAuth.auth_avatar }),
    role: tableField("role", 'text', 'text', { usage: TableFieldUsageAuth.auth_audience }),
    meta: tableField("meta", 'json', 'json', { usage: TableFieldUsageAuth.auth_metadata }),
};
export const triggers = {
    // raise an error if created column is updated (optional, only when executing raw sql)
    created: {
        name: "raise_on_created_update",
        seq: "BEFORE",
        event: "UPDATE",
        updateOf: [fields.created.name],
        body: sql `SELECT RAISE(FAIL, 'Cannot update created column') WHERE OLD.${fields.created.name} != NEW.${fields.created.name}`,
    },
    // update updated column automatically (optional, only when executing raw sql)
    updated: {
        name: "update_updated_on_update",
        seq: "AFTER",
        event: "UPDATE",
        body: sql `UPDATE users SET ${fields.updated.name} = CURRENT_TIMESTAMP WHERE id = NEW.id AND OLD.${fields.updated.name} = NEW.${fields.updated.name}`,
    },
};
export const baseFields = [fields.id, fields.created, fields.updated];
export const authFields = [fields.username, fields.email, fields.email_verified, fields.password, fields.password_salt, fields.name, fields.avatar, fields.role, fields.meta];
export const createdTrigger = triggers.created;
export const updatedTrigger = triggers.updated;
