// teenybase/dist/node/cli.js
// src/node/cli.ts
import { cac } from "cac";
import colors3 from "picocolors";
import prompts from "prompts";

// src/node/constants.ts
import { URL } from "url";
import { readFileSync } from "fs";
var { version } = JSON.parse(
  readFileSync(new URL("../../package.json", import.meta.url)).toString()
);
var VERSION = version;
var DEFAULT_CONFIG_FILES = [
  "teenybase.json",
  "teenybase.js",
  "teenybase.mjs",
  "teenybase.ts",
  "teeny.config.json",
  "teeny.config.js",
  "teeny.config.mjs",
  "teeny.config.ts",
  "teeny.config.cjs",
  "teeny.config.mts",
  "teeny.config.cts"
];
var DEFAULT_WRANGLER_CONFIG_FILES = [
  "wrangler.json",
  // todo prefer json?
  "wrangler.toml"
];

// src/node/config.ts
import path2 from "path";
import * as fs3 from "fs";
import { pathToFileURL } from "url";
import * as esbuild from "esbuild";
import { debug } from "util";

// src/node/logger.ts
import readline from "readline";
import colors from "picocolors";
var LogLevels = {
  silent: 0,
  error: 1,
  warn: 2,
  info: 3,
  debug: 4
};
var lastType;
var lastMsg;
var sameCount = 0;
function clearScreen() {
  const repeatCount = process.stdout.rows - 2;
  const blank = repeatCount > 0 ? "\n".repeat(repeatCount) : "";
  console.log(blank);
  readline.cursorTo(process.stdout, 0, 0);
  readline.clearScreenDown(process.stdout);
}
var timeFormatter;
function getTimeFormatter() {
  timeFormatter ??= new Intl.DateTimeFormat(void 0, {
    hour: "numeric",
    minute: "numeric",
    second: "numeric"
  });
  return timeFormatter;
}
function createLogger(level = "info", options = {}) {
  if (options.customLogger) {
    return options.customLogger;
  }
  const loggedErrors = /* @__PURE__ */ new WeakSet();
  const {
    prefix = "[teeny]",
    allowClearScreen = true,
    console: console2 = globalThis.console
  } = options;
  const thresh = LogLevels[level];
  const canClearScreen = allowClearScreen && process.stdout.isTTY && !process.env.CI;
  const clear = canClearScreen ? clearScreen : () => {
  };
  function format(type, msg, options2 = {}) {
    if (options2.timestamp) {
      let tag = "";
      if (type === "info") {
        tag = colors.cyan(colors.bold(prefix));
      } else if (type === "warn") {
        tag = colors.yellow(colors.bold(prefix));
      } else {
        tag = colors.red(colors.bold(prefix));
      }
      const environment = options2.environment ? options2.environment + " " : "";
      return `${colors.dim(getTimeFormatter().format(/* @__PURE__ */ new Date()))} ${tag} ${environment}${msg}`;
    } else {
      return msg;
    }
  }
  function output(type, msg, options2 = {}) {
    if (thresh >= LogLevels[type]) {
      const method = type === "info" || type === "debug" ? "log" : type;
      if (options2.error) {
        loggedErrors.add(options2.error);
      }
      if (canClearScreen) {
        if (type === lastType && msg === lastMsg) {
          sameCount++;
          clear();
          console2[method](
            format(type, msg, options2),
            colors.yellow(`(x${sameCount + 1})`)
          );
        } else {
          sameCount = 0;
          lastMsg = msg;
          lastType = type;
          if (options2.clear) {
            clear();
          }
          console2[method](format(type, msg, options2));
        }
      } else {
        console2[method](format(type, msg, options2));
      }
    }
  }
  const warnedMessages = /* @__PURE__ */ new Set();
  const logger = {
    hasWarned: false,
    debug(msg, opts) {
      output("debug", msg, opts);
    },
    info(msg, opts) {
      output("info", msg, opts);
    },
    warn(msg, opts) {
      logger.hasWarned = true;
      output("warn", msg, opts);
    },
    warnOnce(msg, opts) {
      if (warnedMessages.has(msg)) return;
      logger.hasWarned = true;
      output("warn", msg, opts);
      warnedMessages.add(msg);
    },
    error(msg, opts) {
      logger.hasWarned = true;
      output("error", msg, opts);
    },
    clearScreen(type) {
      if (thresh >= LogLevels[type]) {
        clear();
      }
    },
    hasErrorLogged(error) {
      return loggedErrors.has(error);
    }
  };
  return logger;
}

// src/node/utils.ts
import * as fs from "fs";
import path from "path";
var isWindows = typeof process !== "undefined" && process.platform === "win32";
var windowsSlashRE = /\\/g;
function slash(p) {
  return p.replace(windowsSlashRE, "/");
}
function tryStatSync(file) {
  try {
    return fs.statSync(file, { throwIfNoEntry: false });
  } catch {
  }
}
function findNearestNodeModules(basedir) {
  while (basedir) {
    const pkgPath = path.join(basedir, "node_modules");
    if (tryStatSync(pkgPath)?.isDirectory()) {
      return pkgPath;
    }
    const nextBasedir = path.dirname(basedir);
    if (nextBasedir === basedir) break;
    basedir = nextBasedir;
  }
  return null;
}
function isObject(value) {
  return Object.prototype.toString.call(value) === "[object Object]";
}
function normalizePath(id) {
  return path.posix.normalize(isWindows ? slash(id) : id);
}
function readableRelative(to) {
  const relativePath = path.relative(process.cwd(), to);
  if (
    // No directory nesting, return as-is
    path.basename(relativePath) === relativePath || // Outside current directory
    relativePath.startsWith(".")
  ) {
    return relativePath;
  } else {
    return "./" + relativePath;
  }
}
function escapeRegExp(string) {
  return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

// src/node/config.ts
import colors2 from "picocolors";

// src/node/wrangler/parse.ts
import * as fs2 from "node:fs";
import { resolve } from "node:path";
import TOML from "@iarna/toml";
import { formatMessagesSync } from "esbuild";
import { parse as jsoncParse, printParseErrorCode } from "jsonc-parser";
var UserError = class extends Error {
  constructor(...args) {
    super(...args);
    Object.setPrototypeOf(this, new.target.prototype);
  }
};
var ParseError = class extends UserError {
  text;
  notes;
  location;
  kind;
  constructor({ text, notes, location, kind }) {
    super(text);
    this.name = this.constructor.name;
    this.text = text;
    this.notes = notes ?? [];
    this.location = location;
    this.kind = kind ?? "error";
  }
};
var TOML_ERROR_NAME = "TomlError";
var TOML_ERROR_SUFFIX = " at row ";
function parseTOML(input, file) {
  try {
    const normalizedInput = input.replace(/\r\n/g, "\n");
    return TOML.parse(normalizedInput);
  } catch (err) {
    const { name, message, line, col } = err;
    if (name !== TOML_ERROR_NAME) {
      throw err;
    }
    const text = message.substring(0, message.lastIndexOf(TOML_ERROR_SUFFIX));
    const lineText = input.split("\n")[line];
    const location = {
      lineText,
      line: line + 1,
      column: col - 1,
      file,
      fileText: input
    };
    throw new ParseError({ text, location });
  }
}
function parseJSONC(input, file) {
  const errors = [];
  const data = jsoncParse(input, errors);
  if (errors.length) {
    throw new ParseError({
      text: printParseErrorCode(errors[0].error),
      location: {
        ...indexLocation({ file, fileText: input }, errors[0].offset + 1),
        length: errors[0].length
      }
    });
  }
  return data;
}
function readFileSync3(file) {
  try {
    return fs2.readFileSync(file, { encoding: "utf-8" });
  } catch (err) {
    const { message } = err;
    throw new ParseError({
      text: `Could not read file: ${file}`,
      notes: [
        {
          text: message.replace(file, resolve(file))
        }
      ]
    });
  }
}
function indexLocation(file, index) {
  let lineText, line = 0, column = 0, cursor = 0;
  const { fileText = "" } = file;
  for (const row of fileText.split("\n")) {
    line++;
    cursor += row.length + 1;
    if (cursor >= index) {
      lineText = row;
      column = row.length - (cursor - index);
      break;
    }
  }
  return { lineText, line, column, ...file };
}

// src/node/config.ts
async function bundleCode(fileName, logLevel) {
  const options = {
    entryPoints: [fileName],
    bundle: true,
    write: false,
    // resolveExtensions: ['.ts', '.js', '.json'],
    minify: false,
    logLevel: logLevel === "warn" ? "warning" : logLevel,
    platform: "node",
    allowOverwrite: false,
    sourcemap: false,
    format: "esm",
    target: "esnext",
    // format: OutputFormat.ESM,
    // mainFields: ['module', 'main'],
    banner: {
      "js": "import { createRequire } from 'module'; const require = createRequire(import.meta.url);"
    }
  };
  const built = await esbuild.build(options);
  const bundledCode = built.outputFiles?.[0].text;
  if (!bundledCode) throw new Error("Failed to load code - " + fileName);
  return bundledCode;
}
async function loadConfigFromBundledFile(bundle, fileName) {
  const nodeModulesDir = findNearestNodeModules(path2.dirname(fileName));
  if (nodeModulesDir) {
    await fs3.promises.mkdir(path2.resolve(nodeModulesDir, ".teeny-temp/"), {
      recursive: true
    });
  }
  const hash = `timestamp-${Date.now()}-${Math.random().toString(16).slice(2)}`;
  const tempFileName = nodeModulesDir ? path2.resolve(
    nodeModulesDir,
    `.teeny-temp/${path2.basename(fileName)}.${hash}.mjs`
  ) : `${fileName}.${hash}.mjs`;
  await fs3.promises.writeFile(tempFileName, bundle);
  try {
    return (await import(pathToFileURL(tempFileName).href)).default;
  } finally {
    fs3.unlink(tempFileName, () => {
    });
  }
}
async function loadConfigFromFile(configRoot, configFile, logLevel, customLogger) {
  const start = performance.now();
  const getTime = () => `${(performance.now() - start).toFixed(2)}ms`;
  let resolvedPath;
  if (configFile) {
    resolvedPath = path2.resolve(configFile);
  } else {
    for (const filename of DEFAULT_CONFIG_FILES) {
      const filePath = path2.resolve(configRoot, filename);
      if (!fs3.existsSync(filePath)) continue;
      resolvedPath = filePath;
      break;
    }
  }
  if (!resolvedPath) {
    throw new Error("no config file found.");
  }
  const isESM = true;
  try {
    const bundled = await bundleCode(resolvedPath, logLevel);
    const userConfig = await loadConfigFromBundledFile(
      bundled,
      resolvedPath
    );
    debug?.(`bundled config file loaded in ${getTime()}`);
    const config = await (typeof userConfig === "function" ? userConfig() : userConfig);
    if (!isObject(config)) {
      throw new Error(`config must export or return an object.`);
    }
    return {
      path: normalizePath(resolvedPath),
      config
      // dependencies: bundled.dependencies,
    };
  } catch (e) {
    createLogger(logLevel, { customLogger }).error(
      colors2.red(`failed to load config from ${resolvedPath}`),
      { error: e }
    );
    throw e;
  }
}
async function loadWranglerConfigFromFile(configRoot, configFile, logLevel, customLogger) {
  const start = performance.now();
  const getTime = () => `${(performance.now() - start).toFixed(2)}ms`;
  let resolvedPath;
  if (configFile) {
    resolvedPath = path2.resolve(configFile);
  } else {
    for (const filename of DEFAULT_WRANGLER_CONFIG_FILES) {
      const filePath = path2.resolve(configRoot, filename);
      if (!fs3.existsSync(filePath)) continue;
      resolvedPath = filePath;
      break;
    }
  }
  if (!resolvedPath) {
    throw new Error("no wrangler config file found.");
  }
  try {
    let loadedConfig = {};
    if (resolvedPath.endsWith(".json")) {
      loadedConfig = parseJSONC(readFileSync3(resolvedPath), resolvedPath);
    } else if (resolvedPath.endsWith(".toml")) {
      loadedConfig = parseTOML(readFileSync3(resolvedPath), resolvedPath);
    }
    return {
      path: normalizePath(resolvedPath),
      config: loadedConfig
    };
  } catch (e) {
    createLogger(logLevel, { customLogger }).error(
      colors2.red(`failed to load config from ${resolvedPath}`),
      { error: e }
    );
    throw e;
  }
}

// src/node/cli.ts
import * as fs5 from "node:fs";
import path4 from "path";
import { unstable_dev } from "wrangler";
import { execSync } from "node:child_process";
import dotenv from "dotenv";
import { URL as URL2 } from "url";

// src/node/wrangler/export.ts
import fs4 from "node:fs/promises";
import path3 from "node:path";
import { Miniflare } from "miniflare";
function getDatabaseInfoFromConfig(config, name) {
  for (const d1Database of config.d1_databases) {
    if (d1Database.database_id && (name === d1Database.database_name || name === d1Database.binding)) {
      return {
        uuid: d1Database.database_id,
        previewDatabaseUuid: d1Database.preview_database_id,
        binding: d1Database.binding,
        name: d1Database.database_name
        // migrationsTableName:
        //     d1Database.migrations_table,
        // migrationsFolderPath:
        //     d1Database.migrations_dir,
        // internal_env: d1Database.database_internal_env,
      };
    }
  }
  return null;
}
async function exportLocal(config, name, output, tables, noSchema, noData, persistencePath, logger) {
  const localDB = getDatabaseInfoFromConfig(config, name);
  if (!localDB) {
    throw new UserError(
      `Couldn't find a D1 DB with the name or binding '${name}' in wrangler.toml.`
    );
  }
  const id = localDB.previewDatabaseUuid ?? localDB.uuid;
  const d1Persist = path3.join(persistencePath, "v3", "d1");
  logger.info(
    `\u{1F300} Exporting local database ${name} (${id}) from ${readableRelative(
      d1Persist
    )}:`
  );
  const mf = new Miniflare({
    modules: true,
    script: "export default {}",
    d1Persist,
    d1Databases: { DATABASE: id }
  });
  const db = await mf.getD1Database("DATABASE");
  logger.info(`\u{1F300} Exporting SQL to ${output}...`);
  try {
    const dump = await db.prepare(`PRAGMA miniflare_d1_export(?,?,?);`).bind(noSchema, noData, ...tables).raw();
    await fs4.writeFile(output, dump[0].join("\n"));
  } catch (e) {
    throw new UserError(e.message);
  } finally {
    await mf.dispose();
  }
  logger.info(`Done!`);
}

// src/security/random.ts
function randomString(len) {
  const vals = crypto.getRandomValues(new Uint8Array(Math.ceil(len / 2)));
  return Array.from(vals).map((b) => b.toString(16).padStart(2, "0")).join("").substring(0, len);
}

// src/sql/build/query.ts
function logSQLQuery(query) {
  if (typeof query.q !== "string") query.q = JSON.stringify(query.l || query.q);
  const params = query.p ? Object.entries(query.p) : null;
  let q = query.q.trim();
  if (params?.length) {
    for (const [k, val] of params) {
      let v = val;
      if (v !== null && (typeof v === "object" || Array.isArray(v))) {
        v = JSON.stringify(v);
      }
      q = q.replaceAll("{:" + k + "}", JSON.stringify(v));
    }
  }
  return q;
}
var literalMapping = {
  0: "0",
  1: "1",
  true: "1",
  TRUE: "1",
  false: "0",
  FALSE: "0",
  null: "NULL",
  NULL: "NULL"
};
function literalToQuery(l, wrap = true) {
  let l1 = l.l;
  const isLiteral = l1 !== void 0;
  const q = l.q;
  if (!isLiteral && typeof l !== "string") {
    if (typeof q !== "string") throw new Error("Invalid query " + JSON.stringify(l));
    return {
      // q: (!q || (q[0] === '(' && q[q.length - 1] === ')')) ? q : `(${q})`, // Note - this is wrong, since it could be like `(a) OR (b)` which should also be escaped.
      q: !q || !wrap ? q : `(${q})`,
      // tag - brackets
      p: l.p,
      dependencies: l.dependencies,
      _readOnly: l._readOnly
    };
  }
  l1 = isLiteral && typeof l === "object" ? l1 : l;
  let lm = typeof l1 !== "object" ? literalMapping[l1] : void 0;
  if (lm !== void 0)
    return { q: lm, _readOnly: true };
  if (l1 === null || l1 === void 0) return { q: "NULL", _readOnly: true };
  const rnd = "f" + Math.random().toString(36).substring(7);
  return {
    q: "{:" + rnd + "}",
    p: { [rnd]: l1 },
    // todo is object check required? right now it's done in sqlQueryToD1Query
    _readOnly: true
  };
}

// src/sql/schema/tableQueries.ts
function buildSql(q) {
  return q || typeof q === "string" ? logSQLQuery(literalToQuery(typeof q === "string" ? { q } : q, false)) : void 0;
}
function createFieldQuery(opts) {
  const def = buildSql(opts.default);
  if (opts.notNull) {
    const isNull = def?.toUpperCase() === "NULL";
    if (isNull) throw new Error(`Field ${opts.name} cannot be notNull with default null`);
  }
  let suffix = opts.sqlType.toUpperCase() ?? "TEXT";
  if (opts.primary) suffix += " PRIMARY KEY";
  if (opts.autoIncrement) suffix += " AUTOINCREMENT";
  if (opts.unique) suffix += " UNIQUE";
  if (opts.notNull) suffix += " NOT NULL";
  if (opts.collate) suffix += ` COLLATE ${opts.collate}`;
  if (def) suffix += ` DEFAULT ${def}`;
  if (opts.check) {
    const chk = buildSql(opts.check);
    suffix += ` CHECK(${chk})`;
  }
  if (opts.foreignKey) {
    const { table, column, onUpdate, onDelete } = opts.foreignKey;
    suffix += ` REFERENCES ${table}(${column})` + (onUpdate ? ` ON UPDATE ${onUpdate}` : "") + (onDelete ? ` ON DELETE ${onDelete}` : "");
  }
  return `${opts.name} ${suffix}`;
}
function createFieldsQuery(Type) {
  return `(
	${Type.map(createFieldQuery).join(", \n	")}
)`;
}
function indexName(index) {
  const fields = Array.isArray(index.fields) ? index.fields : index.fields.split(",");
  const name = index.name ?? fields.map((f) => f.trim().toLowerCase().split(" ")[0].replace(/[^a-z0-9]/g, "_")).join("_");
  return { name, fields };
}
function createIndexQuery(table, index) {
  const where = buildSql(index.where);
  const { name, fields } = indexName(index);
  return `CREATE ${index.unique ? "UNIQUE " : ""}INDEX idx_${table.name}_${name} ON ${table.name} (${fields.join(", ")})${where ? ` WHERE ${where}` : ""}`;
}
function dropIndexQuery(table, index) {
  const { name, fields } = indexName(index);
  return `DROP INDEX IF EXISTS idx_${table.name}_${name}`;
}
function createTriggerQuery(table, trigger) {
  const body = (Array.isArray(trigger.body) ? trigger.body : [trigger.body]).map((b) => buildSql(b)?.trim().replace(/;+$/g, "")).filter((b) => b).join(";\n") + ";";
  const seq = trigger.seq ? trigger.seq + " " : "";
  const when = trigger.when ? `WHEN ${buildSql(trigger.when)} ` : "";
  const updateOf = trigger.updateOf ? `OF ${Array.isArray(trigger.updateOf) ? trigger.updateOf.join(", ") : trigger.updateOf} ` : "";
  const forEach = trigger.forEach ? `FOR EACH ${trigger.forEach} ` : "";
  return `CREATE TRIGGER tgr_${table.name}_${trigger.name} ${seq}${trigger.event} ${updateOf}ON ${table.name} ${forEach}${when}BEGIN ${body} END`;
}
function dropTriggerQuery(table, trigger) {
  return `DROP TRIGGER IF EXISTS tgr_${table.name}_${trigger.name}`;
}
function fts5TableName(name) {
  return `fts5_${name}_idx`;
}
function createFts5Query(data) {
  if (!data.fullTextSearch) return null;
  let options = [];
  if (data.fullTextSearch.tokenize) options.push(`tokenize=${data.fullTextSearch.tokenize}`);
  if (data.fullTextSearch.prefix) options.push(`prefix=${data.fullTextSearch.prefix}`);
  if (data.fullTextSearch.contentless ?? true) options.push(`content="${data.name}"`);
  if (data.fullTextSearch.content_rowid) options.push(`content_rowid=${data.fullTextSearch.content_rowid}`);
  if (data.fullTextSearch.columnsize) options.push(`columnsize=${data.fullTextSearch.columnsize}`);
  if (data.fullTextSearch.detail) options.push(`detail=${data.fullTextSearch.detail}`);
  const tableName = fts5TableName(data.name);
  let sql = `CREATE VIRTUAL TABLE ${tableName} USING fts5(${[...data.fullTextSearch.fields, ...options].map((f) => f).join(", ")});`;
  let sqlRevert = `DROP TABLE IF EXISTS ${tableName};`;
  const fields = data.fullTextSearch.fields.map((f) => f.split(" ")[0]).filter((f) => f);
  const invalidFields = fields.filter((f) => !data.fields.find((f1) => f1.name === f));
  if (invalidFields.length) throw new Error(`Invalid fields for fts - ${invalidFields.join(", ")}`);
  const rowId = data.fullTextSearch.content_rowid || "rowid";
  if (fields.includes(rowId)) throw new Error(`Invalid fields for fts - rowid/content_rowid cannot be included in fields`);
  sql += `
INSERT INTO ${tableName}(rowid, ${fields.join(", ")}) SELECT ${rowId}, ${fields.join(", ")} FROM ${data.name};`;
  sql += `
CREATE TRIGGER ${tableName}_ai AFTER INSERT ON ${data.name} BEGIN
    INSERT INTO ${tableName}(rowid, ${fields.join(", ")}) VALUES (new.${rowId}, new.${fields.join(", new.")});
END;`;
  sql += `
CREATE TRIGGER ${tableName}_ad AFTER DELETE ON ${data.name} BEGIN
    INSERT INTO ${tableName}(${tableName}, rowid, ${fields.join(", ")}) VALUES ('delete', old.${rowId}, old.${fields.join(", old.")});
END;`;
  sql += `
CREATE TRIGGER ${tableName}_au AFTER UPDATE ON ${data.name} BEGIN
    INSERT INTO ${tableName}(${tableName}, rowid, ${fields.join(", ")}) VALUES ('delete', old.${rowId}, old.${fields.join(", old.")});
    INSERT INTO ${tableName}(rowid, ${fields.join(", ")}) VALUES (new.${rowId}, new.${fields.join(", new.")});
END;`;
  sqlRevert = `DROP TRIGGER IF EXISTS ${tableName}_au;
` + sqlRevert;
  sqlRevert = `DROP TRIGGER IF EXISTS ${tableName}_ad;
` + sqlRevert;
  sqlRevert = `DROP TRIGGER IF EXISTS ${tableName}_ai;
` + sqlRevert;
  return { sql, sqlRevert };
}
function createTableQuery(data, _rnd, tempTableName) {
  const fields = data.fields;
  let sql = `CREATE TABLE ${tempTableName ?? data.name} ${createFieldsQuery(fields)};`;
  let sqlRevert = `DROP TABLE IF EXISTS ${data.name};`;
  let sql2 = "";
  if (data.indexes) {
    for (const index of data.indexes) {
      sql2 += "\n" + createIndexQuery(data, index) + ";";
    }
  }
  if (data.triggers) {
    for (const trigger of data.triggers) {
      sql2 += "\n" + createTriggerQuery(data, trigger) + ";";
    }
  }
  if (data.fullTextSearch) {
    const fts = createFts5Query(data);
    if (fts) {
      sql2 += fts.sql;
      sqlRevert += fts.sqlRevert;
    }
  }
  return { sql, sql2, sqlRevert };
}
function migrateTableQuery(table, existingFields) {
  const tempName = table.name + "_" + randomString(5);
  const createTable2 = createTableQuery(table, void 0, tempName);
  return `
        PRAGMA foreign_keys=OFF;
        ${createTable2.sql}
        INSERT INTO ${tempName} (${existingFields})
         SELECT ${existingFields} FROM ${table.name};
        ${createTable2.sqlRevert}
        ALTER TABLE ${tempName} RENAME TO ${table.name};
        ${createTable2.sql2}
        PRAGMA foreign_key_check;
        PRAGMA foreign_keys=ON;
    `.split("\n").map((s) => s.trim()).join("\n");
}
function sqliteSupportedColumnAdd(field) {
  const def = buildSql(field.default);
  const unsupported = field.primary || field.unique || def && (def.startsWith("(") || def.includes("CURRENT_")) || // (field.notNull && def === 'NULL') ||
  field.foreignKey && field.notNull || field.foreignKey && def?.toUpperCase() !== "NULL";
  return !unsupported;
}

// src/sql/schema/dropTable.ts
function dropTable(data, others) {
  let field = void 0;
  const fk = others.find((t) => field = t.fields.find((f) => f.foreignKey && f.foreignKey.table === data.name));
  if (fk) throw new Error(`Table ${data.name} is referenced by ${fk.name}.${field.name}`);
  const query = createTableQuery(data);
  return { sql: query.sqlRevert, sqlRevert: query.sql + query.sql2, logs: [`\u26A0 Table Dropped - ${data.name}`] };
}

// src/types/zod/tableDataSchema.ts
import { z as z3 } from "zod";

// src/types/zod/tableFieldDataSchema.ts
import { z as z2 } from "zod";

// src/types/dataTypes.ts
var TableFieldDataType = /* @__PURE__ */ ((TableFieldDataType2) => {
  TableFieldDataType2["text"] = "text";
  TableFieldDataType2["number"] = "number";
  TableFieldDataType2["bool"] = "bool";
  TableFieldDataType2["email"] = "email";
  TableFieldDataType2["url"] = "url";
  TableFieldDataType2["editor"] = "editor";
  TableFieldDataType2["date"] = "date";
  TableFieldDataType2["select"] = "select";
  TableFieldDataType2["json"] = "json";
  TableFieldDataType2["file"] = "file";
  TableFieldDataType2["relation"] = "relation";
  TableFieldDataType2["password"] = "password";
  TableFieldDataType2["integer"] = "integer";
  TableFieldDataType2["blob"] = "blob";
  return TableFieldDataType2;
})(TableFieldDataType || {});
var TableFieldSqlDataType0 = /* @__PURE__ */ ((TableFieldSqlDataType02) => {
  TableFieldSqlDataType02["text"] = "text";
  TableFieldSqlDataType02["integer"] = "integer";
  TableFieldSqlDataType02["real"] = "real";
  TableFieldSqlDataType02["blob"] = "blob";
  TableFieldSqlDataType02["null"] = "null";
  return TableFieldSqlDataType02;
})(TableFieldSqlDataType0 || {});
var TableFieldSqlDataType1 = /* @__PURE__ */ ((TableFieldSqlDataType12) => {
  TableFieldSqlDataType12["json"] = "json";
  TableFieldSqlDataType12["date"] = "date";
  TableFieldSqlDataType12["datetime"] = "datetime";
  TableFieldSqlDataType12["time"] = "time";
  TableFieldSqlDataType12["timestamp"] = "timestamp";
  TableFieldSqlDataType12["float"] = "float";
  TableFieldSqlDataType12["int"] = "int";
  TableFieldSqlDataType12["boolean"] = "boolean";
  TableFieldSqlDataType12["numeric"] = "numeric";
  return TableFieldSqlDataType12;
})(TableFieldSqlDataType1 || {});

// src/types/zod/sqlSchemas.ts
import { z } from "zod";
var sqlExprSchema = z.string().max(1e3).regex(/^[a-zA-Z0-9_=\s():.*~%'"><!+\-&|@,\/\\${}]*$/, "Invalid expression");
var sqlExprSchemaRecord = z.record(sqlExprSchema);
var sqlValSchema = z.string().or(z.number()).or(z.boolean()).or(z.null());
var sqlValSchemaFile = z.boolean().or(z.string().or(z.number()).or(z.null()).or(z.instanceof(File)));
var sqlValSchemaRecord = z.record(sqlValSchema);
var sqlValSchema2 = sqlValSchema.or(sqlValSchemaRecord);
var sqlValSchemaFile2 = z.record(sqlValSchemaFile);
var sqlValSchemaFile3 = sqlValSchemaFile2.or(z.array(sqlValSchemaFile2));
var sqlColListSchema = z.string().regex(/^[a-zA-Z0-9_,\s]*$/).or(z.literal("*"));
var sqlColListSchemaOrder = z.string().regex(/^[a-zA-Z0-9_,\s+-]*$/);
var sqlColListSchema2 = z.array(sqlColListSchema).or(sqlColListSchema);
var sqlColListSchemaOrder2 = z.array(sqlColListSchemaOrder).or(sqlColListSchemaOrder);
var tableColumnNameSchema = z.string().min(1).max(255).regex(/^[a-zA-Z_][a-zA-Z0-9_]*$/, "Table/Column name must start with a letter or underscore and can only contain letters, numbers and underscores");
var zSQLQuery = z.object({
  q: z.string(),
  p: z.record(sqlValSchema2).optional()
});
var zSQLLiteral = z.object({
  l: sqlValSchema2,
  key: z.string().optional()
});
var zSQLQueryOrLiteral = zSQLQuery.or(zSQLLiteral);
var zOnConflict = z.enum(["ABORT", "FAIL", "IGNORE", "REPLACE", "ROLLBACK"]);
var selectSchema = z.object({
  where: sqlExprSchema.optional(),
  limit: z.coerce.number().optional(),
  offset: z.coerce.number().optional(),
  order: sqlColListSchemaOrder2.optional(),
  group: sqlColListSchema2.optional(),
  select: sqlExprSchema.or(z.array(sqlExprSchema)).optional(),
  // todo rename to fields in api v2
  distinct: z.coerce.boolean().optional()
  // join: joinObjSchema2.array().optional(),
  // having: z.string().regex(/^[a-zA-Z0-9_=\s\(\)\.\*]*$/).optional(),
  // selectOption: z.string().optional(),
});
var updateSchema = z.object({
  or: zOnConflict.optional(),
  set: z.record(sqlExprSchema).optional(),
  setValues: z.record(sqlValSchema).optional(),
  where: sqlExprSchema,
  returning: z.array(sqlExprSchema).or(sqlExprSchema).optional()
});
var insertSchema = z.object({
  or: zOnConflict.optional(),
  values: sqlValSchemaRecord.or(z.array(sqlValSchemaRecord)).optional(),
  expr: sqlExprSchemaRecord.or(z.array(sqlExprSchemaRecord)).optional(),
  returning: z.array(sqlExprSchema).or(sqlExprSchema).optional()
});
var deleteSchema = z.object({
  where: sqlExprSchema,
  returning: z.array(sqlExprSchema).or(sqlExprSchema).optional()
});
var tableDeleteSchema = deleteSchema;
var tableInsertSchema = insertSchema.omit({ values: true }).extend({ values: sqlValSchemaFile3 });
var tableUpdateSchema = updateSchema.omit({ setValues: true }).extend({ setValues: sqlValSchemaFile2.optional() });
var tableSelectSchema = selectSchema;
var tableViewSchema = selectSchema.pick({ select: true, where: true });
var tableEditSchema = z.object({
  setValues: sqlValSchemaFile2,
  or: updateSchema.shape.or.or(z.literal("INSERT")),
  returning: updateSchema.shape.returning
});
var selectStatementSchema = tableSelectSchema.extend({
  type: z.literal("SELECT"),
  table: tableColumnNameSchema
});
var insertStatementSchema = tableInsertSchema.extend({
  type: z.literal("INSERT"),
  table: tableColumnNameSchema
});
var updateStatementSchema = tableUpdateSchema.extend({
  type: z.literal("UPDATE"),
  table: tableColumnNameSchema
});
var deleteStatementSchema = tableDeleteSchema.extend({
  type: z.literal("DELETE"),
  table: tableColumnNameSchema
});
var zSQLIndex = z.object({
  name: tableColumnNameSchema.optional(),
  unique: z.coerce.boolean().optional(),
  // fields: tableColumnNameSchema.or(z.array(tableColumnNameSchema)),
  fields: z.string().or(z.array(z.string())),
  // fields can include collate also
  where: zSQLQuery.optional()
});
var zSQLTrigger = z.object({
  name: tableColumnNameSchema,
  event: z.enum(["INSERT", "DELETE", "UPDATE"]),
  seq: z.enum(["BEFORE", "AFTER", "INSTEAD OF"]).optional(),
  updateOf: tableColumnNameSchema.or(z.array(tableColumnNameSchema)).optional(),
  forEach: z.enum(["ROW"]).optional(),
  body: zSQLQuery.or(z.array(zSQLQuery)),
  when: zSQLQuery.optional()
});
var zTypedSQLQuery = z.record(z.string(), z.any());
var zTypedSQLStatement = selectStatementSchema.or(insertStatementSchema).or(updateStatementSchema).or(deleteStatementSchema);
var zSQLProcedure = z.object({
  name: tableColumnNameSchema,
  rule: z.boolean().default(true),
  params: z.array(tableColumnNameSchema),
  query: z.union([zTypedSQLQuery, z.array(zTypedSQLQuery)]).optional(),
  statement: z.union([zTypedSQLStatement, z.array(zTypedSQLStatement)]).optional()
});

// src/types/zod/tableFieldDataSchema.ts
var zForeignKeyAction = z2.enum(["SET NULL", "SET DEFAULT", "CASCADE", "RESTRICT", "NO ACTION"]);
var tableFieldDataSchema = z2.object({
  name: z2.string(),
  sqlType: z2.nativeEnum(TableFieldSqlDataType0).or(z2.nativeEnum(TableFieldSqlDataType1)),
  type: z2.nativeEnum(TableFieldDataType),
  usage: z2.string().optional(),
  primary: z2.coerce.boolean().optional(),
  autoIncrement: z2.coerce.boolean().optional(),
  unique: z2.coerce.boolean().optional(),
  notNull: z2.coerce.boolean().optional(),
  collate: z2.string().optional(),
  // todo enum
  default: zSQLQueryOrLiteral.or(z2.string()).optional(),
  check: zSQLQueryOrLiteral.or(z2.string()).optional(),
  foreignKey: z2.object({
    table: z2.string(),
    column: z2.string(),
    onUpdate: zForeignKeyAction.optional(),
    onDelete: zForeignKeyAction.optional()
  }).optional(),
  // updateTriggers: z.array(zSQLTrigger.omit({updateOf: true, event: true})).optional(),
  noUpdate: z2.coerce.boolean().optional(),
  noInsert: z2.coerce.boolean().optional(),
  noSelect: z2.coerce.boolean().optional(),
  lastName: z2.string().optional()
});

// src/types/zod/tableDataSchema.ts
var tableDataSchema = z3.object({
  // id: z.string(),
  name: tableColumnNameSchema,
  r2Base: z3.string().optional(),
  idInR2: z3.coerce.boolean().optional(),
  autoDeleteR2Files: z3.coerce.boolean().optional(),
  allowMultipleFileRef: z3.coerce.boolean().optional(),
  allowWildcard: z3.coerce.boolean().optional(),
  fields: z3.array(tableFieldDataSchema),
  indexes: z3.array(zSQLIndex).optional(),
  triggers: z3.array(zSQLTrigger).optional(),
  autoSetUid: z3.coerce.boolean().optional(),
  extensions: z3.array(z3.record(z3.any()).and(z3.object({ name: z3.string() }))),
  lastName: tableColumnNameSchema.optional(),
  fullTextSearch: z3.object({
    enabled: z3.coerce.boolean().optional(),
    fields: z3.array(tableColumnNameSchema),
    tokenize: z3.string().optional(),
    prefix: z3.string().optional(),
    // content: z.string().optional(),
    migrateTableQuery: z3.coerce.boolean().default(true),
    content_rowid: z3.string().optional(),
    columnsize: z3.literal(0).or(z3.literal(1)).optional(),
    detail: z3.enum(["full", "column", "none"]).optional()
  }).optional()
});

// src/sql/schema/alterTable.ts
function alterTable([table, lastTable, changes], others) {
  table = tableDataSchema.parse(table);
  let statements = [];
  const tableName = table.name;
  let clearTriggers = false;
  let clearIndexes = false;
  let resetFts = changes.fts;
  let logs = [];
  if (table.name !== lastTable.name) {
    statements.push([
      `ALTER TABLE ${lastTable.name} RENAME TO ${tableName};`,
      `ALTER TABLE ${tableName} RENAME TO ${lastTable.name};`
    ]);
    clearTriggers = true;
    clearIndexes = true;
    resetFts = true;
    logs.push(`\u26A0 Table Renamed - ${lastTable.name} -> ${tableName}`);
  }
  const toRename = {};
  const renamedFieldNames = {};
  let needsTableChange = false;
  for (const field of changes.drop) {
    const fkCheck = others.find((t) => t.fields.find((f) => f.foreignKey && (f.foreignKey.table === tableName && f.foreignKey.column === field.name)));
    if (fkCheck) throw new Error(`Field ${tableName}:${field.name} - Field cannot be dropped - used in ${fkCheck.name}`);
    if (field.usage) throw new Error(`Field ${tableName}:${field.name} - Field cannot be dropped - ${field.usage} usage exists`);
    if (!sqliteSupportedColumnAdd(field)) needsTableChange = true;
    statements.push([
      `ALTER TABLE ${tableName} DROP COLUMN ${field.name};`,
      `ALTER TABLE ${tableName}
                ADD COLUMN ${createFieldQuery({ ...field, name: field.name })};`
    ]);
    logs.push(`\u26A0 Field Dropped - ${tableName}.${field.name}`);
  }
  if (needsTableChange) statements = [];
  for (const field of changes.create) {
    if (!sqliteSupportedColumnAdd(field)) needsTableChange = true;
    else {
      const tempName = field.name + "_" + randomString(5);
      toRename[tempName] = field.name;
      statements.push([
        `ALTER TABLE ${tableName}
                    ADD COLUMN ${createFieldQuery({ ...field, name: tempName })};`,
        `ALTER TABLE ${tableName} DROP COLUMN ${tempName};`
      ]);
      logs.push(`\u2714 Field Added - ${tableName}.${field.name}`);
    }
  }
  for (const [field, lastField] of changes.alter) {
    if (field.name !== lastField.name) {
      if (field.usage) throw new Error(`Field ${tableName}:${field.name} - Field cannot be renamed - ${field.usage} usage exists`);
      const fkCheck = others.find((t) => t.fields.find((f) => f.foreignKey && (f.foreignKey.table === tableName && f.foreignKey.column === lastField.name)));
      if (fkCheck) throw new Error(`Field ${tableName}:${lastField.name} - Field cannot be renamed - used in ${fkCheck.name}`);
      const tempName = field.name + "_" + randomString(5);
      toRename[tempName] = field.name;
      statements.push([
        `ALTER TABLE ${tableName} RENAME COLUMN ${lastField.name} TO ${tempName};`,
        `ALTER TABLE ${tableName} RENAME COLUMN ${tempName} TO ${lastField.name};`
      ]);
      renamedFieldNames[lastField.name] = field.name;
      logs.push(`\u26A0 Field Renamed - ${tableName}.${lastField.name} -> ${field.name}`);
    }
    const q1 = createFieldQuery({
      ...field,
      name: field.name
    });
    const renamedTables = Object.fromEntries(others.map((o) => {
      if (o.lastName) return [o.lastName, o.name];
      return [o.name, o.name];
    }));
    const q2 = createFieldQuery({
      ...lastField,
      name: field.name,
      foreignKey: lastField.foreignKey ? {
        ...lastField.foreignKey,
        table: renamedTables[lastField.foreignKey.table] ?? lastField.foreignKey.table
      } : void 0
    });
    if (q1 !== q2) {
      needsTableChange = true;
    }
  }
  for (const [tempName, actualName] of Object.entries(toRename)) {
    statements.push([
      `ALTER TABLE ${tableName} RENAME COLUMN ${tempName} TO ${actualName};`,
      `ALTER TABLE ${tableName} RENAME COLUMN ${actualName} TO ${tempName};`
    ]);
  }
  if (statements.length > 0) {
    clearIndexes = true;
  }
  if (needsTableChange) {
    const lastTable2 = {
      ...lastTable,
      name: tableName,
      fields: lastTable.fields.map((f) => ({
        ...f,
        name: renamedFieldNames[f.name] ? renamedFieldNames[f.name] : f.name
      }))
    };
    const existingFields = table.fields.map((f) => f.name).filter((f) => !changes.create.find((f1) => f1.name === f));
    const fieldsStr = existingFields.join(", ");
    statements.push([migrateTableQuery(table, fieldsStr), migrateTableQuery(lastTable2, fieldsStr)]);
    logs.push(`\u26A0 Recreate Table with new schema - ${tableName}`);
  }
  let dropIndexes_ = clearIndexes ? lastTable.indexes : changes.indexes.drop;
  let dropTriggers_ = clearTriggers ? lastTable.triggers : changes.triggers.drop;
  let createIndexes_ = needsTableChange ? [] : clearIndexes ? table.indexes : changes.indexes.create;
  let createTriggers_ = needsTableChange ? [] : clearTriggers ? table.triggers : changes.triggers.create;
  const createIndexes = createIndexes_?.map((index) => [
    createIndexQuery(table, index) + ";",
    dropIndexQuery(table, index) + ";"
  ]) || [];
  const dropIndexes = dropIndexes_?.map((index) => [
    dropIndexQuery(lastTable, index) + ";",
    createIndexQuery(lastTable, index) + ";"
  ]) || [];
  let createTriggers = createTriggers_?.map((trigger) => [
    createTriggerQuery(table, trigger) + ";",
    dropTriggerQuery(table, trigger) + ";"
  ]) || [];
  let dropTriggers = dropTriggers_?.map((trigger) => [
    dropTriggerQuery(lastTable, trigger) + ";",
    createTriggerQuery(lastTable, trigger) + ";"
  ]) || [];
  let dropFts = [];
  let createFts = [];
  if (resetFts && !needsTableChange) {
    const lastFtsq = createFts5Query(lastTable);
    const ftsq = createFts5Query(table);
    if (lastFtsq) dropFts.push([lastFtsq.sqlRevert, lastFtsq.sql]);
    if (ftsq) createFts.push([ftsq.sql, ftsq.sqlRevert]);
  }
  statements = [...dropTriggers, ...dropIndexes, ...dropFts, ...statements];
  logs = [
    // todo better logs
    ...dropTriggers.map((t) => `\u26A0 Trigger Dropped - ${t[0].split("TRIGGER")[1].split("ON")[0].trim()}`),
    ...dropIndexes.map((t) => `\u26A0 Index Dropped - ${t[0].split("INDEX")[1].split("ON")[0].split("IF EXISTS").pop().trim()}`),
    ...dropFts.map((t) => `\u26A0 FTS Dropped for ${lastTable.name}`),
    ...logs
  ];
  statements.push(...createIndexes, ...createTriggers, ...createFts);
  logs.push(
    ...createTriggers.map((t) => `\u2714 Trigger Added - ${t[0].split("TRIGGER")[1].split("ON")[0].trim()}`),
    ...createIndexes.map((t) => `\u2714 Index Added - ${t[0].split("INDEX")[1].split("ON")[0].trim()}`),
    ...createFts.map((t) => `\u2714 FTS Added - ${t[0].split("VIRTUAL TABLE")[1].split(";")[0].trim()}`)
  );
  return { sql: statements.map((s) => s[0]).join("\n"), sqlRevert: statements.map((s) => s[1]).join("\n"), logs };
}

// src/types/zod/tableExtensionsSchema.ts
import { z as z4 } from "zod";
var tableRulesDataSchema = z4.object({
  name: z4.literal("rules"),
  listRule: sqlExprSchema.nullable().default(null),
  viewRule: sqlExprSchema.nullable().default(null),
  createRule: sqlExprSchema.nullable().default(null),
  updateRule: sqlExprSchema.nullable().default(null),
  deleteRule: sqlExprSchema.nullable().default(null)
});
var emailTemplateSchema = z4.object({
  subject: z4.string().optional(),
  variables: z4.record(z4.any()).optional(),
  tags: z4.string().optional(),
  layoutHtml: z4.string().optional()
});
var tableAuthDataSchema = z4.object({
  name: z4.literal("auth"),
  jwtSecret: z4.string(),
  jwtTokenDuration: z4.number(),
  maxTokenRefresh: z4.number(),
  passwordType: z4.literal("sha256").default("sha256"),
  passwordConfirmSuffix: z4.string().optional(),
  passwordCurrentSuffix: z4.string().optional(),
  passwordResetTokenDuration: z4.number().optional(),
  emailVerifyTokenDuration: z4.number().optional(),
  passwordResetEmailDuration: z4.number().optional(),
  emailVerifyEmailDuration: z4.number().optional(),
  autoSendVerificationEmail: z4.coerce.boolean().optional(),
  emailTemplates: z4.object({
    verification: emailTemplateSchema.optional(),
    passwordReset: emailTemplateSchema.optional()
  }).optional()
  // onlyVerified: z.coerce.boolean().optional(),
  // minPasswordLength: z.number().optional(),
  // onlyEmailDomains: z.array(z.string()).nullable().optional(),
  // exceptEmailDomains: z.array(z.string()).nullable().optional(),
  // allowEmailAuth: z.coerce.boolean().optional(),
  // allowUsernameAuth: z.coerce.boolean().optional(),
  // allowOAuth2Auth: z.coerce.boolean().optional(),
  // manageRule: z.string().nullable().optional(),
});
var tableExtensionSchemas = {
  rules: tableRulesDataSchema,
  auth: tableAuthDataSchema
};

// src/sql/schema/createTable.ts
function createTable(data, others, _rnd) {
  data = tableDataSchema.parse(data);
  if (data.name !== data.name.toLowerCase()) throw new Error("Table name must be lowercase");
  if (others.find((t) => t.name === data.name)) throw new Error("Table already exists with this name " + data.name);
  const extensions = data.extensions;
  for (const extension of extensions) {
    const parser = tableExtensionSchemas[extension.name];
    const parsed = parser.parse(extension);
  }
  const query = createTableQuery(data, _rnd);
  return { sql: query.sql + query.sql2, sqlRevert: query.sqlRevert, logs: [`\u2714 Table Created - ${data.name}`] };
}

// src/types/zod/databaseSettingsSchema.ts
import { z as z8 } from "zod";

// src/types/zod/emailTemplatePropsSchema.ts
import { z as z5 } from "zod";
var baseTemplatePropsSchema = z5.object({
  company_name: z5.string(),
  company_url: z5.string(),
  company_address: z5.string(),
  company_copyright: z5.string(),
  support_email: z5.string()
});

// src/types/zod/mailgunBindingsSchema.ts
import { z as z6 } from "zod";
var mailgunBindingsSchema = z6.object({
  MAILGUN_API_KEY: z6.string(),
  MAILGUN_API_SERVER: z6.string(),
  MAILGUN_API_URL: z6.string().optional(),
  MAILGUN_WEBHOOK_SIGNING_KEY: z6.string().optional(),
  DISCORD_MAILGUN_NOTIFY_WEBHOOK: z6.string().optional(),
  EMAIL_BLOCKLIST: z6.string().optional()
});

// src/types/zod/resendBindingsSchema.ts
import { z as z7 } from "zod";
var resendBindingsSchema = z7.object({
  RESEND_API_KEY: z7.string(),
  RESEND_API_URL: z7.string().optional(),
  RESEND_WEBHOOK_SECRET: z7.string().optional(),
  MAILGUN_WEBHOOK_ID: z7.string().optional(),
  DISCORD_RESEND_NOTIFY_WEBHOOK: z7.string().optional(),
  EMAIL_BLOCKLIST: z7.string().optional()
});

// src/types/zod/databaseSettingsSchema.ts
var databaseSettingsSchema = z8.object({
  tables: z8.array(tableDataSchema),
  jwtSecret: z8.string(),
  jwtIssuer: z8.string().optional(),
  jwtAlgorithm: z8.string().optional(),
  version: z8.number().optional(),
  appName: z8.string().optional(),
  appUrl: z8.string(),
  procedures: z8.array(zSQLProcedure).optional(),
  email: z8.object({
    from: z8.string(),
    variables: baseTemplatePropsSchema,
    tags: z8.array(z8.string()).optional(),
    mailgun: mailgunBindingsSchema.optional(),
    resend: resendBindingsSchema.optional()
  }).optional(),
  _kvTableName: z8.string().startsWith("_", { message: "Internal table name must start with _ (underscore)" }).optional(),
  disableTablesEdit: z8.coerce.boolean().optional()
});

// src/utils/string.ts
var jsonStringify = (key) => JSON.stringify(
  key,
  (_, v) => v && v.constructor === Object ? Object.entries(v).sort() : v
);

// src/types/zod/dataTypesSchemas.ts
import { z as z9 } from "zod";
var tableFieldTypeToZod = {
  text: z9.string(),
  number: z9.number(),
  bool: z9.boolean(),
  email: z9.string().email(),
  url: z9.string().url(),
  editor: z9.string(),
  date: z9.string(),
  datetime: z9.string(),
  time: z9.string(),
  timestamp: z9.string(),
  json: z9.record(z9.any()),
  // file: z.instanceof(File),
  file: z9.string(),
  // relation: z.string(),
  // autodate: z.string(),
  integer: z9.number().int(),
  real: z9.number(),
  blob: z9.string(),
  float: z9.number(),
  int: z9.number().int(),
  boolean: z9.boolean(),
  numeric: z9.number()
};
var sqlDataTypeToDataTypeDefaults = {
  ["text" /* text */]: "text" /* text */,
  ["integer" /* integer */]: "integer" /* integer */,
  ["real" /* real */]: "number" /* number */,
  ["blob" /* blob */]: "blob" /* blob */,
  ["null" /* null */]: "text" /* text */,
  ["json" /* json */]: "json" /* json */,
  ["date" /* date */]: "date" /* date */,
  ["datetime" /* datetime */]: "date" /* date */,
  ["time" /* time */]: "date" /* date */,
  ["timestamp" /* timestamp */]: "date" /* date */,
  ["float" /* float */]: "number" /* number */,
  ["int" /* int */]: "integer" /* integer */,
  ["boolean" /* boolean */]: "bool" /* bool */,
  ["numeric" /* numeric */]: "number" /* number */
};
var dataTypeToSqlDataType = Object.fromEntries(Object.entries(sqlDataTypeToDataTypeDefaults).map(([k, v]) => [v, k]));
var sqlDataTypeAliases = {
  ["text" /* text */]: "text" /* text */,
  ["integer" /* integer */]: "integer" /* integer */,
  ["real" /* real */]: "real" /* real */,
  ["blob" /* blob */]: "blob" /* blob */,
  ["null" /* null */]: "null" /* null */,
  ["json" /* json */]: "text" /* text */,
  ["date" /* date */]: "text" /* text */,
  ["datetime" /* datetime */]: "text" /* text */,
  ["time" /* time */]: "text" /* text */,
  ["timestamp" /* timestamp */]: "text" /* text */,
  ["float" /* float */]: "real" /* real */,
  ["int" /* int */]: "integer" /* integer */,
  ["boolean" /* boolean */]: "integer" /* integer */,
  ["numeric" /* numeric */]: "real" /* real */
};
var supportedTypesForSql = {
  ["text" /* text */]: ["text" /* text */, "json" /* json */, "date" /* date */, "file" /* file */, "editor" /* editor */, "select" /* select */, "email" /* email */, "url" /* url */, "relation" /* relation */],
  ["integer" /* integer */]: ["number" /* number */, "integer" /* integer */, "bool" /* bool */, "select" /* select */, "relation" /* relation */],
  ["real" /* real */]: ["number" /* number */, "select" /* select */, "relation" /* relation */],
  ["blob" /* blob */]: ["blob" /* blob */],
  ["null" /* null */]: ["text" /* text */]
};

// src/types/config/validateFieldChange.ts
var allowFieldSchemaChange = true;
function validateFieldChange(field, lastField, tableName, settings) {
  if (!field) {
    if (!lastField) return {};
    if (lastField.usage) throw new Error(`Field ${tableName}:${lastField.name} - Field cannot be dropped - ${lastField.usage} usage exists`);
    const fkCheck = settings?.tables.find((t) => t.fields.find((f) => f.foreignKey && (f.foreignKey.table === tableName && f.foreignKey.column === lastField.name)));
    if (fkCheck) throw new Error(`Field ${tableName}:${lastField.name} - Field cannot be dropped - used in ${fkCheck.name}`);
    const ftCheck = settings?.tables.find((t) => t.name === tableName && t.fullTextSearch?.fields.includes(lastField.name));
    if (ftCheck) throw new Error(`Field ${tableName}:${lastField.name} - Field cannot be dropped - used in full text search`);
    return { drop: lastField };
  }
  if (lastField && jsonStringify(field) === jsonStringify(lastField)) return {};
  const fPrefix = `Table Field ${tableName}:${field.name} - `;
  field.sqlType = field.sqlType.toLowerCase();
  if (!field.type && !field.sqlType) {
    throw new Error(fPrefix + `type or sqlType is required`);
  }
  if (field.sqlType && !field.type) {
    field.type = sqlDataTypeToDataTypeDefaults[field.sqlType];
    if (!field.type) throw new Error(fPrefix + `Unable to determine type. Unknown type for sqlType - ${field.sqlType}`);
  }
  if (field.type && !field.sqlType) {
    field.sqlType = dataTypeToSqlDataType[field.type];
    if (!field.sqlType) throw new Error(fPrefix + `Unable to determine sqlType. Unknown sqlType for type - ${field.type}`);
  }
  const sqlBase = sqlDataTypeAliases[field.sqlType];
  if (!sqlBase) throw new Error(fPrefix + `Unknown sqlType - ${field.sqlType}`);
  const supportedTypes = supportedTypesForSql[sqlBase];
  if (!supportedTypes.includes(field.type)) {
    throw new Error(fPrefix + `Invalid type for type - ${field.type} not in [${supportedTypes.join(", ")}]`);
  }
  if (field.foreignKey && settings) {
    const table2 = settings.tables.find((t) => t.name === field.foreignKey.table);
    const field2 = table2?.fields.find((f) => f.name === field.foreignKey.column);
    if (!field2) throw new Error(fPrefix + `Foreign key field not found - ${field.foreignKey.table}.${field.foreignKey.column}`);
    if (field2.sqlType !== field.sqlType) throw new Error(fPrefix + `Foreign key field sqlType mismatch - ${field.sqlType} !== ${field2.sqlType}`);
  }
  if (!lastField) {
    if (field.default !== void 0) {
    }
    if (field.check) {
    }
    return { create: field };
  }
  if (field.sqlType !== lastField.sqlType) {
    if (!allowFieldSchemaChange) throw new Error(fPrefix + `Field sqlType cannot be changed - ${field.sqlType} !== ${lastField.sqlType}`);
  }
  if (field.primary !== lastField.primary) {
    if (!allowFieldSchemaChange) throw new Error(fPrefix + `Field primary cannot be changed - ${field.primary} !== ${lastField.primary}`);
  }
  if (field.autoIncrement !== lastField.autoIncrement) {
    if (!allowFieldSchemaChange) throw new Error(fPrefix + `Field autoIncrement cannot be changed - ${field.autoIncrement} !== ${lastField.autoIncrement}`);
  }
  if (field.unique !== lastField.unique) {
    if (!allowFieldSchemaChange) throw new Error(fPrefix + `Field unique cannot be changed - ${field.unique} !== ${lastField.unique}`);
  }
  if (field.notNull !== lastField.notNull) {
    if (!allowFieldSchemaChange) throw new Error(fPrefix + `Field notNull cannot be changed - ${field.notNull} !== ${lastField.notNull}`);
  }
  if (buildSql2(field.default) !== buildSql2(lastField.default)) {
    if (!allowFieldSchemaChange) throw new Error(fPrefix + `Field default cannot be changed - ${buildSql2(field.default)} !== ${buildSql2(lastField.default)}`);
  }
  if (buildSql2(field.check) !== buildSql2(lastField.check)) {
    if (!allowFieldSchemaChange) throw new Error(fPrefix + `Field default cannot be changed - ${buildSql2(field.check)} !== ${buildSql2(lastField.check)}`);
  }
  const foreignKey = jsonStringify(field.foreignKey);
  const lastForeignKey = jsonStringify(lastField.foreignKey);
  if (foreignKey !== lastForeignKey) {
    let trivial = false;
    if (field.foreignKey && lastField.foreignKey && field.foreignKey.table !== lastField.foreignKey.table) {
      const fkTable = settings?.tables.find((t) => t.name === field.foreignKey.table);
      if (!fkTable) throw new Error(fPrefix + `Foreign key table not found - ${field.foreignKey.table}`);
      if (fkTable.lastName === lastField.foreignKey.table) trivial = true;
    }
    if (!trivial) {
      if (!allowFieldSchemaChange)
        throw new Error(fPrefix + `Field foreignKey cannot be changed - ${foreignKey} !== ${lastForeignKey}`);
    }
  }
  const changes = jsonStringify(field) !== jsonStringify(lastField);
  return changes ? { alter: [field, lastField] } : {};
}
function buildSql2(q) {
  if (typeof q === "string") {
  }
  return q ? logSQLQuery(literalToQuery(typeof q === "string" ? { q } : q, false)) : void 0;
}

// src/types/config/validateTableChange.ts
function validateTableChange(nextTable, lastTable, nextSettings) {
  if (!nextTable) {
    if (!lastTable) return {};
    return { drop: lastTable };
  }
  if (lastTable && jsonStringify(nextTable) === jsonStringify(lastTable)) return {};
  const prefix = `Table ${nextTable.name} - `;
  if (lastTable && nextTable.name !== lastTable?.name) {
    if (nextTable.lastName !== lastTable?.name) throw new Error(prefix + `Last table name must be set to ${lastTable.name} when changing table name`);
    if (lastTable.r2Base && nextTable.r2Base !== lastTable.r2Base) {
      throw new Error(prefix + `r2Base cannot be changed - ${nextTable.r2Base} !== ${lastTable.r2Base}`);
    }
    if (!lastTable.r2Base && nextTable.r2Base !== lastTable.name) {
      throw new Error(prefix + `r2Base must be set to "${lastTable.name}" because changing table name`);
    }
  }
  if (lastTable && nextTable.r2Base !== lastTable.r2Base && nextTable.name !== lastTable.r2Base && nextTable.r2Base !== lastTable.name) {
    throw new Error(prefix + `r2Base cannot be changed - ${nextTable.r2Base ?? nextTable.name} !== ${lastTable.r2Base}`);
  }
  if (lastTable && nextTable.idInR2 !== lastTable.idInR2) {
    throw new Error(prefix + `idInR2 cannot be changed - ${nextTable.r2Base} !== ${lastTable.r2Base}`);
  }
  if (lastTable && nextTable.allowMultipleFileRef !== lastTable.allowMultipleFileRef) {
  }
  if (lastTable && nextTable.autoDeleteR2Files !== lastTable.autoDeleteR2Files) {
  }
  if (nextTable.idInR2) {
    const idField = nextTable.fields.find((f) => f.usage === "record_uid" /* record_uid */);
    if (!idField || !idField.noUpdate) throw new Error(prefix + `idInR2 requires an id field with 'usage' = 'record_uid' and 'noUpdate' = 'true'.`);
    if (nextTable.allowMultipleFileRef) {
      throw new Error(prefix + `allowMultipleFileRef cannot be true when idInR2 is true`);
    }
  }
  const autoDeleteR2Files = nextTable.autoDeleteR2Files ?? true;
  if (nextTable.allowMultipleFileRef) {
    if (nextTable.idInR2) {
      throw new Error(prefix + `idInR2 cannot be true when allowMultipleFileRef is true`);
    }
    if (autoDeleteR2Files) {
      throw new Error(prefix + `autoDeleteR2Files must be false when allowMultipleFileRef is true`);
    }
  }
  if (nextTable.allowWildcard) {
    throw new Error(prefix + `allowWildcard is not supported yet`);
  }
  if (nextTable.autoSetUid) {
    const idField = nextTable.fields.find((f) => f.usage === "record_uid" /* record_uid */);
    if (!idField || idField.type !== "text") throw new Error(prefix + `autoSetUid requires an id field with 'usage' = 'record_uid' and 'type' = 'text'.`);
  }
  const res = {
    create: [],
    drop: [],
    alter: [],
    indexes: { create: [], drop: [] },
    triggers: { create: [], drop: [] },
    fts: false
  };
  for (const field of nextTable.fields) {
    if (nextTable.fields.find((f) => f.name === field.name) !== field) throw new Error(`Field ${nextTable.name}:${field.name} - Duplicate field name`);
    let lastField = lastTable?.fields?.filter((f) => f.name === field.name || f.name === field.lastName) ?? [];
    lastField = lastField.filter((lf) => {
      const nextField = nextTable.fields.filter((f) => f.lastName === lf.name && f !== field);
      if (nextField.length > 1) throw new Error(`Field ${nextTable.name}:${lf.name} - Duplicate field name`);
      if (!nextField.length) return true;
      const lastField2 = lastTable?.fields?.filter((f) => f.name === nextField[0].name && f.name === nextField[0].lastName) ?? [];
      const renamed = !lastField2.length;
      return !renamed;
    });
    if (lastField.length > 1) {
      let cont = false;
      const exact = lastField.find((f) => f.name === field.name && f.lastName === field.lastName);
      if (exact) {
        lastField = [exact];
        cont = true;
      }
      if (!cont && field.lastName && lastField.length === 2) {
        const match = lastField.filter((f) => f.name === field.lastName);
        if (match.length === 1) {
          cont = true;
          lastField = match;
        }
      }
      if (!cont) throw new Error(`Field ${nextTable.name}:${field.name} - Duplicate field name`);
    }
    const res1 = validateFieldChange(field, lastField[0], nextTable.name, nextSettings);
    if (res1.create) res.create.push(res1.create);
    if (res1.drop) res.drop.push(res1.drop);
    if (res1.alter) res.alter.push(res1.alter);
  }
  for (const field of lastTable?.fields || []) {
    let nextField = nextTable.fields.filter((f) => f.name === field.name || f.lastName === field.name);
    if (nextField.length > 1) {
      let cont = false;
      if (nextField.length === 2) {
        nextField = nextField.filter((f) => f.lastName === field.name);
        if (nextField.length === 1) {
          if (res.create.includes(nextField[0]) || res.alter.find((a) => a[0] === nextField[0]))
            cont = true;
        }
      }
      if (!cont) throw new Error(`Field ${nextTable.name}:${field.name} - Duplicate field name`);
    }
    if (!nextField.length) {
      const res1 = validateFieldChange(void 0, field, nextTable.name);
      if (res1.create) res.create.push(res1.create);
      if (res1.drop) res.drop.push(res1.drop);
      if (res1.alter) res.alter.push(res1.alter);
    }
  }
  if (nextTable.indexes) {
    for (const index of nextTable.indexes) {
      const last = lastTable?.indexes?.find((i) => indexName(i).name === indexName(index).name);
      const changed = !last || jsonStringify(index) !== jsonStringify(last);
      if (changed) {
        res.indexes.create.push(index);
        if (last) res.indexes.drop.push(last);
      }
    }
  }
  if (lastTable?.indexes) {
    for (const index of lastTable?.indexes || []) {
      const next = nextTable.indexes?.find((i) => indexName(i).name === indexName(index).name);
      if (!next) {
        res.indexes.drop.push(index);
      }
    }
  }
  if (nextTable.triggers) {
    for (const trigger of nextTable.triggers) {
      const last = lastTable?.triggers?.find((i) => i.name === trigger.name);
      const changed = !last || jsonStringify(trigger) !== jsonStringify(last);
      if (changed) {
        res.triggers.create.push(trigger);
        if (last) res.triggers.drop.push(last);
      }
    }
  }
  if (lastTable?.triggers) {
    for (const trigger of lastTable?.triggers || []) {
      const next = nextTable.triggers?.find((i) => i.name === trigger.name);
      if (!next) {
        res.triggers.drop.push(trigger);
      }
    }
  }
  if (jsonStringify(nextTable.fullTextSearch ?? {}) !== jsonStringify(lastTable?.fullTextSearch ?? {})) {
    res.fts = true;
  }
  if (nextTable.fullTextSearch) {
    if (!nextTable.fullTextSearch.fields.length) throw new Error(`Table ${nextTable.name} - Full text search fields must be set`);
    const invalidFields = nextTable.fullTextSearch.fields.filter((f) => !nextTable.fields.find((f2) => f2.name === f));
    if (invalidFields.length) throw new Error(`Table ${nextTable.name} - Invalid full text search fields - ${invalidFields.join(", ")}`);
    if (nextTable.fullTextSearch.tokenize && !["unicode61", "ascii", "porter", "trigram"].includes(nextTable.fullTextSearch.tokenize)) {
      throw new Error(`Table ${nextTable.name} - Invalid tokenize value - ${nextTable.fullTextSearch.tokenize}`);
    }
    if (!nextTable.fullTextSearch.contentless && nextTable.fullTextSearch.content_rowid) {
      throw new Error(`Table ${nextTable.name} - content_rowid must not be set when contentless is false`);
    }
    if (nextTable.fullTextSearch.content_rowid) {
      const field = nextTable.fields.find((f) => f.name === nextTable.fullTextSearch.content_rowid);
      if (!field) throw new Error(`Table ${nextTable.name} - content_rowid field not found - ${nextTable.fullTextSearch.content_rowid}`);
      if (field.sqlType !== "integer") throw new Error(`Table ${nextTable.name} - content_rowid field must be integer`);
      if (field.foreignKey) throw new Error(`Table ${nextTable.name} - content_rowid field cannot have foreign key`);
    }
  }
  const extensions = nextTable.extensions;
  for (const extension of extensions) {
    const parser = tableExtensionSchemas[extension.name];
    const parsed = parser.parse(extension);
  }
  if (!lastTable) return { create: nextTable };
  return { alter: [nextTable, lastTable, res] };
}

// src/types/config/validateSettingsChange.ts
function validateSettingsChange(nextConfig, lastConfig) {
  nextConfig = databaseSettingsSchema.parse(nextConfig);
  if (nextConfig.version !== void 0) throw new Error(`Database settings version should not be set manually`);
  {
    const n = { ...nextConfig };
    delete n.version;
    delete n["//"];
    const l = { ...lastConfig };
    delete l.version;
    delete l["//"];
    if (jsonStringify(n) === jsonStringify(l)) return { config: null, changes: null };
  }
  nextConfig.version = (nextConfig?.version ?? -1) + 1;
  const res = { create: [], drop: [], alter: [] };
  for (const table of nextConfig.tables) {
    if (nextConfig.tables.find((t) => t.name === table.name) !== table) throw new Error(`Table ${table.name} - Duplicate table name found in nextConfig`);
    const lastTable = lastConfig?.tables?.filter((t) => t.name === table.name || t.name === table.lastName) ?? [];
    if (lastTable.length > 1) throw new Error(`Table ${table.name} - Duplicate table name found in lastConfig`);
    const res1 = validateTableChange(table, lastTable[0], nextConfig);
    if (res1.create) res.create.push(res1.create);
    if (res1.drop) res.drop.push(res1.drop);
    if (res1.alter) res.alter.push(res1.alter);
  }
  for (const table of lastConfig?.tables || []) {
    const nextTable = nextConfig.tables.filter((t) => t.name === table.name || t.lastName === table.name) ?? [];
    if (nextTable.length > 1) throw new Error(`Table ${table.name} - Duplicate table name found in nextConfig`);
    if (!nextTable.length) {
      const res1 = validateTableChange(void 0, table, nextConfig);
      if (res1.create) res.create.push(res1.create);
      if (res1.drop) res.drop.push(res1.drop);
      if (res1.alter) res.alter.push(res1.alter);
    }
  }
  return { config: nextConfig, changes: res };
}

// src/sql/schema/generateMigrations.ts
function generateMigrations(nextConfig, lastConfig, index = 0) {
  const { config, changes } = validateSettingsChange(nextConfig, lastConfig);
  if (!changes) return { migrations: [], config };
  const name = (op, table) => `${String(index++).padStart(4, "0")}_${op}_table_${table.name}.sql`;
  const others = (table) => config.tables.filter((t) => t.name !== table.name);
  const migrations = [
    ...changes.create.map((item) => ({
      name: name("create", item),
      ...createTable(item, others(item))
    })),
    ...changes.drop.map((item) => ({
      name: name("drop", item),
      ...dropTable(item, others(item))
    })),
    ...changes.alter.map((item) => ({
      name: name("alter", item[0]),
      ...alterTable(item, others(item[0]))
    }))
  ].filter((m) => m.sql.trim().length);
  const extraLogs = [];
  for (const table of config.tables) {
    if (table.lastName !== void 0) {
      extraLogs.push(`\u26A0 You need to Remove ${table.name}.lastName from config file for next migration`);
      delete table.lastName;
    }
    for (const field of table.fields) {
      if (field.lastName !== void 0) {
        extraLogs.push(`\u26A0 You need to Remove ${table.name}.fields.${field.name}.lastName field config file for next migration`);
        delete field.lastName;
      }
    }
  }
  return { migrations, config, extraLogs };
}

// src/node/cli.ts
var adminTokenKey = "ADMIN_SERVICE_TOKEN";
var apiRouteKey = "API_ROUTE";
var devVars = ".dev.vars";
var prodVars = ".prod.vars";
var filterDuplicateOptions = (options) => {
  if (options.local === options.remote) {
    throw new Error("Use either --local or --remote");
  }
  for (const [key, value] of Object.entries(options)) {
    if (Array.isArray(value)) {
      options[key] = value[value.length - 1];
    }
  }
};
var localPersist = ".local-persist";
var CliContext = class {
  constructor(root, options, wranglerConfig, devServerLogLevel = "none") {
    this.options = options;
    this.wranglerConfig = wranglerConfig;
    this.logger = createLogger(options.logLevel);
    this.root = normalizePath(root ? path4.resolve(root) : process.cwd());
    this.migrationsPath = path4.join(this.root, wranglerConfig.migrations);
    this.devServerLogLevel = devServerLogLevel;
    this.devServerPort = parseInt(process.env.TEENY_DEV_PORT || wranglerConfig.config.dev?.port || "8787");
    if (isNaN(this.devServerPort) || this.devServerPort <= 0 || this.devServerPort >= 65536) {
      throw new Error(`Invalid port number specified in TEENY_PORT environment variable: ${process.env.TEENY_PORT}`);
    }
  }
  localWorker;
  migrationsPath;
  get isLocal() {
    return !this.options.remote;
  }
  root;
  logger;
  devServerLogLevel;
  devServerPort;
  get localPersistPath() {
    const persistPath = path4.join(this.root, localPersist);
    return persistPath;
  }
  buildLocal(outDir, log = true, doThrow = true, clean = false, flags = "") {
    const _outDir = outDir ? (
      // explicit path is always resolved from cwd
      path4.resolve(outDir)
    ) : path4.join(this.root, ".teeny", "tmp", "dist-worker");
    const lastWLogEnv = process.env.WRANGLER_LOG;
    process.env.WRANGLER_LOG = log ? "info" : "error";
    try {
      const command = `npx wrangler deploy --dry-run ${flags} --outdir=${JSON.stringify(_outDir)}`;
      execSync(command, { cwd: this.root, stdio: ["inherit", "pipe", "pipe"] });
      return true;
    } catch (e) {
      const err = "" + e.stderr;
      const isBuildError = err.match(/Build failed with (\d+) errors:/g);
      const migrationsDir = readableRelative(this.migrationsPath);
      const configFile = path4.join(migrationsDir, "config.json");
      if (isBuildError?.length) {
        const configNotFound = !!err.match(new RegExp(`Could not resolve "${escapeRegExp(readableRelative(configFile))}"`, "gm"))?.length;
        if (configNotFound) {
          this.logger.error(`\u2718 [ERROR] Unable to build the worker - config.json not found in migrations folder. Run teeny generate or teeny migrate to generate migrations and corresponding config.json.`);
        }
        const isConfigError = !!err.match(new RegExp(`${escapeRegExp(configFile)}:\\d+:\\d+:`, "g"))?.length;
        if (isConfigError) {
          this.logger.error(`\u2718 [ERROR] Unable to build the worker - config.json is not valid. Run teeny generate or teeny migrate to generate migrations and corresponding config.json.`);
        }
        if (!log) {
          const errorCount = isBuildError ? parseInt(isBuildError[1]) : 0;
          this.logger.error(`\u2718 [ERROR] Unable to build the worker - ${errorCount} error(s). Run teeny build to see the errors`);
        }
      }
      if (log) {
        this.logger.info("" + e.stdout);
        if (!doThrow) this.logger.error("" + e.stderr);
      }
      if (doThrow) throw e;
      return false;
    } finally {
      process.env.WRANGLER_LOG = lastWLogEnv;
      if (clean) fs5.rmSync(_outDir, { recursive: true });
    }
  }
  _startingPromise;
  async startLocal() {
    if (this.localWorker) return this.localWorker;
    if (this._startingPromise) return await this._startingPromise;
    const built = this.buildLocal(void 0, false, false, true);
    if (!built) {
      throw new Error('Unable to build and connect to local worker. Run "teeny dev --local" to check the errors.');
    }
    if (this.options.local) this.logger.debug("Local worker starting");
    this._startingPromise = this.options.local ? unstable_dev(this.wranglerConfig.config.main, {
      moduleRoot: this.root,
      // config: wranglerConfig.config,
      port: this.devServerPort,
      local: true,
      bundle: true,
      persistTo: this.localPersistPath,
      logLevel: this.devServerLogLevel,
      //"none",
      // logLevel: "info",
      experimental: {
        disableExperimentalWarning: true
      }
    }) : void 0;
    this.localWorker = await this._startingPromise;
    if (this.localWorker) this.logger.debug("Local worker started");
    return this.localWorker;
  }
  async localFetch(input, init) {
    const w = await this.startLocal();
    if (!w) throw new Error("Unable to start local worker");
    return await w.fetch(input, init);
  }
  _stopping = false;
  async stopLocal() {
    if (!this.localWorker || this._stopping) return;
    this._stopping = true;
    await this.localWorker.stop();
    this.logger.debug("Local worker stopped");
    this.localWorker = void 0;
    this._stopping = false;
  }
};
function readMigrationsFolder(migrationsFolderPath) {
  if (!fs5.existsSync(migrationsFolderPath)) {
    fs5.mkdirSync(migrationsFolderPath);
  }
  const fileNames = fs5.readdirSync(migrationsFolderPath);
  let sqlNames = fileNames.filter((name) => name.endsWith(".sql"));
  let configFile = fileNames.find((name) => name === "config.json");
  let nextConfigFile = fileNames.find((name) => name === "next-config.json");
  sqlNames.sort((a, b) => {
    const aNumber = parseInt(a.split("_")[0]);
    const bNumber = parseInt(b.split("_")[0]);
    return aNumber - bNumber;
  });
  const currentMigrations = sqlNames.map((name) => {
    const migrationPath = path4.join(migrationsFolderPath, name);
    const migration = fs5.readFileSync(migrationPath, "utf8");
    return { name, sql: migration };
  });
  let currentConfig, nextConfig;
  const configPath = configFile ? path4.join(migrationsFolderPath, configFile) : null;
  const nextConfigPath = nextConfigFile ? path4.join(migrationsFolderPath, nextConfigFile) : null;
  try {
    currentConfig = configPath ? JSON.parse(fs5.readFileSync(configPath, "utf8")) : void 0;
  } catch (e) {
    throw new Error(`Unable to read file ${readableRelative(configPath || "")}
${e?.message || e}

This file is not supposed to be manually edited. You may revert any changes, or delete the "${readableRelative(migrationsFolderPath)}" and "${readableRelative(localPersist)}" directory and run again.`);
  }
  try {
    nextConfig = nextConfigPath ? JSON.parse(fs5.readFileSync(nextConfigPath, "utf8")) : void 0;
  } catch (e) {
    throw new Error(`Unable to read file ${readableRelative(nextConfigPath || "")}
${e?.message || e}

This file is not supposed to be manually edited. You may try deleting the file and run again`);
  }
  return { migrations: currentMigrations, config: currentConfig, nextConfig };
}
async function parseWrangler(root, options) {
  try {
    const projectRoot = normalizePath(root ? path4.resolve(root) : process.cwd());
    const wranglerConfig = await loadWranglerConfigFromFile(projectRoot, options.wrangler, options.logLevel);
    const d1s = wranglerConfig.config.d1_databases;
    if (!d1s?.length) throw new Error("No databases found in wrangler config");
    const d1Binding = options.db ?? "PRIMARY_DB";
    const d1 = d1s.find((d) => d.binding === (options.db || d1Binding));
    if (!d1) throw new Error(`Database binding ${d1Binding} not found in wrangler config. ` + (!options.db ? "Use --db to specify a different database binding" : ""));
    const migrationsDir = d1.migrations_dir;
    if (!migrationsDir) throw new Error("No migrations_dir found in wrangler config. Please specify a migrations_dir in wrangler config (like migrations)");
    return { config: wranglerConfig, db: d1, migrations: migrationsDir };
  } catch (e) {
    createLogger(options.logLevel).error(
      colors3.red(`error loading wrangler config:
${e.stack}`),
      { error: e }
    );
    process.exit(1);
  }
}
function readSecretsFile(c, throwError = true, apiRoute2) {
  const secretsPath = path4.join(c.root, c.isLocal ? devVars : prodVars);
  if (!fs5.existsSync(secretsPath)) {
    c.logger.error(`Create a file "${readableRelative(secretsPath)}" with the Admin token(${adminTokenKey}) and API route(${apiRouteKey} to connect to the worker`);
    if (throwError) throw new Error(`Secrets file not found at ${secretsPath}`);
    else return {};
  }
  const secrets = dotenv.parse(fs5.readFileSync(secretsPath, "utf8"));
  const adminToken = secrets[adminTokenKey];
  if (!adminToken && throwError) throw new Error(`Admin token(${adminTokenKey}) not found in secrets file at "${readableRelative(secretsPath)}"`);
  let apiRoute = (secrets[apiRouteKey] ?? apiRoute2)?.trim();
  if (apiRoute) {
    const lowerApiRoute = apiRoute.toLowerCase().trim();
    if (!lowerApiRoute || lowerApiRoute === "null" || lowerApiRoute === "undefined" || lowerApiRoute === "na" || lowerApiRoute === "none") {
      apiRoute = void 0;
    }
  }
  let parsedApiRoute;
  if (apiRoute) {
    try {
      parsedApiRoute = new URL2(apiRoute || "http://example.com");
    } catch (e) {
      throw new Error(`API route(${apiRouteKey}) in secrets file at "${readableRelative(secretsPath)}" is not a valid URL.`);
    }
    if (!["http:", "https:"].includes(parsedApiRoute.protocol)) {
      throw new Error(`API route(${apiRouteKey}) in secrets file at "${readableRelative(secretsPath)}" must start with http:// or https://`);
    }
  }
  if (!apiRoute && !c.isLocal && throwError) throw new Error(`API route(${apiRouteKey}) not found in secrets file at "${readableRelative(secretsPath)}"`);
  if (!c.isLocal && apiRoute?.startsWith("http://localhost")) {
    c.logger.warn(`API route(${apiRouteKey}) in secrets file at "${readableRelative(secretsPath)}" is pointing to localhost. This value will be ignored.
Use the remote API route or remove the variable if not yet deployed.`);
    apiRoute = void 0;
    delete secrets[apiRouteKey];
  }
  if (c.isLocal) {
    if (!apiRoute) {
      apiRoute = "http://localhost:" + c.devServerPort;
    } else if (parsedApiRoute && parsedApiRoute.port && parsedApiRoute.port !== "" + c.devServerPort && parsedApiRoute.hostname === "localhost") {
      c.logger.warn(`API route(${apiRouteKey}) in secrets file at "${readableRelative(secretsPath)}" has port ${parsedApiRoute.port} which is different from the local dev server port ${c.devServerPort}. Using port ${c.devServerPort} instead.`);
      apiRoute = `${parsedApiRoute.protocol}//${parsedApiRoute.hostname}:${c.devServerPort}`;
    }
  }
  return { adminToken, apiRoute, secrets };
}
async function fetchWorker(c, fPath, reqOptions, logging = false, apiRoute2, parseJson = true) {
  const { adminToken, apiRoute } = readSecretsFile(c, true, apiRoute2);
  const fOptions = {
    method: reqOptions?.method ?? "GET",
    ...reqOptions,
    headers: {
      "Authorization": `Bearer ${adminToken}`,
      ...reqOptions?.headers
    },
    signal: reqOptions?.signal ?? AbortSignal.timeout(5e3)
    // 5 second timeout
  };
  let _fetch = fetch;
  if (c.isLocal) {
    const devServerRunning = apiRoute && !c.localWorker ? await fetch(path4.join(apiRoute, "/api/v1/health"), {
      method: "GET",
      signal: AbortSignal.timeout(1e3)
    }).then((r) => r.ok).catch((e) => {
      return !!e && e.cause?.code !== "ECONNREFUSED" && e.name !== "TimeoutError";
    }) : false;
    if (devServerRunning && apiRoute) fPath = path4.join(apiRoute, fPath);
    else _fetch = c.localFetch.bind(c);
  } else {
    if (!apiRoute) throw new Error(`API route(${apiRouteKey}) not found in ${prodVars}`);
    fPath = path4.join(apiRoute, fPath);
  }
  if (logging) {
    c.logger.info(`Fetching ${fOptions.method.toUpperCase()}:${fPath}`);
    if (fOptions.body) c.logger.info(fOptions.headers?.["Content-Type"] === "application/json" ? JSON.parse(fOptions.body) : fOptions.body);
    if (fOptions.headers) c.logger.debug(fOptions.headers);
  } else {
    c.logger.debug(`Fetching ${fPath} ${JSON.stringify(fOptions)}`);
  }
  const res = await _fetch(fPath, fOptions);
  const rest = await res.text();
  if (!res.ok) {
    let message = rest;
    try {
      const json = JSON.parse(rest);
      message = (json.message ? json.message + " " : "") + ((json.error || json.data?.error) + "") || json;
      if (logging) c.logger.error(`Failed to fetch ${fPath}. Status: ${res.status}, Response: ${message}
${JSON.stringify(json, null, 2)}`);
    } catch (_e) {
    }
    throw new Error(`Failed to fetch ${fPath}. Status: ${res.status}, Response: ${typeof message === "string" ? message.slice(0, 1024) : message}`);
  }
  try {
    return rest ? parseJson ? JSON.parse(rest) : rest : null;
  } catch (e) {
    throw new Error(`Failed to parse response from ${fPath}. Status: ${res.status}, Response: ${rest}`);
  }
}
async function fetchDeployedMigrations(c) {
  c.logger.info(`Fetching deployed migrations from ${c.isLocal ? "local" : "remote"} deployment`);
  loadLocalPersistConfig(c);
  const json = await fetchWorker(c, "/api/v1/migrations", {}).catch((e) => {
    if (e.message?.includes("no such table: _db_migrations")) {
      c.logger.info("No migrations found in deployment");
      return { settings: null, migrations: [] };
    }
    throw e;
  });
  return json;
}
function loadLocalPersistConfig(c) {
  if (!c.isLocal) return void 0;
  const localConfigPath = path4.join(c.localPersistPath, "config.json");
  if (fs5.existsSync(localConfigPath)) {
    let localConfig;
    try {
      localConfig = JSON.parse(fs5.readFileSync(localConfigPath, "utf8"));
    } catch (e) {
      throw new Error(`Unable to read file ${readableRelative(localConfigPath || "")}
${e?.message || e}

This file is not supposed to be manually edited. You may revert any changes, or delete the "${readableRelative(c.localPersistPath)}" directory and run again.`);
    }
    if (!fs5.existsSync(c.migrationsPath)) fs5.mkdirSync(c.migrationsPath, { recursive: true });
    const configPath = path4.join(c.migrationsPath, "config.json");
    fs5.writeFileSync(configPath, JSON.stringify(localConfig, null, 2));
    c.logger.info(`Config found in ${readableRelative(c.localPersistPath)}. Written to ${readableRelative(configPath)}`);
    return localConfig;
  }
  return void 0;
}
async function generateMigrationsDir(c, clean = false, write = true) {
  c.logger.info(`Generating/Syncing Migrations folder: ${readableRelative(c.migrationsPath)}`);
  const config = await loadConfigFromFile(c.root, c.options.config, c.options.logLevel);
  let deployed = true;
  let lastConfig = void 0;
  let lastMigrations = [];
  let { migrations: currentMigrations, config: currentConfig, nextConfig: _nextConfig } = readMigrationsFolder(c.migrationsPath);
  if (c.isLocal && deployed) {
    currentConfig = loadLocalPersistConfig(c);
    if (!currentConfig) {
      c.logger.info(`No config found at ${readableRelative(path4.join(c.migrationsPath, "config.json"))} or ${readableRelative(c.localPersistPath)}`);
      deployed = false;
    }
  }
  if (!c.isLocal && deployed) {
    let res = readSecretsFile(c, false);
    if (!res.apiRoute) {
      c.logger.warn(`No API route(${apiRouteKey}) found in ${prodVars}. Assuming not deployed and generating migrations from scratch`);
      deployed = false;
    }
  }
  if (deployed) {
    const json = await fetchDeployedMigrations(c);
    if (json.settings === null) deployed = false;
    lastConfig = json.settings ? databaseSettingsSchema.parse(json.settings) : void 0;
    lastMigrations = json.migrations;
  } else {
  }
  if (!clean && write) {
    const migrationsChanged = jsonStringify(currentMigrations) !== jsonStringify(lastMigrations);
    const configChanged = jsonStringify(currentConfig || {}) !== jsonStringify(lastConfig || {});
    if (migrationsChanged) {
      throw new Error("Migrations folder has been modified. Apply the migrations first or run with --clean to reset migrations from deployment");
    }
    if (configChanged) {
      throw new Error("Config has been modified. Deploy first or run with --clean to reset migrations from deployment");
    }
    c.logger.info(`Migrations folder and config are up to date with the ${c.isLocal ? "local" : "remote"} deployment`);
  }
  const index = lastMigrations.length ? parseInt(lastMigrations[lastMigrations.length - 1].name.split("_")[0]) + 1 : 0;
  const { migrations: newMigrations, config: nextConfig_, extraLogs } = generateMigrations(config.config, lastConfig, index);
  const comment = `Config generated by teenybase on ${(/* @__PURE__ */ new Date()).toISOString()}. Do not modify this file.`;
  const nextConfig = nextConfig_ ? {
    "//": comment,
    ...nextConfig_
  } : void 0;
  if (clean) {
    if (!write) throw new Error("Cannot run with clean with no write");
    for (const m of currentMigrations) {
      fs5.unlinkSync(path4.join(c.migrationsPath, m.name));
    }
    if (currentConfig) {
      fs5.unlinkSync(path4.join(c.migrationsPath, "config.json"));
    }
    if (_nextConfig) {
      fs5.unlinkSync(path4.join(c.migrationsPath, "next-config.json"));
    }
    for (const m of lastMigrations) {
      fs5.writeFileSync(path4.join(c.migrationsPath, m.name), m.sql);
    }
    for (const m of currentMigrations) {
      if (lastMigrations.find((m1) => m1.name === m.name)) continue;
      if (fs5.existsSync(path4.join(c.migrationsPath, m.name))) {
        c.logger.warn(`Deleting extra file in ${c.migrationsPath} - ${m.name}`);
        fs5.unlinkSync(path4.join(c.migrationsPath, m.name));
      }
    }
    if (lastConfig) {
      fs5.writeFileSync(path4.join(c.migrationsPath, "config.json"), JSON.stringify(lastConfig, null, 2));
    }
    c.logger.info("Migrations folder cleaned" + (deployed ? " and reset to deployment" : ""));
  } else {
  }
  if (!nextConfig) {
    c.logger.info("No changes detected in database config");
    return;
  }
  if (write) {
    for (const { name, sql, logs } of newMigrations) {
      const migrationPath = path4.join(c.migrationsPath, name);
      const comment2 = `-- Migration number: ${name.split("_")[0]} generated by teenybase on ${(/* @__PURE__ */ new Date()).toISOString()}. Do not modify this file.`;
      fs5.writeFileSync(migrationPath, [comment2, sql].join("\n\n"));
      c.logger.info(`New migration generated - ${name}`);
      for (const log of logs) {
        c.logger.info(" - " + log);
      }
      c.logger.info("");
    }
    if (!newMigrations.length) {
      c.logger.info("No new migrations generated");
    }
    const configPath = path4.join(c.migrationsPath, "next-config.json");
    fs5.writeFileSync(configPath, JSON.stringify(nextConfig, null, 2));
    c.logger.info(`Config generated and written to - ${readableRelative(configPath)}`);
    if (extraLogs?.length) {
      for (const log of extraLogs) {
        c.logger.warn(log);
      }
    }
  }
  return { newMigrations, nextConfig, lastMigrations, lastConfig, currentConfig };
}
function wranglerDeploy(c, first = false) {
  if (c.isLocal) throw new Error("Cannot deploy with --local");
  try {
    const res = execSync(`wrangler deploy`, { cwd: c.root, stdio: first ? ["inherit", "pipe", "inherit"] : "inherit" });
    if (!first) return;
    const resB = "" + res;
    c.logger.info(resB);
    let route = resB.match(/https:\/\/[a-zA-Z0-9-.]+\.workers\.dev\r?\n.*Current Version ID:/gm)?.[0].split("\n")[0];
    return route;
  } catch (e) {
    const err = "" + e.stdout + e.stderr;
    c.logger.error(err);
    const bucketNotFound = err.includes("workers.api.error.bucket_not_found [code: 10085]");
    if (bucketNotFound) {
      c.logger.error("Bucket specified in wrangler.toml is not found. Create a bucket in cloudflare using wrangler and set it in wrangler.toml");
    }
    throw new Error("Unable to deploy worker, check logs above for more details.");
  }
}
function deploySecrets(c, secrets) {
  if (c.isLocal) throw new Error("Cannot deploy with --local");
  const tempSecretsPath = path4.join(c.root, ".teeny", ".temp-secrets.json");
  try {
    const json = secrets;
    fs5.mkdirSync(path4.dirname(tempSecretsPath), { recursive: true });
    fs5.writeFileSync(tempSecretsPath, JSON.stringify(json, null, 2));
    const res = execSync(`wrangler secret bulk ${JSON.stringify(tempSecretsPath)}`, { cwd: c.root, stdio: "inherit" });
  } catch (e) {
    throw new Error(`Unable to deploy secrets from ${prodVars}, check logs above for more details.`);
  } finally {
    if (fs5.existsSync(tempSecretsPath)) fs5.unlinkSync(tempSecretsPath);
  }
}
async function migrateAndDeploy(c, clean, doMigrate, forceDeploy, skipConfirmation = false) {
  const { newMigrations, nextConfig } = await generateMigrationsDir(c, clean).catch(async (e) => {
    throw e;
  }) || {};
  if (newMigrations?.length) {
    if (!doMigrate) {
      throw new Error("There are unapplied migrations. Run with --migrate to apply them");
    }
  }
  let newSettingsSavedInDb = false;
  const { migrations, config: lastConfig, nextConfig: _nextConfig } = readMigrationsFolder(c.migrationsPath);
  try {
    if (JSON.stringify(nextConfig || {}) !== JSON.stringify(_nextConfig || {})) {
      throw new Error("Unknown error: nextConfig does not match locally");
    }
    if (nextConfig) {
      c.logger.info("Updating config.json with next-config.json");
      fs5.writeFileSync(path4.join(c.migrationsPath, "config.json"), JSON.stringify(nextConfig, null, 2));
    }
    const shouldDeploy = !!nextConfig || newMigrations?.length || forceDeploy;
    if (shouldDeploy) {
      const workerName = c.wranglerConfig.config.config.name;
      const cfAccountId = c.wranglerConfig.config.config.account_id;
      if (!workerName) {
        throw new Error("Required - No worker name(name) found in wrangler config");
      }
      if (!cfAccountId) {
        throw new Error("Required - No cloudflare account id(cfAccountId) found in wrangler config");
      }
      let currentDeployments = [];
      let deployRoute = void 0;
      if (!c.isLocal) {
        const commandCD = `wrangler deployments list --json`;
        try {
          c.logger.debug("Fetching current deployments");
          const cdRes = execSync(commandCD, {
            cwd: c.root,
            stdio: ["inherit", "pipe", "pipe"]
          }).toString();
          currentDeployments = JSON.parse(cdRes);
        } catch (e) {
          const err = "" + e?.stderr + e?.stdout;
          const noDeployment = err.includes("This Worker does not exist on your account. [code: 10007]");
          if (!noDeployment) {
            c.logger.error(`Unknown error when fetching previous deployments of the worker ${workerName} in cloudflare account ${cfAccountId}. Please try again later.`);
            throw e;
          } else {
          }
        }
      }
      if (!lastConfig && nextConfig) {
        if (!c.isLocal) {
          const envSec = readSecretsFile(c, false);
          if (!envSec.apiRoute && currentDeployments.length) {
            throw new Error(`Worker is already deployed. API route(${apiRouteKey}) not found in secrets file(${c.isLocal ? devVars : prodVars}).`);
          }
          if (envSec.apiRoute && !currentDeployments.length) {
            throw new Error(`Worker is not deployed. API route(${apiRouteKey}) found in secrets file(${c.isLocal ? devVars : prodVars}). Remove it and rerun to deploy a new worker.`);
          }
          if (currentDeployments.length) {
            const message = `Worker is already deployed but no settings and migrations found at ${envSec.apiRoute}`;
            if (!skipConfirmation) {
              const res = await prompts({
                type: "confirm",
                name: "value",
                message: message + "\nDo you want to continue and re-deploy the worker and apply migrations from scratch?",
                initial: false
              });
              if (!res.value) {
                throw new Error(`Unknown error - Worker is already deployed but no settings and migrations found at ${envSec.apiRoute}`);
              }
            } else {
              c.logger.info(message);
              c.logger.info("Continuing with re-deployment (--yes flag provided)");
            }
          } else {
            const message = `About to deploy the worker for the first time and upload all the secrets in ${prodVars} with the name ${workerName} in cloudflare account ${cfAccountId}, continue?
Make sure 'wrangler login' has been run before continuing. This will run 'wrangler deploy' and 'wrangler secret bulk'. `;
            if (!skipConfirmation) {
              const res = await prompts({
                type: "confirm",
                name: "value",
                message,
                initial: true
              });
              if (!res.value) {
                throw new Error("First time deployment cancelled by user");
              }
            } else {
              c.logger.info(message);
              c.logger.info("Continuing with first time deployment (--yes flag provided)");
            }
            try {
              c.logger.info("Deploying worker for the first time");
              const configNoTables = {
                ...nextConfig,
                tables: []
              };
              fs5.writeFileSync(path4.join(c.migrationsPath, "config.json"), JSON.stringify(configNoTables, null, 2));
              let route = wranglerDeploy(c, true);
              if (!route) {
                const customDomain = c.wranglerConfig.config.config.routes.find((r) => r.custom_domain)?.pattern;
                if (customDomain) route = `https://${customDomain}`;
              }
              if (!route) {
                c.logger.error("Unknown error after deploying worker. Deployment route not found.");
              } else deployRoute = route;
            } catch (e) {
              throw e;
            } finally {
              fs5.writeFileSync(path4.join(c.migrationsPath, "config.json"), JSON.stringify(nextConfig, null, 2));
            }
            try {
              deploySecrets(c, envSec.secrets);
            } catch (e) {
              c.logger.error(`Unknown error when deploying worker, secrets during first deployment. Cannot continue. First, run "teeny secrets-deploy" then set the ${apiRouteKey} in ${prodVars} and rerun the command to migrate.`);
              throw e;
            }
          }
        } else {
          if (nextConfig) {
            const localConfigPath = path4.join(c.localPersistPath, "config.json");
            fs5.mkdirSync(path4.dirname(localConfigPath), { recursive: true });
            fs5.writeFileSync(localConfigPath, JSON.stringify(nextConfig, null, 2));
          }
        }
      }
      if (nextConfig || newMigrations?.length) {
        if (!c.isLocal && !currentDeployments.length && !deployRoute) {
          c.logger.info("The worker has been deployed for the first time but deployment route not found. Cannot continue with the migration");
          c.logger.info(`Set the ${apiRouteKey} in ${prodVars} and rerun the command to migrate.`);
          throw new Error("Deployment route not found");
        }
        await fetchWorker(c, "/api/v1/setup-db", {
          // todo handle database settings mismatch error from this endpoint. we need to first deploy the worker with the settings in the db and then apply migrations.
          method: "POST",
          headers: {
            "Content-Type": "application/json"
          }
        }, false, deployRoute);
        if (newMigrations?.length) {
          const message = `About to apply ${newMigrations.length} migration(s)
 Your database may not be available to serve requests during the migration, continue?`;
          if (!skipConfirmation) {
            const res = await prompts({
              type: "confirm",
              name: "value",
              message,
              initial: false
            });
            if (!res.value) {
              throw new Error("Migration cancelled by user");
            }
          } else {
            c.logger.info(message);
            c.logger.info("Continuing with migration (--yes flag provided)");
          }
        }
        c.logger.info(`Syncing ${migrations.length} migrations, with ${newMigrations?.length || 0} new migrations`);
        try {
          const json = await fetchWorker(c, "/api/v1/migrations", {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({ migrations, settings: nextConfig || lastConfig })
            // this will also save the settings in the db
          }, false, deployRoute);
          if (!json.applied?.length) {
            if (newMigrations?.length) {
              throw new Error(`Unable to apply ${newMigrations.length} new migrations`);
            }
            c.logger.warn("No migrations applied");
          } else {
            c.logger.info(`Applied ${json.applied.length} migrations successfully`);
            c.logger.info(json.applied.map((m) => ` \u2705 ${m}`).join("\n"));
            if (newMigrations?.length !== json.applied.length) {
              const unapplied = newMigrations?.filter((m) => !json.applied.includes(m.name)) ?? [];
              c.logger.error(unapplied.map((m) => ` \u274C ${m.name}`).join("\n"));
              throw new Error(`Unknown error - Only ${json.applied.length} of ${newMigrations?.length} new migrations applied`);
            }
          }
          newSettingsSavedInDb = true;
        } catch (e) {
          throw e;
        }
      }
      if (true) {
        if (!c.isLocal) {
          wranglerDeploy(c);
        } else {
          if (nextConfig) {
            const localConfigPath = path4.join(c.localPersistPath, "config.json");
            fs5.mkdirSync(path4.dirname(localConfigPath), { recursive: true });
            fs5.writeFileSync(localConfigPath, JSON.stringify(nextConfig, null, 2));
          }
        }
      }
      if (nextConfig) {
        c.logger.info("Removing next-config.json");
        fs5.unlinkSync(path4.join(c.migrationsPath, "next-config.json"));
      }
      if (!c.isLocal && !currentDeployments.length) {
        c.logger.info(`
`);
        c.logger.info(`The worker has been deployed for the first time at ${deployRoute}`);
        c.logger.info(`Set the ${apiRouteKey} to "${deployRoute}" in "${prodVars}" file to connect to the worker from next time.`);
      }
    } else {
      if (!c.isLocal)
        c.logger.info("Skipping worker deployment. No changes detected in database config and migrations. Run with --deploy to deploy anyway");
    }
  } catch (e) {
    if (nextConfig && !newSettingsSavedInDb) {
      c.logger.warn("Reverting config.json due to error");
      const cp = path4.join(c.migrationsPath, "config.json");
      if (lastConfig) {
        fs5.writeFileSync(cp, JSON.stringify(lastConfig, null, 2));
      } else if (fs5.existsSync(cp)) {
        fs5.unlinkSync(cp);
      }
      if (c.isLocal && lastConfig) {
        const localConfigPath = path4.join(c.localPersistPath, "config.json");
        fs5.mkdirSync(path4.dirname(localConfigPath), { recursive: true });
        fs5.writeFileSync(localConfigPath, JSON.stringify(lastConfig, null, 2));
      }
    }
    throw e;
  }
}
process.env.WRANGLER_SEND_METRICS = "false";
process.env.WRANGLER_LOG = "log";
var cli = cac("teeny");
cli.option("-c, --config <file>", `[string] use specified config file`).option("-w, --wrangler <wrangler>", `[string] use specified wrangler config file`).option("--db <db-name>", `[string] d1 database binding in wrangler (default: 'PRIMARY_DB')`).option("--local", `[boolean] use local database (default: true)`).option("--remote", `[boolean] use remote database (default: false)`).option("-l, --logLevel <level>", `[string] info | warn | error | silent`).option("--clearScreen", `[boolean] allow/disable clear screen when logging`).option("-d, --debug [feat]", `[string | boolean] show debug logs`).option("-f, --filter <filter>", `[string] filter debug logs`).option("--root <root>", `[string] project root (default: cwd)`);
cli.command("generate", "generate migrations").option(
  "--clean",
  `[boolean] delete migrations folder and download latest from deployment before generating (default: true)`
).action(
  async (options) => {
    filterDuplicateOptions(options);
    const wranglerConfig = await parseWrangler(options.root, options);
    const c = new CliContext(options.root, options, wranglerConfig);
    await generateMigrationsDir(c, Boolean(options.clean ?? true)).catch(async (e) => {
      c.logger.error(
        colors3.red(`error when generating migrations:
${e.stack}`),
        { error: e }
      );
      await c.stopLocal();
      process.exit(1);
    }).finally(() => {
    });
    await c.stopLocal();
  }
);
cli.command("migrate", "apply migrations and deploy worker").option(
  "--deploy",
  `[boolean] deploy the worker even if there are no teenybase config changes (default: false)`
).option(
  "--migrate",
  `[boolean] apply new migrations. If disabled, this will throw an error if there are new/unapplied migrations (default: true)`
).option(
  "--clean",
  `[boolean] delete migrations folder and download latest from deployment before generating (default: true)`
).option(
  "-y, --yes",
  `[boolean] skip confirmation prompts (default: false)`
).action(
  async (options) => {
    filterDuplicateOptions(options);
    const wranglerConfig = await parseWrangler(options.root, options);
    const c = new CliContext(options.root, options, wranglerConfig);
    const forceDeploy = Boolean(options.deploy ?? false);
    const doMigrate = Boolean(options.migrate ?? true);
    const clean = Boolean(options.clean ?? true);
    const skipConfirm = Boolean(options.y || options.yes);
    await migrateAndDeploy(c, clean, doMigrate, forceDeploy, skipConfirm).catch(async (e) => {
      c.logger.error(
        colors3.red(`error when applying migrations and deploying:
${e.stack}`),
        { error: e }
      );
      await c.stopLocal();
      process.exit(1);
    });
    await c.stopLocal();
  }
);
cli.command("deploy", "deploy worker and optionally apply migrations. This is same as migrate command but with different defaults").option(
  "--deploy",
  `[boolean] deploy the worker even if there are no teenybase config and migrations changes (default: true)`
).option(
  "--migrate",
  `[boolean] apply new migrations. If enabled, this will prompt to apply new migrations otherwise this will throw an error if there are new/unapplied migrations (default: false)`
).option(
  "--clean",
  `[boolean] delete migrations folder and download latest from deployment before generating (default: true)`
).option(
  "-y, --yes",
  `[boolean] skip confirmation prompts (default: false)`
).action(
  async (options) => {
    filterDuplicateOptions(options);
    const wranglerConfig = await parseWrangler(options.root, options);
    const c = new CliContext(options.root, options, wranglerConfig);
    const forceDeploy = Boolean(options.deploy ?? true);
    const doMigrate = Boolean(options.migrate ?? true);
    const clean = Boolean(options.clean ?? true);
    const skipConfirm = Boolean(options.y || options.yes);
    await migrateAndDeploy(c, clean, doMigrate, forceDeploy, skipConfirm).catch(async (e) => {
      c.logger.error(
        colors3.red(`error when applying migrations and deploying:
${e.stack}`),
        { error: e }
      );
      await c.stopLocal();
      process.exit(1);
    });
    await c.stopLocal();
  }
);
cli.command("secrets", "deploy worker and optionally apply migrations. This is same as migrate command but with different defaults").option(
  "--upload",
  `[boolean] deploy the worker even if there are no teenybase config and migrations changes (default: false)`
).action(
  async (options) => {
    filterDuplicateOptions(options);
    const wranglerConfig = await parseWrangler(options.root, options);
    const c = new CliContext(options.root, options, wranglerConfig);
    const doUpload = Boolean(options.upload ?? false);
    const envRes = readSecretsFile(c, false);
    const secretKeys = [...Object.keys(envRes.secrets ?? {})];
    if (secretKeys.length) {
      c.logger.info(`${secretKeys.length} values found in ${c.isLocal ? devVars : prodVars}`);
      secretKeys.forEach((k) => {
        c.logger.info(` - ${k}`);
      });
      c.logger.info("");
    }
    if (!doUpload) {
      c.logger.warn("Run with --upload to deploy secrets to the worker");
      return;
    }
    if (c.isLocal) {
      c.logger.error("No need to deploy secrets in local development, they are automatically read by the dev server");
      return;
    }
    const workerName = c.wranglerConfig.config.config.name;
    const cfAccountId = c.wranglerConfig.config.config.account_id;
    if (!workerName) {
      throw new Error("Required - No worker name(name) found in wrangler config");
    }
    if (!cfAccountId) {
      throw new Error("Required - No cloudflare account id(cfAccountId) found in wrangler config");
    }
    const message = `About to upload(create/replace) ${secretKeys.length} secrets to the worker ${workerName} in cloudflare account ${cfAccountId}?
Make sure 'wrangler login' has been run before continuing. 
This will run 'wrangler secret bulk'. Do you want to continue? `;
    const res = await prompts({
      type: "confirm",
      name: "value",
      message,
      initial: true
    });
    if (!res.value) {
      throw new Error("Secrets upload cancelled by user");
    }
    try {
      deploySecrets(c, envRes.secrets);
    } catch (e) {
      c.logger.error(
        colors3.red(`error when applying migrations and deploying:
${e.stack}`),
        { error: e }
      );
      await c.stopLocal();
      process.exit(1);
    }
    await c.stopLocal();
  }
);
cli.command("dev", "start wrangler dev server").action(
  async (options) => {
    filterDuplicateOptions(options);
    const logger = createLogger(options.logLevel);
    const wranglerConfig = await parseWrangler(options.root, options);
    const c = new CliContext(options.root, options, wranglerConfig);
    try {
      if (!options.local) {
        throw new Error("Remote dev server not support at the moment.");
      }
      loadLocalPersistConfig(c);
      const configPath = path4.join(c.migrationsPath, "config.json");
      if (!fs5.existsSync(c.migrationsPath) || !fs5.existsSync(configPath)) {
        throw new Error(`Migrations or Config file not found. First run "teeny migrate ${c.isLocal ? "--local" : "--remote"}" to generate and apply migrations to the ${c.isLocal ? "local" : "remote"}" database.`);
      }
      const command = `npx wrangler dev --persist-to ${JSON.stringify(c.localPersistPath)} --port ${c.devServerPort}`;
      execSync(command, { cwd: c.root, stdio: "inherit" });
      await c.stopLocal();
    } catch (e) {
      logger.error(
        colors3.red(`error when running wrangler dev:
${e.stack}`),
        { error: e }
      );
      await c.stopLocal();
      process.exit(1);
    } finally {
    }
  }
);
cli.command("backup", "backup database and schema").action(
  async (options) => {
    filterDuplicateOptions(options);
    const wranglerConfig = await parseWrangler(options.root, options);
    const c = new CliContext(options.root, options, wranglerConfig);
    try {
      const dir = `db_backups/${options.local ? "local" : "remote"}/${wranglerConfig.db.binding}/${(/* @__PURE__ */ new Date()).toISOString()}`;
      const absDir = path4.join(c.root, dir);
      fs5.mkdirSync(absDir, { recursive: true });
      const migrations = await fetchDeployedMigrations(c);
      if (migrations.settings)
        fs5.writeFileSync(path4.join(absDir, "config.json"), JSON.stringify(migrations.settings, null, 2));
      if (migrations.migrations.length) {
        const historyDir = path4.join(absDir, "history");
        fs5.mkdirSync(historyDir, { recursive: true });
        for (const m of migrations.migrations) {
          fs5.writeFileSync(path4.join(historyDir, m.name), m.sql);
        }
      }
      if (options.local) {
        await exportLocal(wranglerConfig.config.config, wranglerConfig.db.binding, path4.join(absDir, "schema.sql"), [], false, true, c.localPersistPath, c.logger);
        await exportLocal(wranglerConfig.config.config, wranglerConfig.db.binding, path4.join(absDir, "data.sql"), [], true, false, c.localPersistPath, c.logger);
      } else {
        const command = `npx wrangler d1 export ${wranglerConfig.db.binding} ${options.local ? `--local --persist-to ${JSON.stringify(c.localPersistPath)}` : "--remote"} --no-data --output ${JSON.stringify(path4.join(absDir, "schema.sql"))}`;
        const command2 = `npx wrangler d1 export ${wranglerConfig.db.binding} ${options.local ? `--local --persist-to ${JSON.stringify(c.localPersistPath)}` : "--remote"} --no-schema --output ${JSON.stringify(path4.join(absDir, "data.sql"))}`;
        execSync(command, { cwd: c.root, stdio: "inherit" });
        execSync(command2, { cwd: c.root, stdio: "inherit" });
      }
      await c.stopLocal();
    } catch (e) {
      c.logger.error(
        colors3.red(`error when exporting backend backup:
${e.stack}`),
        { error: e }
      );
      await c.stopLocal();
      process.exit(1);
    } finally {
    }
  }
);
cli.command("build", "build for production").option("--outDir <dir>", `[string] output directory (default: dist)`).action(
  async (options) => {
    filterDuplicateOptions(options);
    const wranglerConfig = await parseWrangler(options.root, options);
    const c = new CliContext(options.root, options, wranglerConfig);
    const { nextConfig } = await generateMigrationsDir(c, false, false).catch(async (e) => {
      c.logger.error(
        colors3.red(`error when generating migrations:
${e.stack}`),
        { error: e }
      );
      await c.stopLocal();
      process.exit(1);
    }).finally(() => {
    }) || {};
    c.logger.info("Starting Build");
    if (!fs5.existsSync(c.migrationsPath)) fs5.mkdirSync(c.migrationsPath, { recursive: true });
    const configFile = path4.join(c.migrationsPath, "config.json");
    const configData = fs5.existsSync(configFile) ? fs5.readFileSync(configFile, "utf8") : null;
    try {
      if (nextConfig) {
        c.logger.debug("Temporarily writing next-config to config.json");
        fs5.writeFileSync(configFile, JSON.stringify(nextConfig, null, 2));
      } else {
      }
      const outDir = options.outDir || path4.join(c.root, "dist");
      const res = c.buildLocal(outDir, true, true, false, "--minify");
      if (res) c.logger.info(`Finished Build. Written to "${readableRelative(outDir)}"`);
      await c.stopLocal();
    } catch (e) {
      createLogger(options.logLevel).error(
        colors3.red(`error during build:
${e.stack}`),
        { error: e }
      );
      await c.stopLocal();
      process.exit(1);
    } finally {
      if (nextConfig) {
        c.logger.debug("Reverting config.json");
        if (!configData) {
          if (fs5.existsSync(configFile)) fs5.unlinkSync(configFile);
        } else fs5.writeFileSync(configFile, configData);
      }
    }
  }
);
cli.command("exec <route>", "execute a route on the worker with admin token").option("-m, --method <method>", `[string] http method (default: GET)`).option("-d, -b, --data, --body <body>", `[string] http body`).option("-y, --yes", `[boolean] skip confirmation. only for non-GET requests (default: false)`).option("-r, --raw", `[boolean] output raw response (default: false)`).option("--explain", `[boolean] dry run and return the queries that will be executed (default: false)`).action(
  async (route, options) => {
    filterDuplicateOptions(options);
    const wranglerConfig = await parseWrangler(options.root, options);
    const c = new CliContext(options.root, options, wranglerConfig, "warn");
    try {
      if (!options.yes && options.method && options.method?.toUpperCase() !== "GET") {
        const res2 = await prompts({
          type: "confirm",
          name: "value",
          message: `Requesting "${route}" on the worker. This might change the database state.
Are you sure you want to continue?`,
          initial: true
        });
        if (!res2.value) {
          throw new Error("Execution cancelled by user");
        }
      }
      const routePath = path4.join(
        !options.explain ? "/api/v1/table/" : "/api/v1/explain/table/",
        route
      );
      const res = await fetchWorker(c, routePath, {
        method: options.method || "GET",
        body: options.body,
        headers: {
          "Content-Type": "application/json"
        }
      }, !options.raw, void 0, !options.raw);
      if (!options.raw) {
        c.logger.info("Request completed. Response:");
        c.logger.info(res);
      } else {
        console.log(res);
      }
      await c.stopLocal();
    } catch (e) {
      createLogger(options.logLevel).error(
        colors3.red(`error during exec:
${e.stack}`),
        { error: e }
      );
      await c.stopLocal();
      process.exit(1);
    } finally {
    }
  }
);
cli.help();
cli.version(VERSION);
cli.parse();
