// teenybase/dist/worker/email/mailgun.js
import { HTTPException } from 'hono/http-exception';
import { discordNotify } from './discordNotify';
import { z } from 'zod';
import { pathJoin } from '../../utils/helpers';
import { checkBlocklist } from './block-list';
const emailBlockList = 'yopmail.com,mailinator.com,guerrillamail.com,sharklasers.com,maildrop.cc';
export class MailgunHelper {
    baseUrl = 'https://api.mailgun.net/v3/';
    bindings;
    constructor(bindings) {
        this.bindings = bindings;
        if (bindings.MAILGUN_API_URL?.startsWith('https://'))
            this.baseUrl = bindings.MAILGUN_API_URL;
        else if (bindings.MAILGUN_API_URL)
            throw new HTTPException(400, { message: 'Invalid mailgun configuration - MAILGUN_API_URL must start with https://' });
        if (!this.bindings.MAILGUN_API_SERVER)
            throw new HTTPException(400, { message: 'Invalid mailgun configuration - missing MAILGUN_API_SERVER or MAILGUN_API_KEY' });
    }
    async sendEmail({ from, to, subject, html, tags }) {
        checkBlocklist(to, this.bindings.EMAIL_BLOCKLIST);
        const form = new FormData();
        form.append('from', from);
        form.append('to', to);
        form.append('subject', subject);
        form.append('html', html);
        // form.append('h:X-Mailgun-Variables', JSON.stringify(variables))
        tags.forEach(tag => form.append('o:tag', tag));
        const _key = this.bindings.MAILGUN_API_KEY;
        const key = typeof _key === 'string' || !_key ? _key : await _key();
        if (!key)
            throw new HTTPException(500, { message: 'Invalid mailgun configuration - missing MAILGUN_API_KEY' });
        let res = await fetch(pathJoin([this.baseUrl, this.bindings.MAILGUN_API_SERVER, '/messages']), {
            method: 'POST',
            headers: {
                'Authorization': 'Basic ' + btoa('api:' + key),
                // don't add content-type header for multipart
            },
            body: form
        }).catch(e => {
            console.error('Failed to fetch mailgun', e);
            return new Response('Failed to fetch mailgun - ' + e.message || e, { status: 500 });
        });
        let resJson = undefined;
        let resp = '';
        try {
            resp = await res.text();
            resJson = JSON.parse(resp);
        }
        catch (e) {
            console.error('Failed to parse mailgun response', e);
            // resp = e?.message || e
        }
        if (!resJson) {
            console.error('Error sending email, ', res?.status, resp);
            if (this.bindings.DISCORD_MAILGUN_NOTIFY_WEBHOOK)
                await discordNotify(this.bindings.DISCORD_MAILGUN_NOTIFY_WEBHOOK, `Failed to send email(\`${res.status}\`) to \`${to}\` \n\`\`\`${resp}\`\`\``);
            throw new HTTPException(500, { message: 'Failed to send email' });
        }
        return resJson;
    }
    async receiveWebhook(data) {
        // https://help.mailgun.com/hc/en-us/articles/202236504-Webhooks
        // If Mailgun receives a 200 (Success) code it will determine the webhook POST is successful and not retry.
        // If Mailgun receives a 406 (Not Acceptable) code, Mailgun will determine the POST is rejected and not retry.
        // For any other code, Mailgun will retry POSTing according to the following schedule (other than the delivery notification): 10 minutes, 10 minutes, 15 minutes, 30 minutes, 1 hour, 2 hour and 4 hours.
        const _key = this.bindings.MAILGUN_WEBHOOK_SIGNING_KEY;
        const key = typeof _key === 'string' || !_key ? _key : await _key();
        const signature = data.signature;
        const verified = (!key || !signature) ? false : await verify({ signingKey: key, ...signature });
        let message = '';
        if (!verified) {
            console.error('Invalid Webhook Signature', JSON.stringify(data));
            message += '[Invalid Webhook Signature]\n';
        }
        const eventData = data['event-data'];
        // mailgun is sending empty GET request for some reason...
        if (!eventData && !verified) {
            if (this.bindings.DISCORD_MAILGUN_NOTIFY_WEBHOOK)
                await discordNotify(this.bindings.DISCORD_MAILGUN_NOTIFY_WEBHOOK, `Empty webhook notification from mailgun, make sure https is used in webhook. https://app.mailgun.com/mg/sending/${this.bindings.MAILGUN_API_SERVER}/logs`);
            return;
        }
        if (!eventData)
            throw new HTTPException(400, { message: 'Invalid mailgun webhook data' });
        const timestamp = eventData.timestamp ?? Date.now();
        const logLevel = eventData['log-level'];
        const { event, recipient } = eventData;
        message += `${logLevel?.toUpperCase() || 'UNKNOWN'} - ${event?.toUpperCase() || 'UNKNOWN'} at ${new Date(parseInt(timestamp) * 1000).toISOString()}\n`;
        if (event.toLowerCase() === 'failed') {
            message += `Severity: ${eventData.severity}\n`;
            message += `Reason: ${eventData.reason}\n`;
        }
        if (recipient)
            message += `Recipient: ${recipient}\n`;
        if (logLevel !== 'info') {
            const json = JSON.stringify(eventData, null, 2);
            const res = this.bindings.DISCORD_MAILGUN_NOTIFY_WEBHOOK ? await discordNotify(this.bindings.DISCORD_MAILGUN_NOTIFY_WEBHOOK, message, [new File([json], 'event.json', {
                    type: 'application/json',
                })]) : false; // todo - notify by email if no discord?
            if (!res) {
                logLevel === 'warn' ? console.warn(message, json) : console.error(message, json);
                if (this.bindings.DISCORD_MAILGUN_NOTIFY_WEBHOOK)
                    throw new HTTPException(500, { message: 'Failed to notify destination' });
            }
        }
        else {
            // todo batch cache
            console.log(message);
        }
        if (!verified)
            throw new HTTPException(406, { message: 'Invalid mailgun webhook signature' + key ? '' : ', no key to verify' });
    }
    getRoutes(c) {
        const path = '/mailgun/webhook/:wid?';
        const handler = {
            raw: async (params) => {
                const wid = params?.wid;
                if ((c.env.MAILGUN_WEBHOOK_ID || wid) && wid !== c.env.MAILGUN_WEBHOOK_ID)
                    throw new HTTPException(404, { message: 'Not found' });
                let body;
                try {
                    body = JSON.parse((await c.req.text()) || '{}');
                    // todo - validation
                    // body = zMailgunWebhookData.parse(body)
                }
                catch (e) {
                    throw new HTTPException(400, { message: 'Invalid mailgun webhook data' });
                }
                await this.receiveWebhook(body);
                return c.json({ message: 'Received' }, 200);
            }
        };
        const zod = () => ({
            description: 'Mailgun webhook. Webhook can send email errors/events to this endpoint',
            request: {
                headers: z.object({}),
                params: z.object({ wid: z.string().optional().describe('Optional id for the webhook.') }),
                body: {
                    required: true, content: { 'application/json': { schema: zMailgunWebhookData } },
                },
            },
            responses: {
                '200': {
                    description: 'Success',
                    content: { 'application/json': { schema: z.object({ message: z.literal('Received') }) } },
                },
                '400': { description: 'Invalid mailgun webhook data, retry in some time.' },
                '406': { description: 'Invalid mailgun webhook signature, don\'t retry' },
                '500': { description: 'Failed to notify destination or other error, retry in some time.' },
            },
        });
        // mailgun makes get request for validation and post for actual events
        return [{ method: 'get', path, handler, zod }, { method: 'post', path, handler, zod }];
    }
}
export const zMailgunWebhookData = z.object({
    signature: z.object({
        timestamp: z.string(),
        token: z.string(),
        signature: z.string()
    }),
    'event-data': z.record(z.string()) // todo could it be anything else?
});
async function verify({ signingKey, timestamp, token, signature }) {
    const encoder = new TextEncoder();
    const keyData = encoder.encode(signingKey);
    const message = encoder.encode(timestamp + token);
    const key = await crypto.subtle.importKey('raw', keyData, { name: 'HMAC', hash: 'SHA-256' }, false, ['sign']);
    const signatureBuffer = await crypto.subtle.sign('HMAC', key, message);
    const signatureArray = Array.from(new Uint8Array(signatureBuffer));
    const encodedToken = signatureArray.map(b => b.toString(16).padStart(2, '0')).join('');
    return encodedToken === signature;
}
