// teenybase/dist/worker/email/resend.js
import { HTTPException } from 'hono/http-exception';
import { checkBlocklist } from './block-list';
import { discordNotify } from './discordNotify';
import { z } from '@hono/zod-openapi';
export class ResendHelper {
    baseUrl = 'https://api.resend.com/emails';
    bindings;
    constructor(bindings) {
        this.bindings = bindings;
        if (bindings.RESEND_API_URL?.startsWith('https://'))
            this.baseUrl = bindings.RESEND_API_URL;
        else if (bindings.RESEND_API_URL)
            throw new HTTPException(400, { message: 'Invalid Resend configuration' });
        if (!this.bindings.RESEND_API_KEY) {
            throw new Error('Invalid Resend configuration - missing RESEND_API_KEY');
        }
        // if (!this.bindings.RESEND_WEBHOOK_SECRET) {
        //     throw new Error('Invalid Resend configuration - missing RESEND_WEBHOOK_SECRET');
        // }
    }
    async sendEmail({ from, to, subject, html, tags }) {
        to = Array.isArray(to) ? to : [to];
        to.forEach((recipient) => checkBlocklist(recipient, this.bindings.EMAIL_BLOCKLIST));
        const payload = { from, to, subject, html };
        if (tags)
            payload.tags = tags;
        const _key = this.bindings.RESEND_API_KEY;
        const key = typeof _key === 'string' || !_key ? _key : await _key();
        if (!key)
            throw new HTTPException(500, { message: 'Invalid resend configuration - missing RESEND_API_KEY' });
        const res = await fetch(this.baseUrl, {
            method: 'POST',
            headers: {
                Authorization: `Bearer ${key}`,
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(payload),
        }).catch(e => {
            console.error('Failed to fetch resend', e);
            return new Response('Failed to fetch resend - ' + e.message || e, { status: 500 });
        });
        let resJson = undefined;
        let resp = '';
        try {
            resp = await res.text();
            resJson = JSON.parse(resp);
        }
        catch (e) {
            console.error('Failed to parse resend response', e);
            // resp = e?.message || e
        }
        if (!resJson || !res.ok) {
            console.error('Error sending email, ', res?.status, resp);
            if (this.bindings.DISCORD_RESEND_NOTIFY_WEBHOOK)
                await discordNotify(this.bindings.DISCORD_RESEND_NOTIFY_WEBHOOK, `Failed to send email(\`${res.status}\`) to \`${to}\` \n\`\`\`${resp}\`\`\``);
            throw new HTTPException(500, { message: 'Failed to send email' });
        }
        return resJson;
    }
    async receiveWebhook(headers, body) {
        let data;
        try {
            data = JSON.parse(body);
        }
        catch (e) {
            console.error('Failed to parse resend webhook data', e);
            throw new HTTPException(400, { message: 'Invalid resend webhook data' });
        }
        const _key = this.bindings.RESEND_WEBHOOK_SECRET;
        const key = typeof _key === 'string' || !_key ? _key : await _key();
        const verified = (!key || !data) ? false : await verify({ headers, body, secret: key });
        let message = '';
        if (!verified) {
            console.error('Invalid Webhook Signature', body);
            message += '[Invalid Webhook Signature]\n';
        }
        const parsed = zResendWebhookData.safeParse(data);
        let eventData;
        if (!parsed.success) {
            console.error('Invalid Webhook Data', parsed.error);
            message += `[Invalid Webhook Data] ${parsed.error}\n`;
            eventData = data;
        }
        else {
            eventData = parsed.data;
        }
        // console.log('Webhook event received:', event);
        // {
        //   "type": "email.bounced",
        //   "created_at": "2024-11-22T23:41:12.126Z",
        //   "data": {
        //     "created_at": "2024-11-22T23:41:11.894719+00:00",
        //     "email_id": "56761188-7520-42d8-8898-ff6fc54ce618",
        //     "from": "Acme <onboarding@resend.dev>",
        //     "to": ["delivered@resend.dev"],
        //     "subject": "Sending this example"
        //     "bounce": {
        //       "message": "The recipient's email address is on the suppression list because it has a recent history of producing hard bounces.",
        //       "subType": "Suppressed",
        //       "type": "Permanent",
        //     }
        //   }
        // }
        const timestamp = headers['svix-timestamp'] ?? Date.now();
        const event = eventData.data;
        const recipient = (Array.isArray(event.to) ? event.to : [event.to]).join(', ');
        message += `${eventData.type} at ${new Date(parseInt(timestamp) * 1000).toISOString()}\n`;
        if (recipient)
            message += `Recipient: ${JSON.stringify(eventData.data.to)}\n`;
        const notifyEvents = [
            'email.bounced',
            'email.complained',
            'email.delivery_delayed',
            'contact.deleted',
            'domain.deleted',
        ]; // todo param
        // Handle specific event types
        if (notifyEvents.includes(eventData.type)) {
            const res = this.bindings.DISCORD_RESEND_NOTIFY_WEBHOOK ? await discordNotify(this.bindings.DISCORD_RESEND_NOTIFY_WEBHOOK, message, [new File([body], 'event.json', {
                    type: 'application/json',
                })]) : false; // todo - notify by email if no discord?
            if (!res) {
                console.error(message, event);
                if (this.bindings.DISCORD_RESEND_NOTIFY_WEBHOOK)
                    throw new HTTPException(406, { message: 'Failed to notify destination' });
            }
        }
        else {
            console.log('Unhandled event type:', eventData.type);
        }
    }
    getRoutes(c) {
        const path = '/resend/webhook/:wid?';
        const handler = {
            raw: async (params) => {
                const wid = params?.wid;
                if ((c.env.RESEND_WEBHOOK_ID || wid) && wid !== c.env.RESEND_WEBHOOK_ID)
                    throw new HTTPException(404, { message: 'Not found' });
                let body;
                try {
                    body = await c.req.text();
                    // todo - validation
                    // body = zResendWebhookData.parse(body)
                }
                catch (e) {
                    throw new HTTPException(400, { message: 'Invalid resend webhook data' });
                }
                await this.receiveWebhook(c.req.header(), body);
                return c.json({ message: 'Received' }, 200);
            }
        };
        const zod = () => ({
            description: 'Resend webhook. Webhook can send email errors/events to this endpoint',
            request: {
                headers: z.object({}),
                params: z.object({ wid: z.string().optional().describe('Optional id for the webhook.') }),
                body: {
                    required: true, content: { 'application/json': { schema: zResendWebhookData } },
                },
            },
            responses: {
                '200': {
                    description: 'Success',
                    content: { 'application/json': { schema: z.object({ message: z.literal('Received') }) } },
                },
                '400': { description: 'Invalid resend webhook data, retry in some time.' },
                '406': { description: 'Invalid resend webhook signature, don\'t retry' },
                '500': { description: 'Failed to notify destination or other error, retry in some time.' },
            },
        });
        // resend makes get request for validation and post for actual events
        return [{ method: 'get', path, handler, zod }, { method: 'post', path, handler, zod }];
    }
}
const zResendWebhookData = z.object({
    type: z.string(),
    created_at: z.string(),
    data: z.object({
        created_at: z.string(),
        email_id: z.string(),
        from: z.string(),
        to: z.union([z.string(), z.array(z.string())]),
        subject: z.string().optional(),
        bounce: z.object({
            message: z.string().optional(),
            subType: z.string().optional(),
            type: z.string().optional(),
        }).optional()
    }),
});
async function verify({ headers, body, secret }) {
    const svixId = headers['svix-id'];
    const svixTimestamp = headers['svix-timestamp'];
    const svixSignature = headers['svix-signature'];
    if (!svixId || !svixTimestamp || !svixSignature) {
        console.error('Invalid webhook headers', headers);
        return false;
    }
    const signedContent = `${svixId}.${svixTimestamp}.${body}`;
    const encoder = new TextEncoder();
    // const signature = crypto
    //     .createHmac('sha256', secretBytes)
    //     .update(signedContent)
    //     .digest('base64');
    // const secretBytes = Buffer.from(secret.split('_')[1], 'base64');
    const keyData = encoder.encode(atob(secret.split('_')[1]));
    const message = encoder.encode(signedContent);
    const key = await crypto.subtle.importKey('raw', keyData, { name: 'HMAC', hash: 'SHA-256' }, false, ['sign']);
    const signatureBuffer = await crypto.subtle.sign('HMAC', key, message);
    const b64encoded = btoa(String.fromCharCode(...new Uint8Array(signatureBuffer)));
    return b64encoded === svixSignature;
}
