// teenybase/dist/worker/email/send-email.js
import { MailgunHelper } from './mailgun';
import { replaceTemplateVariables } from '../util/replaceTemplateVariables';
import { HTTPException } from 'hono/http-exception';
import { baseLayout1 } from './templates/base-layout-1';
import { messageLayout1 } from './templates/message-layout-1';
import { actionLinkTemplate } from './templates/action-link';
import { actionTextTemplate } from './templates/action-text';
import { $DBExtension } from '../$DBExtension';
import { ResendHelper } from './resend';
export function buildEmailTemplate(templates) {
    return templates.reduce((acc, t) => acc.replace('{{EMAIL_CONTENT}}', t), '{{EMAIL_CONTENT}}');
}
export class EmailSendClient extends $DBExtension {
    props;
    mg;
    rd;
    constructor(db, props, mg, rd) {
        super(db);
        this.props = props;
        this.mg = mg ? new MailgunHelper(mg) : null;
        this.rd = rd ? new ResendHelper(rd) : null;
        // todo also make a webhook that logs everything to an r2 bucket with encryption
        this.mg && this.routes.push(...this.mg.getRoutes(this.db.c));
        this.rd && this.routes.push(...this.rd.getRoutes(this.db.c));
    }
    templates = {
        actionLink: [baseLayout1, messageLayout1, actionLinkTemplate],
        actionText: [baseLayout1, messageLayout1, actionTextTemplate],
    };
    sendActionLink(prop) {
        if (!prop.variables.action_link)
            throw new Error('action_link is required');
        if (!prop.variables.action_text)
            prop.variables.action_text = 'Click here';
        return this.sendEmail({
            html: buildEmailTemplate(this.templates.actionLink),
            ...prop,
            tags: ['action-link', ...(prop.tags || [])],
        });
    }
    sendActionText(prop) {
        if (!prop.variables.action_text)
            throw new Error('action_text is required');
        return this.sendEmail({
            html: buildEmailTemplate(this.templates.actionLink),
            ...prop,
            tags: ['action-text', ...(prop.tags || [])],
        });
    }
    sendEmail(prop) {
        const props = {
            ...this.props,
            ...prop,
            variables: {
                ...this.props.variables,
                ...prop.variables,
            },
            tags: [...(this.props.tags || []), ...(prop.tags || [])],
        };
        if (!props.html)
            throw new Error('html is required');
        if (!props.subject)
            throw new Error('html is required');
        if (props.variables) {
            props.html = replaceTemplateVariables(props.html, props.variables, 3);
            props.subject = replaceTemplateVariables(props.subject, props.variables, 2);
            delete props.variables;
        }
        if (!props.to)
            throw new Error('to is required');
        if (!props.subject)
            throw new Error('subject is required');
        if (!props.from)
            throw new Error('from is required');
        if (this.mg) {
            return this.mg.sendEmail({
                from: props.from,
                html: props.html,
                subject: props.subject,
                tags: props.tags || [],
                to: props.to,
            });
        }
        else if (this.rd) {
            return this.rd.sendEmail({
                from: props.from,
                html: props.html,
                subject: props.subject,
                tags: props.tags.map(t => ({ name: t, value: 'true' })) || [],
                to: props.to,
            });
        }
        else
            throw new HTTPException(500, { message: 'Email provider not configured' });
    }
}
