// teenybase/dist/sql/parse/select.js
import { parseColumnList, parseColumnListOrder, } from './jsep';
import { queryToSqlQuery } from './parse';
import { selectSchema } from '../../types/zod/sqlSchemas';
export function parseSelectQuery(q, jc) {
    const selectData = selectSchema.parse(q);
    const select = {
        selects: selectData.select ? parseColumnList(selectData.select, jc, true, true, true) : undefined,
        from: jc.tableName,
        where: selectData.where ? queryToSqlQuery(selectData.where, jc) : undefined,
        groupBy: selectData.group ? parseColumnList(selectData.group, jc, false, false) : undefined,
        orderBy: selectData.order ? parseColumnListOrder(selectData.order, jc) : undefined,
        limit: selectData.limit,
        offset: selectData.offset,
        distinct: selectData.distinct,
        // having: selectData.having ? queryToD1Query(selectData.having, table, c) : undefined,
        // selectOption: selectData
        // join
    };
    let readOnly = !select.selects || select.selects?.every(s => typeof s === 'string' || s._readOnly);
    readOnly = readOnly && !!(Array.isArray(select.where) ? select.where.every(s => !s.q || s._readOnly) : (!select.where?.q || select.where?._readOnly));
    // todo having
    if (readOnly)
        select._readOnly = true;
    return select;
}
