// teenybase/dist/worker/$Table.js
import { extensions } from './extensions';
import { appendJoin, appendOrderBy, getSubSelectQueries } from '../sql/build/select';
import { HTTPException } from 'hono/http-exception';
import { parseUpdateQuery } from '../sql/parse/update';
import { parseInsertQuery } from '../sql/parse/insert';
import { parseSelectQuery } from '../sql/parse/select';
import { parseDeleteQuery } from '../sql/parse/delete';
import { z } from 'zod';
import { generateUid } from '../security/random';
import { ProcessError } from './util/error';
import { TableCrudExtension } from './extensions/tableCrudExtention';
import { columnify, ident, uncolumnify } from '../sql/parse/jsep';
import { LinearRouter } from 'hono/router/linear-router';
import { normalizeFileName } from './util/string';
import { fts5TableName } from '../sql/schema/tableQueries';
import { sqlValSchemaRecord } from '../types/zod/sqlSchemas';
import { tableFieldTypeToZod } from '../types/zod/dataTypesSchemas';
export class $Table {
    data;
    jc;
    $db;
    extensions = [];
    fileFields;
    // jsonFields?: string[]
    // editorFields?: string[]
    mapping;
    fields;
    /**
     * Maps usage strings to field name/array of field names(for multiple)
     */
    fieldsUsage;
    autoDeleteR2Files;
    allowMultipleFileRef;
    allowWildcard; // this is allow direct wildcard in select/returning instead of expanding it automatically. todo maybe this should be renamed
    extension(key) {
        const ext = this.extensions.find(e => e.name === key);
        if (!ext)
            throw new ProcessError('Extension not found', 500);
        return ext;
    }
    get name() {
        return this.data.name;
    }
    // get id(){
    //     return this.data.id
    // }
    constructor(data, jc, $db) {
        this.data = data;
        this.jc = jc;
        this.$db = $db;
        this.fieldsUsage = data.fields.reduce((acc, f) => {
            if (!f.usage)
                return acc;
            let v = acc[f.usage];
            if (!v)
                v = f.name;
            else if (Array.isArray(v))
                v.push(f.name);
            else
                v = [v, f.name];
            acc[f.usage] = v;
            return acc;
        }, {});
        // todo generic error for any non array usages
        if (Array.isArray(this.fieldsUsage.record_uid))
            throw new ProcessError('Multiple fields for record_uid not supported', 500);
        if (Array.isArray(this.fieldsUsage.record_created))
            throw new ProcessError('Multiple fields for record_created not supported', 500);
        if (Array.isArray(this.fieldsUsage.record_updated))
            throw new ProcessError('Multiple fields for record_updated not supported', 500);
        this.mapping = {
            uid: z.string().optional().parse(this.fieldsUsage.record_uid), // uid. this is for unique id(used in view route, auth etc).
            created: z.string().optional().parse(this.fieldsUsage.record_created), // created
            updated: z.string().optional().parse(this.fieldsUsage.record_updated), // updated
        };
        this.allowWildcard = data.allowWildcard ?? false;
        this.fields = Object.fromEntries(this.data.fields.map(f => [f.name, f]));
        this.fileFields = this.data.fields.filter(f => f.type === 'file').map(f => f.name);
        this.extensions.push(new TableCrudExtension({ name: 'crud' }, this, jc));
        for (const ext of data.extensions) {
            const extClass = extensions[ext.name];
            // todo dont pass $db.c here, remove dep from TableAuthExtension
            if (extClass)
                this.extensions.push(new extClass(ext, this, jc, $db.c));
            else
                throw new ProcessError('Invalid extension ' + ext.name + ' for table ' + data.name, 500);
        }
        if (this.data.idInR2 && (!this.mapping.uid || !this.fields[this.mapping.uid].noUpdate))
            throw new ProcessError('Invalid Configuration for idInR2 - uid field must be present and set to noUpdate', 500);
        this.autoDeleteR2Files = this.data.autoDeleteR2Files !== undefined ? this.data.autoDeleteR2Files : true;
        this.allowMultipleFileRef = this.data.allowMultipleFileRef !== undefined ? this.data.allowMultipleFileRef : false;
        if (this.allowMultipleFileRef && this.data.idInR2)
            throw new ProcessError('Invalid Configuration for idInR2 - allowMultipleFileRef cannot be true', 500);
        if (this.allowMultipleFileRef && this.autoDeleteR2Files)
            throw new ProcessError('Invalid Configuration for autoDeleteR2Files - allowMultipleFileRef cannot be true', 500);
    }
    // setup db etc
    async setup(version) {
        // for (const ext of this.extensions) {
        //     ext.setup && ext.setup()
        // }
        return this;
    }
    // todo why is initialize required?
    // if it needs to be initialized then wait for it in any function call.
    initialize() {
        for (const ext of this.extensions) {
            ext.initialize && ext.initialize();
        }
        return this;
    }
    // crud
    async edit(data, id) {
        return await (await this.rawEdit(data, id))?.run();
    }
    async view(data, id) {
        return await (await this.rawView(data, id))?.run();
    }
    async select(data, countTotal) {
        return typeof countTotal === 'boolean' ?
            await (await this.rawSelect(data, countTotal))?.run() :
            await (await this.rawSelect(data, countTotal))?.run();
    }
    async selectCount(data) {
        return await (await this.rawSelectCount(data))?.run();
    }
    async selectRaise(data) {
        return await (await this.rawSelectRaise(data))?.run();
    }
    async update(data, then) {
        return await (await this.rawUpdate(data, then))?.run();
    }
    async insert(data, then) {
        return await (await this.rawInsert(data, then))?.run();
    }
    async delete(data, then) {
        return await (await this.rawDelete(data, then))?.run();
    }
    async rawDelete(data, then) {
        let query;
        try {
            query = parseDeleteQuery(data, this.jc);
        }
        catch (e) {
            throw new ProcessError('Error parsing DELETE data', 400, { input: data }, e);
        }
        await this.onDeleteParse(query);
        // const filesToDelete = await this.filesToDelete(query)
        const fileFields = this.fileFields;
        const res = this.$db.rawDelete(this, query, fileFields, then);
        return res;
    }
    async rawInsert(data, then) {
        // console.error('insert', data)
        const values = Array.isArray(data.values) ? data.values : data.values ? [data.values] : undefined;
        let filesToUpload = {};
        let filesToRef = [];
        const data2 = {
            ...data,
            values: []
        };
        // todo - make beforeInsertParse
        if (values) {
            for (const setValues of values) {
                if (!setValues)
                    continue;
                this.checkNewRecordValues(setValues);
                const { values: _values } = this.filesToUpload(setValues, filesToUpload, filesToRef);
                data2.values.push(_values);
            }
        }
        let query;
        try {
            query = parseInsertQuery(data2, this.jc);
        }
        catch (e) {
            throw new ProcessError('Error parsing INSERT data', 400, data, e);
        }
        await this.onInsertParse(query);
        const values0 = Array.isArray(data2.values) ? data2.values[0] : data2.values;
        const fileFields = values0 && this.fileFields?.filter(f => values0[f] !== undefined);
        const res = this.$db.rawInsert(this, query, filesToUpload, filesToRef, fileFields, then);
        return res;
    }
    async rawUpdate(data, then) {
        const set = data.set;
        if (set && this.fileFields)
            for (const [key, val] of Object.entries(set)) {
                if (this.fileFields.includes(key))
                    throw new ProcessError('File fields can only be set using setValues');
            }
        // todo - make beforeUpdateParse
        this.checkNewRecordValues(data.setValues);
        const { filesToUpload, filesToRef, values } = data.setValues ? this.filesToUpload(data.setValues) : {};
        const data2 = {
            ...data,
            setValues: values,
        };
        let updateQuery;
        try {
            updateQuery = parseUpdateQuery(data2, this.jc);
        }
        catch (e) {
            throw new ProcessError('Error parsing UPDATE data', 400, data, e);
        }
        await this.onUpdateParse(updateQuery);
        const fileFields = this.fileFields?.filter(f => updateQuery.set[f] !== undefined);
        // const filesToDelete = fileFields ? await this.filesToDelete(updateQuery, fileFields) : undefined
        const res = this.$db.rawUpdate(this, updateQuery, filesToUpload, filesToRef, [], fileFields, then);
        return res;
    }
    async rawView(data, id) {
        if (!this.mapping.uid)
            throw new HTTPException(404, { message: 'Not supported without uidField.' });
        let query;
        try {
            query = parseSelectQuery({
                where: `${this.data.name}.${this.mapping.uid}=${JSON.stringify(id)}` + (data.where ? `& (${data.where})` : ''),
                limit: 1,
                select: data.select,
            }, this.jc);
        }
        catch (e) {
            throw new ProcessError('Error parsing SELECT(view) data', 400, data, e);
        }
        await this.onViewParse(query, id);
        const res = this.$db.rawSelect(this, query, (r) => r?.length ? r[0] : null);
        return res;
    }
    async rawEdit(data, id) {
        if (!this.mapping.uid)
            throw new HTTPException(404, { message: 'Not supported without uidField.' });
        const res = data.or !== 'INSERT' ? await this.rawUpdate({
            where: `${this.data.name}.${this.mapping.uid}=${JSON.stringify(id)}`,
            setValues: data.setValues,
            returning: data.returning,
            or: data.or,
        }, r => r?.length ? r[0] : null) : await this.rawInsert({
            values: {
                ...data.setValues,
                [this.mapping.uid]: id,
            },
            returning: data.returning,
            or: 'REPLACE',
        }, r => r?.length ? r[0] : null);
        return res;
    }
    async rawSelect(data, countTotal) {
        let query;
        try {
            query = parseSelectQuery(data, this.jc);
        }
        catch (e) {
            throw new ProcessError('Error parsing SELECT data', 400, data, e);
            // throw new ProcessErrore: 'Invalid query')
        }
        await this.onSelectParse(query);
        if (countTotal === undefined || typeof countTotal === 'function')
            return this.$db.rawSelect(this, query, countTotal);
        return this.$db.rawSelect(this, query, countTotal);
    }
    async rawSelectCount(data) {
        let query;
        try {
            query = parseSelectQuery(data, this.jc);
        }
        catch (e) {
            throw new ProcessError('Error parsing SELECT data', 400, data, e);
        }
        await this.onSelectParse(query);
        if (query.selects?.length ?? 0 > 1)
            throw new ProcessError('Invalid select count query, only 1 field allowed');
        const countField = query.selects?.length === 1 ? query.selects[0] :
            this.mapping.uid ? `${this.data.name}.${columnify(this.mapping.uid)}` : '*';
        const res = this.$db.rawSelect(this, {
            ...query,
            selects: [`count(${countField}) as total`],
            limit: undefined,
            offset: undefined,
            orderBy: undefined,
            groupBy: undefined,
            distinct: false,
        }, r => r?.length ? r[0].total : -1);
        return res;
    }
    async rawSelectRaise(data) {
        const { message, code } = z.object({
            message: z.string(),
            code: z.enum(['ROLLBACK', 'ABORT', 'FAIL']).optional().default('FAIL'),
        }).parse(data);
        const data2 = {
            ...data,
            select: `RAISE(${code}, ${JSON.stringify(message)})`,
        };
        let query;
        try {
            query = parseSelectQuery(data2, this.jc);
        }
        catch (e) {
            throw new ProcessError('Error parsing SELECT data', 400, data, e);
        }
        await this.onSelectParse(query);
        const res = this.$db.rawSelect(this, query, r => {
            if (r.length)
                throw new ProcessError(message, 409, { code: code });
            return null;
        });
        return res;
    }
    // hooks
    async onInsertParse(query) {
        let v = Array.isArray(query.values) ? query.values : [query.values];
        if (v.length && v[0]) {
            const setKeys = Object.keys(v[0]); // its checked in the buildInsert if all the rows share the same key. todo should it be done here?
            for (const key of setKeys) {
                const field = this.fields[key];
                // if (!field) throw new ProcessError: 'Invalid field ' + key}) // this check is moved below.
                if (!field)
                    continue;
                // if (field.usage && noInsertSetUsages.includes(field.usage)) throw new ProcessError('Cannot insert ' + key + ' field')
                if (field.noInsert)
                    throw new ProcessError('Cannot insert ' + key + ' field', this.jc.globals.auth?.uid ? 403 : 401);
            }
            if (this.mapping.uid && this.data.autoSetUid) {
                for (const val of v) {
                    if (!val[this.mapping.uid])
                        val[this.mapping.uid] = { l: generateUid() };
                }
            }
        }
        // todo notNull check
        if (query.returning)
            query.returning = this.checkColsWildcard(Array.isArray(query.returning) ? query.returning : [query.returning]);
        for (const ext of this.extensions) {
            ext.onInsertParse && await ext.onInsertParse(query);
        }
        // todo check for notNull fields so we dont hit the db
        v = Array.isArray(query.values) ? query.values : [query.values];
        if (v.length && v[0]) {
            const setKeys = Object.keys(v[0]);
            for (const key of setKeys) {
                const field = this.fields[key];
                if (!field)
                    throw new ProcessError('Invalid field ' + key);
            }
        }
    }
    async onDeleteParse(query) {
        if (query.returning)
            query.returning = this.checkColsWildcard(Array.isArray(query.returning) ? query.returning : [query.returning]);
        for (const ext of this.extensions) {
            ext.onDeleteParse && await ext.onDeleteParse(query, this.jc.globals.auth?.admin ?? false);
        }
    }
    async onUpdateParse(query) {
        let setKeys = Object.keys(query.set);
        for (const key of setKeys) {
            const field = this.fields[key];
            if (!field)
                continue;
            // if(field.usage && noUpdateSetUsages.includes(field.usage)) throw new ProcessError('Cannot update ' + key + ' field')
            if (field.noUpdate)
                throw new ProcessError('Cannot update ' + key + ' field', this.jc.globals.auth?.uid ? 403 : 401);
        }
        // todo notNull check
        // todo autodate?
        if (this.mapping.updated)
            query.set[this.mapping.updated] = { q: 'CURRENT_TIMESTAMP' };
        if (query.returning)
            query.returning = this.checkColsWildcard(Array.isArray(query.returning) ? query.returning : [query.returning]);
        for (const ext of this.extensions) {
            ext.onUpdateParse && await ext.onUpdateParse(query);
        }
        setKeys = Object.keys(query.set); // again, could have been changed
        for (const key of setKeys) {
            const field = this.fields[key];
            if (!field)
                throw new ProcessError('Invalid field ' + key);
        }
    }
    async onSelectParse(query) {
        // console.log('query', query)
        // todo make data.allowSubqueries and check
        const allowSubqueries = true;
        query.selects = this.checkColsWildcard(Array.isArray(query.selects) ? query.selects : [query.selects || '*'], allowSubqueries);
        this.handleSelectDeps(query);
        // extensions
        for (const ext of this.extensions) {
            ext.onSelectParse && await ext.onSelectParse(query);
        }
        await this.handleSelectSubQueries(query, allowSubqueries);
    }
    async onViewParse(query, id) {
        // todo make data.allowSubqueries and check
        const allowSubqueries = true;
        query.selects = this.checkColsWildcard(Array.isArray(query.selects) ? query.selects : [query.selects || '*'], allowSubqueries);
        this.handleSelectDeps(query);
        for (const ext of this.extensions) {
            ext.onViewParse && await ext.onViewParse(query, id);
        }
        await this.handleSelectSubQueries(query, allowSubqueries);
    }
    handleSelectDeps(query) {
        // sql query dependencies. todo move to a function or extension
        const allQueries = [query.where, query.having].flat();
        for (const sqlQuery of allQueries) {
            if (!sqlQuery)
                continue;
            const deps = sqlQuery.dependencies;
            if (!deps?.length)
                continue;
            for (const dep of deps) {
                if (dep.type === 'fts') {
                    const { table, column } = dep;
                    if (table !== this.jc.tableName)
                        throw new ProcessError('Invalid FTS dependency table ' + table);
                    if (!this.data.fullTextSearch || this.data.fullTextSearch.enabled === false)
                        throw new ProcessError('FTS not available or disabled on table ' + table);
                    if (column && !this.data.fullTextSearch.fields.map(f => columnify(f)).includes(column))
                        throw new ProcessError('Invalid FTS column ' + column);
                    const content_rowid = this.data.fullTextSearch.content_rowid ?? 'rowid';
                    // add join
                    // SELECT main_table.uuid, main_table.content
                    // FROM main_table
                    // JOIN fts_table ON main_table.rowid = fts_table.rowid
                    // WHERE fts_table MATCH 'search_term';
                    const ftsTable = columnify(fts5TableName(this.name));
                    const join = {
                        table: ftsTable,
                        on: { q: `${columnify(this.name)}.${content_rowid} = ${ftsTable}.rowid` },
                    };
                    appendJoin(query, join);
                    appendOrderBy(query, `${ftsTable}.rank`);
                    continue;
                }
                throw new ProcessError('Invalid SQLQuery dependency type ' + dep.type);
            }
        }
    }
    async handleSelectSubQueries(query, allowSubqueries) {
        const subQueries = getSubSelectQueries(query);
        for (const subQuery of subQueries) {
            if (!allowSubqueries)
                throw new ProcessError('Subqueries not allowed');
            if (typeof subQuery.from !== 'string')
                throw new ProcessError('Subquery from must be a string');
            const table = this.$db.table(uncolumnify(subQuery.from));
            if (!table)
                throw new ProcessError('Subquery table not found');
            await table.onSelectParse(subQuery); // todo use Promise.all? shouldnt matter actually
        }
    }
    checkNewRecordValues(setValues) {
        if (!setValues)
            return;
        for (const [key, val] of Object.entries(setValues)) { // todo
            if (Array.isArray(val))
                throw new ProcessError('Array values not supported at the moment');
            const field = this.fields[key];
            // if(!field) throw new ProcessError('Invalid field ' + key)
            if (!field)
                continue; // field check is done in onUpdateParse
            // for these types, any string "null" will be coersed to null(this is required for FormData as it only sends strings)
            // todo relation, password, editor, text, select
            const directNullableTypes = ['file', 'number', 'bool', 'email', 'url', 'date', 'json', 'integer', 'blob'];
            if (directNullableTypes.includes(field.type) && val === 'null') {
                // todo notnull check required?
                setValues[key] = null; // any existing file will be deleted in filesToDelete
                continue;
            }
            if (field.type === 'json') {
                if (typeof val === 'string' && val !== '') {
                    try {
                        // this should return null if val is "null"
                        const l = JSON.parse(val);
                        if (l === null)
                            setValues[key] = l;
                    }
                    catch (e) {
                        throw new ProcessError('Invalid JSON value for ' + key);
                    }
                }
            }
        }
    }
    // D1 Helpers
    parseD1Error(e) {
        const msg = e?.errorMessage;
        // todo check for other error types.
        // NOT NULL constraint failed: files.config: SQLITE_CONSTRAINT
        // FOREIGN KEY constraint failed: SQLITE_CONSTRAINT
        // UNIQUE constraint failed: users.email: SQLITE_CONSTRAINT
        const errorMap = {
            'UNIQUE constraint failed': 'A record with this value already exists.',
            'NOT NULL constraint failed': 'This field cannot be NULL.',
            // 'FOREIGN KEY constraint failed': 'Unable to find related record (Foreign Key).',
            'FOREIGN KEY constraint failed': 'Foreign Key constraint failed.',
        };
        const parts = msg.split(':').map(p => p.trim());
        const column = parts.find(p => {
            const c = p.split('.').pop();
            return c && this.fields[c];
        })?.split('.').pop();
        const mess = parts[0];
        if (column) {
            const code = parts.pop();
            e.data = {
                [column]: {
                    code,
                    message: errorMap[mess] ?? mess,
                }
            };
        }
        else if (errorMap[mess]) {
            e.message += ' - ' + errorMap[mess];
        }
        return e;
    }
    // R2 Helpers
    // private async filesToDelete(desQuery: UpdateQuery|DeleteQuery, fileFields?: string[]) {
    //     if(!fileFields) fileFields = this.fileFields || []
    //     if (!fileFields.length) return undefined
    //     if(desQuery.table !== this.jc.tableName) throw new ProcessError('Invalid table for filesToDelete')
    //     const selects = fileFields.map(f => ident(f, this.jc))
    //     // const res = await this.run<any>(selectQuery, 'Failed to run select query')
    //     const res = await this.$db.rawSelect(this, {
    //         from: desQuery.table,
    //         where: desQuery.where, // this will have the rule as well.
    //         selects,
    //         // limit: desQuery.limit,
    //         params: desQuery.params,
    //     })?.run() || []
    //     return res.flatMap(r => Object.values(r)).filter(v => !!v) as string[]
    // }
    // returning is not used, todo remove?
    filesToUpload(values, filesToUpload, filesToRef, returning) {
        filesToUpload = filesToUpload ?? {};
        filesToRef = filesToRef ?? [];
        returning = returning ?? [];
        if (!values || !this.fileFields)
            return { filesToUpload, filesToRef, values: sqlValSchemaRecord.parse(values), returning };
        const transformed = {};
        for (let [key, val] of Object.entries(values)) {
            const isFileField = this.fileFields.includes(key);
            // ignore file field. This is done because file fields in html form set value to empty string when no file is set. we dont want to send the file every time we are updating some record, so empty string is set in autoform
            //  so empty string when passed in update for file fields means NO CHANGE
            //  to delete a file, set it to null
            if (isFileField && Array.isArray(val)) { // remove extra empty string values
                val = val.filter(v => v !== '');
                if (val.length === 1)
                    val = val[0];
            }
            if (isFileField && val === '')
                continue;
            if (!(val instanceof File)) {
                transformed[key] = val;
            }
            if (!isFileField || val === null || val === undefined)
                continue;
            // when deleting file, null should be passed.
            if (val === '@null') { // todo global constant
                transformed[key] = null;
                continue;
            }
            if (typeof val === 'string') {
                // todo
                const allowArbitraryStringsInFile = false;
                if (allowArbitraryStringsInFile) {
                    transformed[key] = val;
                    continue;
                }
                if (this.allowMultipleFileRef) {
                    // todo check if file exists?
                    const name = normalizeFileName(val, '');
                    // todo this will fail if normalizeFileName implementation changes, we should just check for the file in the bucket or could that open up for security issues?
                    if (name !== val)
                        throw new ProcessError('Invalid file name ' + val);
                    // values[key] = name // not needed
                    filesToRef.push(name);
                    continue;
                }
                // todo check if the string value is same as the current value in db.
                // todo try to download file?
                throw new ProcessError('File URLs not supported. Set "allowMultipleFileRef" to refer to files from other records.');
            }
            if (val instanceof File) {
                const name = normalizeFileName(val.name);
                filesToUpload[name] = val;
                transformed[key] = name;
                if (!returning.includes(key))
                    returning.push(key); // we will return the file name to be sure it was updated in db
                continue;
            }
            throw new ProcessError('File or string expected in file field ' + key);
        }
        return { filesToUpload, filesToRef, values: sqlValSchemaRecord.parse(transformed), returning };
    }
    fileRoute(key, recordId) {
        // todo hardcoded api path
        return `/api/v1/files/${this.name}/${recordId}/${key}`;
    }
    // todo record id like in pocketbase? not using right now because id can change which would break.
    fileKey(key, _recordId) {
        let basePath = this.data.r2Base /*??this.data.id*/ ?? this.data.name;
        if (this.data.idInR2) {
            if (!_recordId)
                throw new HTTPException(500, { message: 'Record id required when idInR2 is true' });
            basePath += '/' + _recordId;
        }
        return `${basePath}/${key}`;
    }
    async getFile(key, _recordId) {
        const fileKey = this.fileKey(key, _recordId);
        const object = await this.$db.getFileObject(fileKey).catch(e => {
            console.error('Failed to get file', key, e);
            throw new HTTPException(500, { message: 'Failed to get file' });
        });
        if (!object)
            throw new HTTPException(404, { message: 'File not found' });
        return object;
    }
    checkColsWildcard(cols, allowSubqueries = false) {
        const res = [];
        const isAdmin = !!this.jc.globals.auth?.admin;
        // const hidden = isAdmin ? [] : this.data.fields.filter(f=>!f.noSelect).map(f=>f.name)
        // console.log('select ', cols, hidden)
        const aliases = new Set();
        const addAlias = (alias) => {
            if (aliases.has(alias))
                throw new ProcessError('Duplicate field access ' + alias);
            aliases.add(alias);
        };
        for (const col of cols) {
            if (col === '*') {
                if (!this.allowWildcard /* || hidden.length*/) {
                    const fields = Object.entries(this.fields).filter(f => isAdmin || !f[1].noSelect);
                    const fieldsSql = fields.map(s => ident(s[0], this.jc));
                    res.push(...fieldsSql);
                    fields.forEach(f => addAlias(f[0]));
                    continue;
                }
                res.push('*');
                addAlias('*');
                continue;
            }
            if (typeof col === 'string' || typeof col.q === 'string') {
                const strCol = typeof col === 'string' ? col : col.q;
                if (strCol.includes('*') && strCol !== 'COUNT(*)')
                    throw new ProcessError('Invalid column ' + strCol);
                const alias = typeof col !== 'string' ? col.as ?? strCol : strCol;
                addAlias(alias);
                // const column = col.split(' AS')[0]
                // if(hidden.includes(column)) throw new ProcessError('Cannot query ' + col)
            }
            else {
                if (!allowSubqueries)
                    throw new ProcessError('Subqueries not allowed');
                if (!col.as)
                    throw new Error('Subquery must have an alias');
                addAlias(col.as);
                // subquery
                // it will be checked in its own table
            }
            res.push(col);
        }
        // console.log(res)
        return res;
    }
    // Routing (separate from $Database so we only init the routes for the table when required)
    router;
    _routesInit = false;
    routes = [];
    _initRoutes() {
        if (!this.router)
            this.router = new LinearRouter();
        if (this._routesInit)
            return this.router;
        this.routes.push(...this.extensions.flatMap(e => e.routes));
        for (const route of this.routes) {
            this.router.add(route.method.toUpperCase(), route.path, this.$db.rawRouteHandler(route));
        }
        this._routesInit = true;
        return this.router;
    }
    async route(path) {
        const router = this._initRoutes();
        const match = router.match(this.$db.requestMethod, path);
        const [handler, params] = match[0]?.[0] ?? [undefined];
        if (!handler)
            return undefined;
        return await handler(params, path);
    }
    getRoutes() {
        this._initRoutes();
        return this.routes;
    }
    _zodSchema;
    get zodSchema() {
        if (this._zodSchema)
            return this._zodSchema;
        const schema = {};
        for (const field of this.data.fields) {
            schema[field.name] = tableFieldTypeToZod[field.type]
                ?? tableFieldTypeToZod[field.sqlType];
        }
        this._zodSchema = z.object(schema);
        return this._zodSchema;
    }
}
