// teenybase/dist/worker/internalKV.js
export class InternalKV {
    db;
    tableName;
    constructor(db, tableName = '_ddb_internal_kv') {
        this.db = db;
        this.tableName = tableName;
    }
    async setup(version) {
        if (version >= 0) {
            const sql = `
            CREATE TABLE IF NOT EXISTS "${this.tableName}" (
                key TEXT PRIMARY KEY,
                value TEXT NOT NULL,
                expiry INTEGER NULL
            )`.split('\n').map(l => l.trim()).join(' ');
            await this.db.rawSQL({ q: sql, v: [] }).run();
        }
    }
    async get(key, field) {
        const rnd = '_' + Math.random().toString(36).substring(7);
        const q = `SELECT (${field || 'value'}) as ${rnd} FROM ${this.tableName} WHERE key = ? AND (expiry IS NULL OR expiry > unixepoch(CURRENT_TIMESTAMP))`;
        const res = await this.db.rawSQL({ q, v: [key] }).run();
        return res?.[0] ? res[0][rnd] : null;
    }
    async pop(key, field) {
        const rnd = '_' + Math.random().toString(36).substring(7);
        const q = `DELETE FROM ${this.tableName} WHERE key = ? AND (expiry IS NULL OR expiry > unixepoch(CURRENT_TIMESTAMP)) RETURNING (${field || 'value'}) as ${rnd}`;
        const res = await this.db.rawSQL({ q, v: [key] }).run();
        return res?.[0] ? res[0][rnd] : null;
    }
    async get2(key, ...fields) {
        const q = `SELECT value, expiry, ${fields.join(', ')} FROM ${this.tableName} WHERE key = ? AND (expiry IS NULL OR expiry > unixepoch(CURRENT_TIMESTAMP))`;
        const res = await this.db.rawSQL({ q, v: [key] }).run();
        return res?.[0] || null;
    }
    async set(key, value, expiryOffsetSeconds) {
        return this.db.rawSQL(this.setQuery(key, value, expiryOffsetSeconds)).run();
    }
    setQuery(key, value, expiryOffsetSeconds) {
        const s = typeof value !== 'string';
        if (typeof expiryOffsetSeconds !== 'number' || isNaN(expiryOffsetSeconds)) {
            expiryOffsetSeconds = 1;
        }
        const q = `INSERT OR REPLACE INTO ${this.tableName} (key, value, expiry) VALUES (?, ${s ? value.sql || "''" : "?"}, ${expiryOffsetSeconds ? `unixepoch(CURRENT_TIMESTAMP) + ${expiryOffsetSeconds}` : 'NULL'})`;
        const b = [key];
        if (!s)
            b.push(value);
        return { q, v: b };
    }
    async setMultiple(data, expiryOffsetSeconds) {
        const qs = [];
        const b = [];
        const expiry = expiryOffsetSeconds ? `unixepoch(CURRENT_TIMESTAMP) + ${expiryOffsetSeconds}` : 'NULL';
        for (const [key, value] of Object.entries(data)) {
            const s = typeof value !== 'string';
            qs.push(`(?, ${s ? value.sql || "''" : "?"}, ${expiry})`);
            b.push(key);
            if (!s)
                b.push(value);
        }
        const q = `INSERT OR REPLACE INTO ${this.tableName} (key, value, expiry) VALUES ${qs.join(', ')}`;
        return this.db.rawSQL({ q, v: b }).run();
    }
    async remove(key) {
        const q = `DELETE FROM ${this.tableName} WHERE key = ?`;
        await this.db.rawSQL({ q, v: [key] }).run();
    }
    async setSql(key, value, expiryOffsetSeconds) {
        return await this.set(key, { sql: value }, expiryOffsetSeconds);
    }
}
