// teenybase/dist/worker/usages.js
export var TableFieldUsageRecord;
(function (TableFieldUsageRecord) {
    TableFieldUsageRecord["record_uid"] = "record_uid";
    TableFieldUsageRecord["record_created"] = "record_created";
    TableFieldUsageRecord["record_updated"] = "record_updated";
})(TableFieldUsageRecord || (TableFieldUsageRecord = {}));
export var TableFieldUsageAuth;
(function (TableFieldUsageAuth) {
    TableFieldUsageAuth["auth_username"] = "auth_username";
    TableFieldUsageAuth["auth_email"] = "auth_email";
    TableFieldUsageAuth["auth_email_verified"] = "auth_email_verified";
    TableFieldUsageAuth["auth_password"] = "auth_password";
    TableFieldUsageAuth["auth_password_salt"] = "auth_password_salt";
    TableFieldUsageAuth["auth_name"] = "auth_name";
    TableFieldUsageAuth["auth_avatar"] = "auth_avatar";
    TableFieldUsageAuth["auth_audience"] = "auth_audience";
    TableFieldUsageAuth["auth_metadata"] = "auth_metadata";
    // auth_reset_sent_at = 'auth_reset_sent_at',
    // auth_verification_sent_at = 'auth_verification_sent_at',
})(TableFieldUsageAuth || (TableFieldUsageAuth = {}));
