// teenybase/dist/types/zod/databaseSettingsSchema.js
import { z } from 'zod';
import { tableDataSchema } from './tableDataSchema';
import { zSQLProcedure } from './sqlSchemas';
import { baseTemplatePropsSchema } from './emailTemplatePropsSchema';
import { mailgunBindingsSchema } from './mailgunBindingsSchema';
import { resendBindingsSchema } from './resendBindingsSchema';
export const databaseSettingsSchema = z.object({
    tables: z.array(tableDataSchema),
    jwtSecret: z.string(),
    jwtIssuer: z.string().optional(),
    jwtAlgorithm: z.string().optional(),
    version: z.number().optional(),
    appName: z.string().optional(),
    appUrl: z.string(),
    procedures: z.array(zSQLProcedure).optional(),
    email: z.object({
        from: z.string(),
        variables: baseTemplatePropsSchema,
        tags: z.array(z.string()).optional(),
        mailgun: mailgunBindingsSchema.optional(),
        resend: resendBindingsSchema.optional(),
    }).optional(),
    _kvTableName: z.string().startsWith('_', { message: 'Internal table name must start with _ (underscore)' }).optional(),
    disableTablesEdit: z.coerce.boolean().optional(),
});
