// teenybase/dist/types/zod/sqlSchemas.js
import { z } from 'zod';
export const sqlExprSchema = z.string().max(1000).regex(/^[a-zA-Z0-9_=\s():.*~%'"><!+\-&|@,\/\\${}]*$/, 'Invalid expression');
export const sqlExprSchemaRecord = z.record(sqlExprSchema);
export const sqlValSchema = z.string().or(z.number()).or(z.boolean()).or(z.null());
export const sqlValSchemaFile = z.boolean().or(z.string().or(z.number()).or(z.null()).or(z.instanceof(File)));
export const sqlValSchemaRecord = z.record(sqlValSchema);
export const sqlValSchema2 = sqlValSchema.or(sqlValSchemaRecord);
export const sqlValSchemaFile2 = z.record(sqlValSchemaFile);
export const sqlValSchemaFile3 = sqlValSchemaFile2.or(z.array(sqlValSchemaFile2));
export const sqlColListSchema = z.string().regex(/^[a-zA-Z0-9_,\s]*$/).or(z.literal('*'));
export const sqlColListSchemaOrder = z.string().regex(/^[a-zA-Z0-9_,\s+-]*$/);
// for search query params
export const sqlColListSchema2 = z.array(sqlColListSchema).or(sqlColListSchema);
export const sqlColListSchemaOrder2 = z.array(sqlColListSchemaOrder).or(sqlColListSchemaOrder);
export const tableColumnNameSchema = z.string()
    .min(1).max(255)
    .regex(/^[a-zA-Z_][a-zA-Z0-9_]*$/, 'Table/Column name must start with a letter or underscore and can only contain letters, numbers and underscores');
//.or(z.literal('*')) // todo *
export const zSQLQuery = z.object({
    q: z.string(),
    p: z.record(sqlValSchema2).optional()
});
export const zSQLLiteral = z.object({
    l: sqlValSchema2,
    key: z.string().optional()
});
export const zSQLQueryOrLiteral = zSQLQuery.or(zSQLLiteral);
export const zOnConflict = z.enum(['ABORT', 'FAIL', 'IGNORE', 'REPLACE', 'ROLLBACK']);
export const selectSchema = z.object({
    where: sqlExprSchema.optional(),
    limit: z.coerce.number().optional(),
    offset: z.coerce.number().optional(),
    order: sqlColListSchemaOrder2.optional(),
    group: sqlColListSchema2.optional(),
    select: sqlExprSchema.or(z.array(sqlExprSchema)).optional(), // todo rename to fields in api v2
    distinct: z.coerce.boolean().optional(),
    // join: joinObjSchema2.array().optional(),
    // having: z.string().regex(/^[a-zA-Z0-9_=\s\(\)\.\*]*$/).optional(),
    // selectOption: z.string().optional(),
});
export const updateSchema = z.object({
    or: zOnConflict.optional(),
    set: z.record(sqlExprSchema).optional(),
    setValues: z.record(sqlValSchema).optional(),
    where: sqlExprSchema,
    returning: z.array(sqlExprSchema).or(sqlExprSchema).optional()
});
export const insertSchema = z.object({
    or: zOnConflict.optional(),
    values: sqlValSchemaRecord.or(z.array(sqlValSchemaRecord)).optional(),
    expr: sqlExprSchemaRecord.or(z.array(sqlExprSchemaRecord)).optional(),
    returning: z.array(sqlExprSchema).or(sqlExprSchema).optional(),
});
export const deleteSchema = z.object({
    where: sqlExprSchema,
    returning: z.array(sqlExprSchema).or(sqlExprSchema).optional(),
});
export const tableDeleteSchema = deleteSchema;
export const tableInsertSchema = insertSchema.omit({ values: true }).extend({ values: sqlValSchemaFile3 });
export const tableUpdateSchema = updateSchema.omit({ setValues: true }).extend({ setValues: sqlValSchemaFile2.optional() });
export const tableSelectSchema = selectSchema;
export const tableViewSchema = selectSchema.pick({ select: true, where: true });
export const tableEditSchema = z.object({
    setValues: sqlValSchemaFile2,
    or: updateSchema.shape.or.or(z.literal('INSERT')),
    returning: updateSchema.shape.returning
});
export const selectStatementSchema = tableSelectSchema.extend({
    type: z.literal('SELECT'),
    table: tableColumnNameSchema,
});
export const insertStatementSchema = tableInsertSchema.extend({
    type: z.literal('INSERT'),
    table: tableColumnNameSchema,
});
export const updateStatementSchema = tableUpdateSchema.extend({
    type: z.literal('UPDATE'),
    table: tableColumnNameSchema,
});
export const deleteStatementSchema = tableDeleteSchema.extend({
    type: z.literal('DELETE'),
    table: tableColumnNameSchema,
});
export const zSQLIndex = z.object({
    name: tableColumnNameSchema.optional(),
    unique: z.coerce.boolean().optional(),
    // fields: tableColumnNameSchema.or(z.array(tableColumnNameSchema)),
    fields: z.string().or(z.array(z.string())), // fields can include collate also
    where: zSQLQuery.optional(),
});
export const zSQLTrigger = z.object({
    name: tableColumnNameSchema,
    event: z.enum(['INSERT', 'DELETE', 'UPDATE']),
    seq: z.enum(['BEFORE', 'AFTER', 'INSTEAD OF']).optional(),
    updateOf: tableColumnNameSchema.or(z.array(tableColumnNameSchema)).optional(),
    forEach: z.enum(['ROW']).optional(),
    body: zSQLQuery.or(z.array(zSQLQuery)),
    when: zSQLQuery.optional(),
});
export const zTypedSQLQuery = z.record(z.string(), z.any()); // todo SelectQuery|UpdateQuery...
export const zTypedSQLStatement = selectStatementSchema.or(insertStatementSchema).or(updateStatementSchema).or(deleteStatementSchema);
export const zSQLProcedure = z.object({
    name: tableColumnNameSchema,
    rule: z.boolean() /*.or(z.string())*/.default(true),
    params: z.array(tableColumnNameSchema),
    query: z.union([zTypedSQLQuery, z.array(zTypedSQLQuery)]).optional(),
    statement: z.union([zTypedSQLStatement, z.array(zTypedSQLStatement)]).optional(),
});
// not final, cant pass in query param
// export const joinObjSchema = z.object({
//     table: z.string(),
//     on: sqlExprSchema,
//     type: z.string().optional(),
// })
// export const joinObjSchema2 = z.array(joinObjSchema).or(joinObjSchema)
