// teenybase/dist/types/zod/tableFieldDataSchema.js
import { z } from 'zod';
import { TableFieldDataType, TableFieldSqlDataType0, TableFieldSqlDataType1 } from '../dataTypes';
import { zSQLQueryOrLiteral } from './sqlSchemas';
export const zForeignKeyAction = z.enum(['SET NULL', 'SET DEFAULT', 'CASCADE', 'RESTRICT', 'NO ACTION']);
export const tableFieldDataSchema = z.object({
    name: z.string(),
    sqlType: z.nativeEnum(TableFieldSqlDataType0).or(z.nativeEnum(TableFieldSqlDataType1)),
    type: z.nativeEnum(TableFieldDataType),
    usage: z.string().optional(),
    primary: z.coerce.boolean().optional(),
    autoIncrement: z.coerce.boolean().optional(),
    unique: z.coerce.boolean().optional(),
    notNull: z.coerce.boolean().optional(),
    collate: z.string().optional(), // todo enum
    default: zSQLQueryOrLiteral.or(z.string()).optional(),
    check: zSQLQueryOrLiteral.or(z.string()).optional(),
    foreignKey: z.object({
        table: z.string(),
        column: z.string(),
        onUpdate: zForeignKeyAction.optional(),
        onDelete: zForeignKeyAction.optional(),
    }).optional(),
    // updateTriggers: z.array(zSQLTrigger.omit({updateOf: true, event: true})).optional(),
    noUpdate: z.coerce.boolean().optional(),
    noInsert: z.coerce.boolean().optional(),
    noSelect: z.coerce.boolean().optional(),
    lastName: z.string().optional(),
});
