// teenybase/dist/types/zod/tableDataSchema.js
import { z } from 'zod';
import { tableFieldDataSchema } from './tableFieldDataSchema';
import { tableColumnNameSchema, zSQLIndex, zSQLTrigger } from './sqlSchemas';
export const tableDataSchema = z.object({
    // id: z.string(),
    name: tableColumnNameSchema,
    r2Base: z.string().optional(),
    idInR2: z.coerce.boolean().optional(),
    autoDeleteR2Files: z.coerce.boolean().optional(),
    allowMultipleFileRef: z.coerce.boolean().optional(),
    allowWildcard: z.coerce.boolean().optional(),
    fields: z.array(tableFieldDataSchema),
    indexes: z.array(zSQLIndex).optional(),
    triggers: z.array(zSQLTrigger).optional(),
    autoSetUid: z.coerce.boolean().optional(),
    extensions: z.array(z.record(z.any()).and(z.object({ name: z.string() }))),
    lastName: tableColumnNameSchema.optional(),
    fullTextSearch: z.object({
        enabled: z.coerce.boolean().optional(),
        fields: z.array(tableColumnNameSchema),
        tokenize: z.string().optional(),
        prefix: z.string().optional(),
        // content: z.string().optional(),
        migrateTableQuery: z.coerce.boolean().default(true),
        content_rowid: z.string().optional(),
        columnsize: z.literal(0).or(z.literal(1)).optional(),
        detail: z.enum(["full", "column", "none"]).optional(),
    }).optional(),
});
