// teenybase/dist/types/zod/tableExtensionsSchema.js
import { z } from 'zod';
import { sqlExprSchema } from './sqlSchemas';
export const tableRulesDataSchema = z.object({
    name: z.literal('rules'),
    listRule: sqlExprSchema.nullable().default(null),
    viewRule: sqlExprSchema.nullable().default(null),
    createRule: sqlExprSchema.nullable().default(null),
    updateRule: sqlExprSchema.nullable().default(null),
    deleteRule: sqlExprSchema.nullable().default(null),
});
export const emailTemplateSchema = z.object({
    subject: z.string().optional(),
    variables: z.record(z.any()).optional(),
    tags: z.string().optional(),
    layoutHtml: z.string().optional(),
});
export const tableAuthDataSchema = z.object({
    name: z.literal('auth'),
    jwtSecret: z.string(),
    jwtTokenDuration: z.number(),
    maxTokenRefresh: z.number(),
    passwordType: z.literal('sha256').default('sha256'),
    passwordConfirmSuffix: z.string().optional(),
    passwordCurrentSuffix: z.string().optional(),
    passwordResetTokenDuration: z.number().optional(),
    emailVerifyTokenDuration: z.number().optional(),
    passwordResetEmailDuration: z.number().optional(),
    emailVerifyEmailDuration: z.number().optional(),
    autoSendVerificationEmail: z.coerce.boolean().optional(),
    emailTemplates: z.object({
        verification: emailTemplateSchema.optional(),
        passwordReset: emailTemplateSchema.optional(),
    }).optional(),
    // onlyVerified: z.coerce.boolean().optional(),
    // minPasswordLength: z.number().optional(),
    // onlyEmailDomains: z.array(z.string()).nullable().optional(),
    // exceptEmailDomains: z.array(z.string()).nullable().optional(),
    // allowEmailAuth: z.coerce.boolean().optional(),
    // allowUsernameAuth: z.coerce.boolean().optional(),
    // allowOAuth2Auth: z.coerce.boolean().optional(),
    // manageRule: z.string().nullable().optional(),
});
export const tableExtensionSchemas = {
    rules: tableRulesDataSchema,
    auth: tableAuthDataSchema,
};
