// teenybase/dist/worker/email/templates/base-layout-1.js
export const baseLayout1 = `
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
        "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml"
>
<head>
    <title>{{company_name}}</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
    <meta http-equiv="X-UA-Compatible" content="IE=edge"/>
    <meta name="viewport" content="width=device-width, initial-scale=1.0 "/>
    <meta name="format-detection" content="telephone=no"/>
    <style type="text/css">
        body {
            margin: 0;
            padding: 0;
            -webkit-text-size-adjust: 100% !important;
            -ms-text-size-adjust: 100% !important;
            -webkit-font-smoothing: antialiased !important;
            background-color: #F0F2F8;
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Ubuntu, sans-serif;
        }

        img {
            border: 0 !important;
            outline: none !important;
        }

        p {
            Margin: 0px !important;
            Padding: 0px !important;
        }

        table {
            border-collapse: collapse;
            mso-table-lspace: 0px;
            mso-table-rspace: 0px;
        }

        td, a, span {
            border-collapse: collapse;
            mso-line-height-rule: exactly;
        }

        .ExternalClass * {
            line-height: 100%;
        }

        .em_blue a {
            text-decoration: none;
            color: #264780;
        }

        .em_grey a {
            text-decoration: none;
            color: #434343;
        }

        .em_white a {
            text-decoration: none;
            color: #ffffff;
        }

        .em_aside5 {
            padding: 0 20px !important;
        }

        @media only screen and (min-width: 481px) and (max-width: 649px) {
            .em_main_table {
                width: 100% !important;
            }

            .em_wrapper {
                width: 100% !important;
            }

            .em_hide {
                display: none !important;
            }

            .em_aside10 {
                padding: 0px 10px !important;
            }

            .em_h20 {
                height: 20px !important;
                font-size: 1px !important;
                line-height: 1px !important;
            }

            .em_h10 {
                height: 10px !important;
                font-size: 1px !important;
                line-height: 1px !important;
            }

            .em_aside5 {
                padding: 0px 10px !important;
            }
        }

        @media only screen and (min-width: 375px) and (max-width: 480px) {
            .em_main_table {
                width: 100% !important;
            }

            .em_wrapper {
                width: 100% !important;
            }

            .em_hide {
                display: none !important;
            }

            .em_aside10 {
                padding: 0px 10px !important;
            }

            .em_aside5 {
                padding: 0px 8px !important;
            }

            .em_h20 {
                height: 20px !important;
                font-size: 1px !important;
                line-height: 1px !important;
            }

            .em_h10 {
                height: 10px !important;
                font-size: 1px !important;
                line-height: 1px !important;
            }

            .em_font_11 {
                font-size: 12px !important;
            }

            .em_font_22 {
                font-size: 22px !important;
                line-height: 25px !important;
            }

            .em_w5 {
                width: 7px !important;
            }

            u + .em_body .em_full_wrap {
                width: 100% !important;
                width: 100vw !important;
            }
        }

        @media only screen and (max-width: 374px) {
            .em_main_table {
                width: 100% !important;
            }

            .em_wrapper {
                width: 100% !important;
            }

            .em_hide {
                display: none !important;
            }

            .em_aside10 {
                padding: 0px 10px !important;
            }

            .em_aside5 {
                padding: 0px 8px !important;
            }

            .em_h20 {
                height: 20px !important;
                font-size: 1px !important;
                line-height: 1px !important;
            }

            .em_h10 {
                height: 10px !important;
                font-size: 1px !important;
                line-height: 1px !important;
            }

            .em_font_11 {
                font-size: 11px !important;
            }

            .em_font_22 {
                font-size: 22px !important;
                line-height: 25px !important;
            }

            .em_w5 {
                width: 5px !important;
            }

            u + .em_body .em_full_wrap {
                width: 100% !important;
                width: 100vw !important;
            }
        }
    </style>
</head>
<body class="em_body" style="margin:0px auto; padding:0px;" bgcolor="#F0F2F8">
<table width="100%" border="0" cellspacing="0" cellpadding="0" class="em_full_wrap" align="center" bgcolor="#F0F2F8">
    <tr>
        <td align="center" valign="top">
            <table align="center" width="480" border="0" cellspacing="0" cellpadding="0" class="em_main_table"
                   style="width:480px; min-width:480px; max-width:480px;">
                <tr>
                    <td align="left" valign="top" style="padding:0 ;" class="em_aside10">
                        <table width="100%" border="0" cellspacing="0" cellpadding="0" align="left">
                            <tr>
                                <td height="58" style="height:58px;" class="em_h20">&nbsp;</td>
                            </tr>
                            <tr>
                                <td align="left" valign="top">
                                    <a href="{{company_url}}" target="_blank" style="text-decoration:none;">
                                        <span style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Ubuntu, sans-serif; font-size:20px; line-height:24px; color:rgb(0,0,0); font-weight:500;">{{company_name}}</span>
                                    </a>
                                </td>
                            </tr>
                            <tr>
                                <td height="32" style="height:32px;" class="em_h20">&nbsp;</td>
                            </tr>
                        </table>
                    </td>
                </tr>
            </table>
        </td>
    </tr>
</table>
<table width="100%" border="0" cellspacing="0" cellpadding="0" class="em_full_wrap" align="center" bgcolor="#F0F2F8">
    <tr>
        <td align="center" valign="top" class="em_aside5">
            <table align="center" width="480" border="0" cellspacing="0" cellpadding="0" class="em_main_table"
                   style="width:480px; min-width:480px; max-width:480px;">
                <tr>
                    <td align="center" valign="top"
                        style="padding:0 20px; background-color:#ffffff; border-radius:12px;">
                        {{EMAIL_CONTENT}}
                    </td>
                </tr>
            </table>
        </td>
    </tr>
</table>
<table width="100%" border="0" cellspacing="0" cellpadding="0" class="em_full_wrap" align="center" bgcolor="#F0F2F8">
    <tr>
        <td align="center" valign="top">
            <table align="center" width="480" border="0" cellspacing="0" cellpadding="0" class="em_main_table"
                   style="width:480px; min-width:480px; max-width:480px;">
                <tr>
                    <td align="center" valign="top" style="padding:0 20px;" class="em_aside10">
                        <table width="100%" border="0" cellspacing="0" cellpadding="0" align="center">
                            <tr>
                                <td height="32" style="height:32px;" class="em_h20">&nbsp;</td>
                            </tr>
                            <tr>
                                <td class="em_grey" align="center" valign="top"
                                    style="font-family: Arial, sans-serif; font-size: 15px; line-height: 18px; color:#434343; font-weight:bold;">
                                    Problems or questions?
                                </td>
                            </tr>
                            <tr>
                                <td height="10" style="height:10px; font-size:1px; line-height:1px;">&nbsp;</td>
                            </tr>
                            <tr>
                                <td align="center" valign="top" style="font-size:0px; line-height:0px;">
                                    <table border="0" cellspacing="0" cellpadding="0" align="center">
                                        <tr>
                                            <td class="em_grey em_font_11" align="left" valign="middle"
                                                style="font-family: Arial, sans-serif; font-size: 13px; line-height: 15px; color:#434343;">
                                                <a href="mailto:{{support_email}}"
                                                   style="text-decoration:none; color:#434343;">{{support_email}}</a>
                                                <a href="mailto:{{support_email}}"
                                                   style="text-decoration:none; color:#434343;">[mailto:{{support_email}}]</a>
                                            </td>
                                        </tr>
                                    </table>
                                </td>
                            </tr>
                            <tr>
                                <td height="9" style="font-size:0px; line-height:0px; height:9px;" class="em_h10">
                                </td>
                            </tr>
                            <tr>
                                <td align="center" valign="top">
                                    <table border="0" cellspacing="0" cellpadding="0" align="center">
                                        <tr>
                                            <td width="12" align="left" valign="middle"
                                                style="font-size:0px; line-height:0px; width:12px;">
                                                <!--                                                <a href="#" target="_blank" style="text-decoration:none;"></a>-->
                                            </td>
                                            <td width="7" style="width:7px; font-size:0px; line-height:0px;"
                                                class="em_w5">&nbsp;
                                            </td>
                                            <td class="em_grey em_font_11" align="left" valign="middle"
                                                style="font-family: Arial, sans-serif; font-size: 13px; line-height: 15px; color:#434343;">
                                                <a href="{{company_url}}" target="_blank"
                                                   style="text-decoration:none; color:#434343;">{{company_name}}</a>
                                                &bull; {{company_address}}
                                            </td>
                                        </tr>
                                    </table>
                                </td>
                            </tr>
                            <tr>
                                <td height="35" style="height:35px;" class="em_h20">&nbsp;</td>
                            </tr>
                        </table>
                    </td>
                </tr>
                <tr>
                    <td height="1" bgcolor="#dadada" style="font-size:0px; line-height:0px; height:1px;">
                    </td>
                </tr>
                <tr>
                    <td align="center" valign="top" style="padding:0 20px;" class="em_aside10">
                        <table width="100%" border="0" cellspacing="0" cellpadding="0" align="center">
                            <tr>
                                <td height="16" style="font-size:0px; line-height:0px; height:16px;">&nbsp;</td>
                            </tr>
                            <tr>
                                <td align="center" valign="top">
                                    <table border="0" cellspacing="0" cellpadding="0" align="left" class="em_wrapper">
                                        <tr>
                                            <td class="em_grey" align="center" valign="middle"
                                                style="font-family: Arial, sans-serif; font-size: 11px; line-height: 16px; color:#434343;">
                                                &copy; {{company_copyright | Copyright}} &nbsp;
                                                <!--|&nbsp;  <a href="#" target="_blank" style="text-decoration:underline; color:#434343;">Unsubscribe</a>-->
                                            </td>
                                        </tr>
                                    </table>
                                </td>
                            </tr>
                            <tr>
                                <td height="16" style="font-size:0px; line-height:0px; height:16px;">&nbsp;</td>
                            </tr>
                        </table>
                    </td>
                </tr>
                <tr>
                    <td class="em_hide" style="line-height:1px;min-width:480px;background-color:#F0F2F8;">
                    </td>
                </tr>
            </table>
        </td>
    </tr>
</table>
</body>
</html>
`;
