// teenybase/dist/worker/extensions/tableRulesExtension.js
import { HTTPException } from 'hono/http-exception';
import { ProcessError } from '../util/error';
import { TableExtension } from '../tableExtension';
import { appendWhere } from '../../sql/build/select';
import { UPDATE_NEW_COL_ID } from '../../sql/build/update';
import { queryToSqlQuery } from '../../sql/parse/parse';
import { sqlExprSchema } from '../../types/zod/sqlSchemas';
import { tableRulesDataSchema } from '../../types/zod/tableExtensionsSchema';
export class TableRulesExtension extends TableExtension {
    static name = 'rules';
    constructor(data, table, jc) {
        super(tableRulesDataSchema.parse(data), table, jc);
        if (data.name !== TableRulesExtension.name)
            throw new HTTPException(500, { message: 'Invalid Configuration' });
    }
    _applyWhere(jc, query, rule) {
        if (jc.globals.auth?.admin)
            return; // todo add flag?
        // console.log('rule', rule)
        const ruleSql = parseRuleQuery(jc, rule);
        // const simplify = jc.c.req.header('$DB_TEST_PARSE_SIMPLIFY_DISABLE') !== 'true'
        // query.where = query.where ? applyBinaryOperator(query.where, ruleSql, 'AND', simplify) : ruleSql
        appendWhere(query, ruleSql);
    }
    async onInsertParse(query) {
        // todo add allowed columns based on what data is passed in the insert
        //  so if the rule is `role = 'guest'` and role is not in the insert data, throw an error
        this._applyWhere({
            ...this.jc,
            tableName: UPDATE_NEW_COL_ID,
            allowedTables: { [UPDATE_NEW_COL_ID]: this.jc.allowedTables[this.jc.tableName] },
        }, query, this.data.createRule);
    }
    async onDeleteParse(query) {
        this._applyWhere(this.jc, query, this.data.deleteRule);
    }
    async onSelectParse(query) {
        this._applyWhere(this.jc, query, this.data.listRule);
    }
    async onViewParse(query) {
        this._applyWhere(this.jc, query, this.data.viewRule);
    }
    async onUpdateParse(query) {
        this._applyWhere(query.contextWithNew || this.jc, query, this.data.updateRule);
    }
}
export function parseRuleQuery(c, rule) {
    if (rule === null || rule === undefined) {
        // todo check only admin
        throw new ProcessError('Forbidden', 403);
    }
    if (rule === 'true' || rule === '1')
        return { q: "1" };
    if (typeof rule !== 'string')
        throw new ProcessError('Invalid Configuration', 500);
    if (!rule.trim().length)
        throw new ProcessError('Not Found', 404);
    rule = sqlExprSchema.parse(rule);
    try {
        return queryToSqlQuery(rule, c);
    }
    catch (e) {
        throw new ProcessError('Error parsing rule', 500, { input: rule }, e);
    }
}
