// teenybase/dist/worker/extensions/tableAuthExtension.routes.js
import { z } from 'zod';
import { defaultSchema, jwtTokenSchema, uidTokenSchema } from './tableAuthExtension.schema';
export function setupTableAuthExtensionRoutes() {
    const routeZod = {
        '/auth/login-token': () => ({
            description: 'Login with token',
            request: { headers: z.object({ Authorization: z.string().min(10) }) },
            responses: {
                '200': {
                    description: 'Success',
                    content: { 'application/json': {
                            schema: z.object({
                                token: jwtTokenSchema,
                                refresh_token: uidTokenSchema,
                                verified: z.boolean().optional().describe('Email verified. (only returned if false)')
                            }),
                        } },
                },
            }
        }),
        '/auth/refresh-token': () => ({
            description: 'Refresh login token',
            request: { headers: z.object({ Authorization: z.string().min(10) }), body: {
                    required: true,
                    content: { 'application/json': { schema: {
                                refresh_token: uidTokenSchema.describe('Refresh token provided with the token on login/refresh'),
                            } } },
                } },
            responses: {
                '200': {
                    description: 'Success',
                    content: { 'application/json': {
                            schema: z.object({
                                token: jwtTokenSchema,
                                refresh_token: uidTokenSchema,
                                verified: z.boolean().optional().describe('Email verified. (only returned if false)')
                            }),
                        } },
                },
            }
        }),
        '/auth/logout': () => ({
            description: 'Logout. Invalidates the current session. (requires login)',
            request: {},
            responses: {
                '200': {
                    description: 'Success',
                    content: { 'application/json': {
                            schema: z.object({ success: z.literal(true), }),
                        } },
                },
            }
        }),
        '/auth/sign-up': () => ({
            description: 'Sign up. (similar to insert, but returns the login token along with the record)',
            request: { body: {
                    required: true,
                    content: { 'application/json': { schema: this.table.zodSchema } }, // todo we can pick the fields here based on user fields mapping
                    description: 'Data to insert in the table along with passwordConfirm',
                } },
            responses: {
                '200': {
                    description: 'Success',
                    content: { 'application/json': {
                            schema: z.object({
                                token: jwtTokenSchema,
                                refresh_token: uidTokenSchema,
                                verified: z.boolean().optional().describe('Email verified.'),
                                record: this.table.zodSchema, // todo we can pick the fields here based on user fields mapping
                            }),
                        } },
                },
            }
        }),
        '/auth/login-password': () => ({
            description: 'Login with password',
            request: { body: {
                    required: true,
                    content: { 'application/json': { schema: z.object({
                                identity: defaultSchema.identity.describe('Email or username'),
                                // [this.mapping.password]: defaultSchema.password,
                            }).or(z.record(z.string(), z.unknown())) } }, // todo we can pick the fields here based on user fields mapping
                    description: 'identity/email/username and password',
                } },
            responses: {
                '200': {
                    description: 'Success',
                    content: { 'application/json': {
                            schema: z.object({
                                token: jwtTokenSchema,
                                refresh_token: uidTokenSchema,
                                record: this.table.zodSchema, // todo we can pick the fields here based on user fields mapping
                            }),
                        } },
                },
            }
        }),
        '/auth/change-password': () => ({
            description: 'Change password',
            request: { body: {
                    required: true,
                    content: { 'application/json': { schema: z.object({
                            // [this.mapping.password]: defaultSchema.password,
                            }).or(z.record(z.string(), z.unknown())) } }, // todo we can pick the fields here based on user fields mapping
                    description: 'The current password, new password and new password confirm and changes the password',
                } },
            responses: {
                '200': {
                    description: 'Success',
                    content: { 'application/json': {
                            schema: z.object({
                                success: z.literal(true),
                            }),
                        } },
                },
            }
        }),
        '/auth/request-password-reset': () => ({
            description: 'Request password reset. Sends an email to the user with a token to reset the password. (doesn\'t require login)',
            request: {
                body: {
                    required: true,
                    content: { 'application/json': { schema: z.object({
                            // email: defaultSchema.identity.describe('Email field'), // todo
                            }).or(z.record(z.string(), z.unknown())) } },
                    description: 'identity/email/username',
                }
            },
            responses: {
                '200': {
                    description: 'Success',
                    content: { 'application/json': {
                            schema: z.object({ success: z.literal(true), }),
                        } },
                },
            }
        }),
        '/auth/confirm-password-reset': () => ({
            description: 'Confirm password reset. Sends an email to the user with a token to reset the password.',
            request: {
                body: {
                    required: true,
                    content: { 'application/json': { schema: z.object({
                                token: uidTokenSchema,
                            }).and(z.record(z.string(), z.unknown())) } }, // todo password fields
                    description: 'token and new password and new password confirm',
                }
            },
            responses: {
                '200': {
                    description: 'Success',
                    schema: z.object({
                        token: jwtTokenSchema,
                        refresh_token: uidTokenSchema,
                        record: this.table.zodSchema, // todo we can pick the fields here based on user fields mapping
                    }),
                },
            }
        }),
        '/auth/request-verification': () => ({
            description: 'Request email verification. Sends an email to the user with a token to verify. (requires login)',
            request: {},
            responses: {
                '200': {
                    description: 'Success',
                    content: { 'application/json': {
                            schema: z.object({ success: z.literal(true), }),
                        } },
                },
            }
        }),
        '/auth/confirm-verification': () => ({
            description: 'Confirm email verification with token. Returns token and refresh token if not authenticated',
            request: {
                body: {
                    required: true,
                    content: { 'application/json': { schema: z.object({
                                token: uidTokenSchema,
                            }) } },
                    description: 'jwt token received in the email',
                }
            },
            responses: {
                '200': {
                    description: 'Success',
                    schema: z.object({
                        token: jwtTokenSchema.optional(),
                        refresh_token: uidTokenSchema.optional(),
                        record: this.table.zodSchema, // todo we can pick the fields here based on user fields mapping
                    }),
                },
            }
        }),
    };
    // todo make parameter to enable/disable this endpoint when external auth is used
    this.routes.push({
        method: 'post', path: '/auth/login-token',
        handler: this.loginWithToken.bind(this),
        zod: routeZod['/auth/login-token'],
    });
    this.routes.push({
        method: 'post', path: '/auth/refresh-token',
        handler: this.refreshToken.bind(this),
        zod: routeZod['/auth/refresh-token'],
    });
    this.routes.push({
        method: 'post', path: '/auth/logout',
        handler: { raw: this.logout.bind(this) },
        zod: routeZod['/auth/logout'],
    });
    if (this.mapping.password) {
        this.routes.push({
            method: 'post', path: '/auth/sign-up',
            handler: this.signUp.bind(this),
            zod: routeZod['/auth/sign-up'],
        });
        this.routes.push({
            method: 'post', path: '/auth/login-password',
            handler: this.loginWithPassword.bind(this),
            zod: routeZod['/auth/login-password'],
        });
        this.routes.push({
            method: 'post', path: '/auth/change-password',
            handler: { raw: this.changePassword.bind(this) },
            zod: routeZod['/auth/change-password'],
        });
        this.routes.push({
            method: 'post', path: '/auth/request-password-reset',
            handler: { raw: this.requestPasswordReset.bind(this) },
            zod: routeZod['/auth/request-password-reset'],
        });
        this.routes.push({
            method: 'post', path: '/auth/confirm-password-reset',
            handler: { raw: this.confirmPasswordReset.bind(this) },
            zod: routeZod['/auth/confirm-password-reset'],
        });
    }
    if (this.mapping.emailVerified) {
        this.routes.push({
            method: 'post', path: '/auth/request-verification',
            handler: { raw: this.requestVerification.bind(this) },
            zod: routeZod['/auth/request-verification'],
        });
        this.routes.push({
            method: 'post', path: '/auth/confirm-verification',
            handler: { raw: this.confirmVerification.bind(this) },
            zod: routeZod['/auth/confirm-verification'],
        });
    }
}
