// teenybase/dist/worker/extensions/tableAuthExtension.js
import { z } from 'zod';
import { TableExtension } from '../tableExtension';
import { HTTPException } from 'hono/http-exception';
import { ident } from '../../sql/parse/jsep';
import { literalToQuery } from '../../sql/build/query';
import { appendWhere } from '../../sql/build/select';
import { recordToSqlValues } from '../../sql/parse/parse';
import { generateUid, randomString } from '../../security/random';
import { passwordProcessors } from '../util/passwordProcessors';
import { normalizeEmail } from '../util/normalizeEmail';
import { tableAuthDataSchema } from '../../types/zod/tableExtensionsSchema';
import { decode } from '@tsndr/cloudflare-worker-jwt';
import { defaultSchema, uidTokenSchema } from './tableAuthExtension.schema';
import { setupTableAuthExtensionRoutes } from './tableAuthExtension.routes';
const SALT_LENGTH = 20;
export class TableAuthExtension extends TableExtension {
    c;
    static name = 'auth';
    mapping;
    jwtSecret;
    // todo remove c from here
    constructor(data, table, jc, c) {
        super(tableAuthDataSchema.parse(data), table, jc);
        this.c = c;
        if (data.name !== TableAuthExtension.name)
            throw new HTTPException(500, { message: 'Invalid Configuration' });
        if (!this.table.mapping.uid)
            throw new HTTPException(500, { message: 'Invalid Configuration - id field required' });
        this.jwtSecret = this.table.$db.secretResolver.resolver(this.data.jwtSecret, true, `JWT_SECRET for ${data.name}`);
        const aud = this.table.fieldsUsage.auth_audience;
        this.mapping = {
            uid: this.table.mapping.uid,
            username: z.string().optional().parse(this.table.fieldsUsage.auth_username),
            email: z.string().optional().parse(this.table.fieldsUsage.auth_email),
            emailVerified: z.string().optional().parse(this.table.fieldsUsage.auth_email_verified),
            password: z.string().optional().parse(this.table.fieldsUsage.auth_password),
            passwordSalt: z.string().optional().parse(this.table.fieldsUsage.auth_password_salt),
            name: z.string().optional().parse(this.table.fieldsUsage.auth_name),
            avatar: z.string().optional().parse(this.table.fieldsUsage.auth_avatar),
            audience: (aud && !Array.isArray(aud)) ? [z.string().parse(aud)] : aud,
            metadata: z.string().optional().parse(this.table.fieldsUsage.auth_metadata),
            // resetSentAt: z.string().optional().parse(this.table.fieldsUsage.auth_reset_sent_at),
            // verificationSentAt: z.string().optional().parse(this.table.fieldsUsage.auth_verification_sent_at),
        };
        if (this.mapping.emailVerified && !this.mapping.email)
            throw new HTTPException(500, { message: 'Invalid Configuration - email field required for emailVerified' });
        if (this.mapping.passwordSalt && !this.mapping.password)
            throw new HTTPException(500, { message: 'Invalid Configuration - passwordSalt field required for password' });
        setupTableAuthExtensionRoutes.call(this);
    }
    async onInsertParse(query) {
        const v = query.values;
        const records = Array.isArray(v) ? v : [v];
        for (const record of records) {
            await this._parsePasswordFieldsInsert(record);
            this._parseUsername(record);
            if (this.mapping.email)
                this._parseEmail(record);
            if (this.mapping.name)
                defaultSchema.name.parse(record[this.mapping.name]?.l, { path: [this.mapping.name] });
            if (this.mapping.emailVerified) {
                let val = defaultSchema.emailVerified.parse(record[this.mapping.emailVerified]?.l, { path: [this.mapping.emailVerified] });
                if (!this.jc.globals.auth?.admin) {
                    if (val !== undefined)
                        throw new HTTPException(400, { message: `${this.mapping.emailVerified} must not be set` });
                    record[this.mapping.emailVerified] = { l: false };
                }
            }
        }
        // todo send verification email after insert?
    }
    async onUpdateParse(query) {
        const record = query.set;
        await this._parsePasswordFieldUpdate(record, query);
    }
    async onDeleteParse(query, admin) {
        // todo
        if (!admin)
            throw new HTTPException(400, { message: `Not allowed` });
    }
    async onSelectParse(query) {
        // do nothing
    }
    async onViewParse(query) {
        return this.onSelectParse(query);
    }
    // why is this required? because this returns a token...
    async signUp(data) {
        if (!this.mapping.username || !this.mapping.password)
            throw new HTTPException(500, { message: 'Invalid Configuration' });
        const user = await this._createUser(data);
        const ret = {
            ...await this.createSession(user),
            record: this._userToFields(user),
        };
        if (this.mapping.emailVerified) {
            ret.verified = user.emailVerified;
            if (!user.emailVerified && this.data.autoSendVerificationEmail) {
                this.c.executionCtx.waitUntil(this._requestVerification(user));
            }
        }
        return ret;
    }
    async requestVerification() {
        if (!this.mapping.email || !this.mapping.emailVerified)
            throw new HTTPException(500, { message: 'Invalid Configuration' });
        const auth = this.jc.globals.auth;
        if (!auth?.uid)
            throw new HTTPException(403, { message: 'Unauthorized' });
        if (auth.verified)
            throw new HTTPException(400, { message: 'Already verified' });
        const user = await this.getUser(auth.uid);
        if (!user)
            throw new HTTPException(400, { message: 'User not found' });
        await this._requestVerification(user);
        return this.c.json({ success: true });
    }
    confirmRequiresAuth = false; // todo make setting
    async confirmVerification() {
        const data = await this.table.$db.getRequestBody();
        if (!data)
            throw new HTTPException(400, { message: 'Usage: POST /auth/confirm-verification {data} in JSON/FormData' });
        // console.log(data.token)
        const token = uidTokenSchema.parse(data.token);
        if (!token)
            throw new HTTPException(400, { message: 'Token required' });
        const auth = this.jc.globals.auth;
        if (this.confirmRequiresAuth && (!auth || !auth.uid || !auth.sid))
            throw new HTTPException(403, { message: 'Unauthorized' });
        const uid = this.confirmRequiresAuth ? auth.uid : undefined;
        const user = await this._confirmVerification(token, uid);
        const ret = {
            ...(!auth?.sid && await this.createSession(user)),
            record: this._userToFields(user),
        };
        return this.c.json(ret);
    }
    async requestPasswordReset() {
        if (!this.mapping.email || !this.mapping.password)
            throw new HTTPException(500, { message: 'Invalid Configuration' });
        const data = await this.table.$db.getRequestBody();
        if (!data)
            throw new HTTPException(400, { message: 'Usage: POST /auth/request-password-reset {data} in JSON/FormData' });
        const record = recordToSqlValues(data);
        const email = this._parseEmail(record);
        const user = await this.findUser(email, true);
        if (!user)
            throw new HTTPException(400, { message: 'User not found' });
        const auth = this.jc.globals.auth;
        // auth is optional here. todo make this configurable
        if (auth?.uid && (auth.uid !== user.id || auth.email !== email))
            throw new HTTPException(403, { message: 'Unauthorized' });
        await this._requestPasswordReset(user);
        return this.c.json({ success: true });
    }
    async confirmPasswordReset() {
        if (!this.mapping.email || !this.mapping.password)
            throw new HTTPException(500, { message: 'Invalid Configuration' });
        const data = await this.table.$db.getRequestBody();
        if (!data)
            throw new HTTPException(400, { message: 'Usage: POST /auth/confirm-password-reset {data} in JSON/FormData' });
        const token = uidTokenSchema.parse(data.token);
        if (!token)
            throw new HTTPException(400, { message: 'Token required' });
        delete data.token;
        const record = recordToSqlValues(data);
        const password = await this._parsePasswordFieldsInsert(record);
        if (!password)
            throw new HTTPException(400, { message: 'Invalid password' });
        const user = await this._confirmPasswordReset(token, record);
        const ret = {
            ...await this.createSession(user),
            record: this._userToFields(user),
        };
        return this.c.json(ret);
    }
    async loginWithPassword(data) {
        if (!this.mapping.password)
            throw new HTTPException(500, { message: 'Invalid Configuration' });
        if (!data)
            throw new HTTPException(400, { message: 'Usage: POST /auth/login-password {data} in JSON' });
        const record = recordToSqlValues(data);
        const username = (this.mapping.email && record[this.mapping.email]) ? this._parseEmail(record) :
            record['identity'] ? defaultSchema.identity.parse(record['identity'].l, { path: ["identity"] }) :
                this._parseUsername(record);
        const invalid = async () => {
            await new Promise((r) => setTimeout(r, Math.random() * 200)); // todo
            return new HTTPException(400, { message: 'Invalid username or password' });
        };
        const user = username ? await this.findUser(username, false, true) : null;
        if (!user)
            throw await invalid();
        const salt = user.passwordSalt || '';
        const password = await this._parsePasswordFieldsLogin(record, salt).catch(e => {
            console.error(e);
            return null;
        });
        // console.log(password, res)
        if (!password || password !== user.password)
            throw await invalid();
        // todo
        // if(this.mapping.emailVerified && !user.emailVerified) throw new HTTPException(400, {message: 'Email not verified'})
        const ret = {
            ...await this.createSession(user),
            record: this._userToFields(user),
        };
        if (this.mapping.emailVerified && !user.emailVerified)
            ret.verified = false;
        // return this.c.json(ret)
        return ret;
    }
    /**
     * validates token from google, github etc or self(disabled for now) and creates a new session
     */
    async loginWithToken() {
        const tok = this.c.req.header('Authorization')?.replace(/^Bearer /, '');
        if (!tok)
            throw new HTTPException(400, { message: 'Authorization header required' });
        const payload = await this.table.$db.jwt.decodeAuth(tok, await this.jwtSecret(), false).catch(() => null);
        // payload will be null when sign not valid or expired etc
        if (!payload)
            throw new HTTPException(401, { message: 'Unauthorized' });
        let res = await this.findUser(payload.sub, true);
        if (payload.iss !== this.table.$db.jwt.issuer) {
            // todo why this false and 0 check? because it can be 'true'?
            if (payload.verified === false || payload.verified === 0)
                throw new HTTPException(400, { message: 'Not verified' });
            // todo make a setting in this.data to control whether a new user should be created on new token or should it be explicit
            if (!res)
                res = await this._createUserToken(payload.sub, true, payload?.issData);
        }
        else {
            // do nothing
            // todo throw error or refresh session?
            throw new HTTPException(400, { message: 'Use refresh-token endpoint instead' });
        }
        if (!res)
            throw new HTTPException(401, { message: 'Unauthorized' });
        // if(this.mapping.emailVerified && !res.emailVerified) throw new HTTPException(400, {message: 'Email not verified'})
        if (res.email !== payload.sub)
            throw new HTTPException(400, { message: 'Invalid email' });
        const ret = {
            ...await this.createSession(res),
            record: this._userToFields(res),
        };
        if (this.mapping.emailVerified && !res.emailVerified)
            ret.verified = false;
        return ret;
    }
    async refreshToken(data) {
        const tok = this.c.req.header('Authorization')?.replace(/^Bearer /, '');
        if (!tok)
            throw new HTTPException(400, { message: 'Authorization header required' });
        // todo not really required but can we verify the signature without verifying exp
        const payloadUntrusted = decode(tok).payload;
        if (payloadUntrusted.iss !== this.table.$db.jwt.issuer || !payloadUntrusted.sub || !payloadUntrusted.sid)
            throw new HTTPException(401, { message: 'Invalid token' });
        const { sub, sid } = payloadUntrusted;
        const refreshToken = uidTokenSchema.parse(data.refresh_token);
        const sessionId = z.string().min(10).parse(sid); // todo zod
        let res = await this.findUser(sub, true);
        if (!res)
            throw new HTTPException(401, { message: 'Unauthorized' });
        // if(this.mapping.emailVerified && !res.emailVerified) throw new HTTPException(400, {message: 'Email not verified'})
        if (res.email !== sub)
            throw new HTTPException(400, { message: 'Invalid email' });
        const token = await this.refreshSession(res, sessionId, refreshToken);
        const ret = {
            ...token,
            record: this._userToFields(res),
        };
        if (this.mapping.emailVerified && !res.emailVerified)
            ret.verified = false;
        return ret;
    }
    async changePassword() {
        if (!this.data.passwordType)
            throw new HTTPException(500, { message: 'Invalid Configuration' });
        const email = this.jc.globals.auth?.email;
        if (!this.jc.globals.auth?.uid || !email)
            throw new HTTPException(401, { message: 'Unauthorized' });
        const data = await this.c.req.json();
        if (!data)
            throw new HTTPException(400, { message: 'Usage: POST /auth/change-password {data} in JSON' });
        const record = recordToSqlValues(data);
        const res = await this.findUser(email, false, true);
        if (!res)
            throw new HTTPException(400, { message: 'Invalid user' });
        const salt = res.passwordSalt || '';
        if (this.data.passwordCurrentSuffix) {
            let current = await this._parseCurrentPasswordField(record, salt);
            if (current !== res.password)
                throw new HTTPException(400, { message: 'Invalid current password' });
        }
        const password = await this._parsePasswordFieldsInsert(record);
        if (!password)
            throw new HTTPException(400, { message: 'Invalid password' });
        const res2 = await this.updateUser(res.id, record);
        if (!res2 || res2.id !== res.id)
            throw new HTTPException(500, { message: 'Failed to update password' });
        // Invalidate all sessions after password change for security
        const _token = await this._invalidateAllUserSessions(res.id); // return token
        return this.c.json({ success: true });
    }
    async logout() {
        const auth = this.jc.globals.auth;
        if (!auth?.uid || !auth?.sid)
            throw new HTTPException(401, { message: 'Unauthorized' });
        // const user = await this.getUser(auth.uid)
        // if(!user) throw new HTTPException(400, {message: 'User not found'})
        const sessionKey = this._kvSessionKey(/*user*/ { id: auth.uid }, auth.sid);
        await this.table.$db.kv.remove(sessionKey);
        return this.c.json({ success: true });
    }
    // todo
    // login with token
    // reset/forgot password
    // username change
    // username recovery
    // no point of this for this frontend, just use normal token. auth0 also says no point - https://auth0.com/blog/refresh-tokens-what-are-they-and-when-to-use-them/
    // actually useful to update meta and stuff
    // refresh token
    async _refreshSession(res, session) {
        const data = {
            cid: this.table.data.name,
            user: res.username, // todo: should this be added?
            sub: res.email,
            id: res.id,
            meta: res.metadata,
            sid: session.id, // session id
            // admin: false,
        };
        if (res.audience && res.audience.length)
            data.aud = res.audience;
        if (res.emailVerified !== undefined)
            data.verified = res.emailVerified;
        const now = this.getTimestamp();
        const refreshToken = generateUid();
        session.tok = refreshToken;
        session.rc++;
        session.rat = now; // Update refreshed at timestamp
        // todo parameter
        const refreshTokenDuration = /*this.data.refreshTokenDuration || */ 60 * 60 * 24 * 7; // default to 7 days
        const sessionDuration = /*this.data.sessionDuration || */ 60 * 60 * 24 * 30; // default to 30 days
        // Implement sliding expiration - extend session exp on each refresh
        session.exp = now + sessionDuration;
        const sessionKey = this._kvSessionKey(res, session.id);
        // todo limit number of sessions - check how many sessions in table for the user, if more than allowed, deleted the first one
        // overwrite existing session
        await this.table.$db.kv.setMultiple({
            [sessionKey]: JSON.stringify(session),
        }, refreshTokenDuration);
        const token = await this.table.$db.jwt.createJwtToken(data, await this.jwtSecret(), this.data.jwtTokenDuration);
        return {
            token,
            refresh_token: refreshToken,
        };
    }
    getTimestamp() {
        return Math.floor(Date.now() / 1000); // seconds since epoch
    }
    async createSession(user) {
        // save session and refresh token in the database
        const now = this.getTimestamp();
        const session = {
            id: generateUid(),
            tok: '', // refresh token
            cid: this.table.name,
            uid: user.id,
            sub: user.email,
            rc: 0, // refresh count
            cat: now, // created at
            rat: now, // refreshed at
            exp: now, // expires at (updated in refresh)
            // todo ip etc
        };
        return await this._refreshSession(user, session);
    }
    async refreshSession(user, id, token) {
        // save session and refresh token in the database
        const sessionKey = this._kvSessionKey(user, id);
        const sessionStr = await this.table.$db.kv.get(sessionKey);
        const now = this.getTimestamp();
        // Parse the JSON string to get the session object
        let session = null;
        if (sessionStr) {
            try {
                session = JSON.parse(sessionStr);
            }
            catch (e) {
                console.error(e);
            }
        }
        const isValid = !!session
            && session.tok === token
            && session.uid === user.id
            && (!this.data.maxTokenRefresh || session.rc < this.data.maxTokenRefresh)
            && (session.exp === undefined || session.exp > now);
        if (!isValid || !session)
            throw new HTTPException(401, { message: 'Invalid session' });
        return await this._refreshSession(user, session);
    }
    async findUser(loginId, emailOnly = false, asAdmin = false) {
        const username = (!emailOnly && this.mapping.username) ? `${ident(this.mapping.username, this.jc)} = {:loginId}` : null;
        const email = this.mapping.email ? `${ident(this.mapping.email, this.jc)} = {:loginId}` : null;
        if (!username && !email)
            throw new HTTPException(500, { message: 'Invalid Configuration' });
        const res = await this.table.$db.rawSelect(this.table, {
            from: this.jc.tableName,
            where: {
                q: (username && email) ? `${username} OR ${email}` : email || username,
                p: { loginId: loginId }
            },
            selects: this._userFields(true, asAdmin),
            limit: 1,
            _readOnly: true,
        })?.run() ?? [];
        return this._fieldsToUser(res[0], true);
    }
    async getUser(id) {
        const res = await this.table.$db.rawSelect(this.table, {
            from: this.jc.tableName,
            where: {
                q: `${ident(this.mapping.uid, this.jc)} = {:uid}`,
                p: { uid: id }
            },
            selects: this._userFields(),
            limit: 1,
            _readOnly: true,
        })?.run() ?? [];
        return this._fieldsToUser(res[0]);
    }
    async updateUser(id, data) {
        const res = await this.table.$db.rawUpdate(this.table, {
            table: this.jc.tableName,
            where: {
                q: `${ident(this.mapping.uid, this.jc)} = {:uid}`,
                p: { uid: id }
            },
            set: data,
            returning: this._userFields(),
        })?.run() ?? [];
        return this._fieldsToUser(res[0]);
    }
    async _createUser(data) {
        const res = await this.table.insert({ values: data, returning: this._userFields(false) }) ?? [];
        const user = res ? this._fieldsToUser(res[0]) : null;
        if (!user)
            throw new HTTPException(400, { message: 'Unable to create user' }); // todo maybe tell user it can be because of rules?
        return user;
    }
    // todo we need to save id of the user from the external platform also.
    // this should only be used for auto user creation
    async _createUserToken(email, verified, externalData, record, values2) {
        if (!this.table.data.autoSetUid)
            throw new HTTPException(500, { message: 'Invalid Configuration - autoSetUid required for auth extension' });
        const values = {
            ...record,
            ...values2,
            [this.mapping.uid]: { l: generateUid() },
        };
        // todo check if email, username, name is not null, unique etc and handle that
        if (this.mapping.email && email)
            values[this.mapping.email] = { l: email };
        if (this.mapping.emailVerified)
            values[this.mapping.emailVerified] = { l: verified };
        if (this.mapping.username)
            values[this.mapping.username] = { l: externalData.username || email?.split('@')[0].toLowerCase().replace(/[^a-z0-9_]/g, '') || '' };
        if (this.mapping.name)
            values[this.mapping.name] = { l: externalData.name || 'User' };
        // if(this.mapping.avatar) // todo avatar url
        // todo put whole externalData in raw_user_meta_data like supabase
        if (this.data.passwordType && this.mapping.password && !values[this.mapping.password]) {
            let pass = generateUid();
            const salt = this.mapping.passwordSalt ? randomString(SALT_LENGTH) : '';
            pass = await this._hashPassword(pass, salt);
            values[this.mapping.password] = { l: pass };
            if (this.mapping.passwordSalt)
                values[this.mapping.passwordSalt] = { l: salt };
        }
        const res = await this.table.$db.rawInsert(this.table, {
            table: this.jc.tableName,
            values,
            returning: this._userFields(),
        })?.run() ?? [];
        return this._fieldsToUser(res[0]);
    }
    _fieldsToUser(res, asAdmin = false) {
        return res ? {
            id: res[this.mapping.uid],
            username: this.mapping.username ? res[this.mapping.username] : undefined,
            email: this.mapping.email ? res[this.mapping.email] : undefined,
            password: asAdmin && this.mapping.password ? res[this.mapping.password] : undefined,
            passwordSalt: asAdmin && this.mapping.passwordSalt ? res[this.mapping.passwordSalt] : undefined,
            emailVerified: this.mapping.emailVerified ? Boolean(res[this.mapping.emailVerified]) : undefined,
            audience: this.mapping.audience ? this.mapping.audience.map((f) => res[f]).filter(v => v) : undefined,
            metadata: this.mapping.metadata ? JSON.parse(res[this.mapping.metadata]) : undefined,
        } : null;
    }
    _userToFields(user) {
        const res = {
            [this.mapping.uid]: user.id,
        };
        if (this.mapping.username)
            res[this.mapping.username] = user.username;
        if (this.mapping.email)
            res[this.mapping.email] = user.email;
        // if(this.mapping.password) res[this.mapping.password] = user.password!
        // if(this.mapping.passwordSalt) res[this.mapping.passwordSalt] = user.passwordSalt!
        if (this.mapping.emailVerified)
            res[this.mapping.emailVerified] = user.emailVerified;
        if (this.mapping.metadata)
            res[this.mapping.metadata] = JSON.stringify(user.metadata);
        return res;
    }
    _userFields(identifier = true, asAdmin = false) {
        const f = [
            this.mapping.uid, this.mapping.username, this.mapping.email,
            this.mapping.password, this.mapping.passwordSalt,
            this.mapping.emailVerified, ...this.mapping.audience ?? [],
            this.mapping.metadata,
        ].filter(v => {
            const f = v ? this.table.fields[v] : undefined;
            return f && (!f.noSelect || asAdmin);
        });
        const jc = !asAdmin ? this.jc : {
            ...this.jc,
            _checkColumns: !asAdmin, // todo see if we can pass asAdmin(or something) and that would include the password columns
        };
        return !identifier ? f : f.map(v => ident(v, jc));
    }
    _kvVerificationSentKey = (user) => '@email_verification_sent_at_' + this.table.name + user.id;
    _kvTokenKey = (token) => '@token_' + token; // + this.table.name + user.id
    _kvSessionKey = (user, id) => '@session_' + this.table.name + user.id + '_' + id; // + this.table.name + user.id
    _kvPasswordResetSentKey = (user) => '@password_reset_sent_at_' + this.table.name + user.id;
    async _invalidateAllUserSessions(userId) {
        // todo get all keys with prefix and delete them
    }
    _getEmailVars() {
        return {
            APP_NAME: this.table.$db.settings.appName || 'Dollar App',
            APP_URL: this.table.$db.settings.appUrl,
            // todo add RECORD:* for record specific data(username, name, email, role etc) like in pocketbase
        };
    }
    async _requestVerification(user) {
        if (!this.mapping.emailVerified)
            throw new HTTPException(500, { message: 'Invalid Configuration - emailVerified field required' });
        if (!this.mapping.email)
            throw new HTTPException(500, { message: 'Invalid Configuration - email field required' });
        if (!user.email)
            throw new HTTPException(500, { message: 'No user' });
        if (user.emailVerified)
            return;
        const verificationKey = this._kvVerificationSentKey(user);
        const emailVerifyEmailDuration = this.data.emailVerifyEmailDuration || 2 * 60; // 2 min
        // this check is to prevent sending verification email too often
        // every email will generate and send a new token but previous ones will still keep working
        const res = await this.table.$db.kv.get(verificationKey, '(unixepoch(CURRENT_TIMESTAMP) - unixepoch(value))');
        if (res !== null && res < emailVerifyEmailDuration - 1) {
            // throw new Error('Verification email already sent') // todo better error for frontend
            throw new HTTPException(400, { message: 'Verification email already sent' });
        }
        // const token = await this.table.$db.jwt.createJwtToken<{typ: string, cid: string}>({
        //     id: user.id,
        //     sub: user.email,
        //     typ: 'verify_email',
        //     cid: this.table.name,
        // }, await this.jwtSecret(), this.data.emailVerifyTokenDuration || 1 * 60 * 60)
        const token = generateUid();
        const email = this.table.$db.email;
        if (!email)
            throw new HTTPException(500, { message: 'Email not configured' });
        const template = this.data.emailTemplates?.verification || {};
        // todo move defaults out of class
        await email.sendActionLink({
            subject: template.subject || 'Verify your {{APP_NAME}} email',
            tags: ['email-verification', 'table-' + this.table.name, ...(template.tags || [])],
            to: user.email,
            variables: {
                message_title: 'Email Verification',
                message_description: 'Thank you for joining us at {{APP_NAME}}. Click the button below to verify your email address.',
                message_footer: 'If you did not request this, please ignore this email.',
                action_text: 'Verify Email',
                action_link: '{{APP_URL}}/verify-email/{{TOKEN}}',
                ...template.variables,
                TOKEN: token,
                ...this._getEmailVars(),
            }
        });
        const verificationTokenKey = this._kvTokenKey(token);
        const emailVerifyTokenDuration = this.data.emailVerifyTokenDuration || 1 * 60 * 60; // 1 hour
        // await this.table.$db.kv.setSql(verificationKey, 'CURRENT_TIMESTAMP')
        await this.table.$db.kv.setMultiple({
            [verificationKey]: { sql: 'CURRENT_TIMESTAMP' },
            [verificationTokenKey + token]: JSON.stringify({ id: user.id, sub: user.email, typ: 'verify_email', cid: this.table.name, }),
        }, emailVerifyTokenDuration);
        // no need to save token as the user is verified when the token is used, or the token expires.
    }
    async _confirmVerification(token, uid) {
        if (!this.mapping.emailVerified)
            throw new HTTPException(500, { message: 'Invalid Configuration - emailVerified field required' });
        const { id, sub } = await this._useToken(token, 'verify_email');
        if (uid && uid !== id)
            throw new Error('User mismatch');
        const user = await this.findUser(sub, true);
        if (!user || user.id !== id)
            throw new HTTPException(400, { message: 'Invalid user' });
        if (user.emailVerified)
            return user;
        const user2 = await this.updateUser(user.id, {
            [this.mapping.emailVerified]: { l: true },
        });
        if (!user2)
            throw new HTTPException(500, { message: 'Failed to verify email' });
        return user2;
    }
    async _requestPasswordReset(user) {
        if (!this.mapping.password)
            throw new HTTPException(500, { message: 'Invalid Configuration - password field required' });
        if (!user.email)
            throw new HTTPException(500, { message: 'No user' });
        const resetKey = this._kvPasswordResetSentKey(user);
        const passwordResetEmailDuration = this.data.passwordResetEmailDuration || 2 * 60; // 2 min
        // every email will generate and send a new token but previous ones will still keep working
        const res = await this.table.$db.kv.get(resetKey, '(unixepoch(CURRENT_TIMESTAMP) - unixepoch(value))');
        if (res !== null && res < passwordResetEmailDuration - 1) {
            throw new Error('Password reset email already sent'); // todo better error for frontend
        }
        // const token = await this.table.$db.jwt.createJwtToken<{typ: string, cid: string}>({
        //     id: user.id,
        //     sub: user.email,
        //     typ: 'reset_password',
        //     cid: this.table.name,
        // }, await this.jwtSecret(), this.data.passwordResetTokenDuration || 60 * 60)
        const token = generateUid();
        const email = this.table.$db.email;
        if (!email)
            throw new HTTPException(500, { message: 'Email not configured' });
        const template = this.data.emailTemplates?.passwordReset || {};
        await email.sendActionLink({
            subject: template.subject || 'Reset your {{APP_NAME}} password',
            tags: ['password-reset', 'table-' + this.table.name, ...(template.tags || [])],
            to: user.email,
            variables: {
                message_title: 'Password Reset',
                message_description: 'Click the button below to reset the password for your {{APP_NAME}} account.',
                message_footer: 'If you did not request this, you can safely ignore this email.',
                action_text: 'Reset Password',
                action_link: '{{APP_URL}}/reset-password/{{TOKEN}}',
                ...template.variables,
                TOKEN: token,
                ...this._getEmailVars(),
            }
        });
        const resetTokenKey = this._kvTokenKey(token);
        const passwordResetTokenDuration = this.data.passwordResetTokenDuration || 60 * 60; // 1 hour
        await this.table.$db.kv.setMultiple({
            [resetKey]: { sql: 'CURRENT_TIMESTAMP' },
            [resetTokenKey + token]: JSON.stringify({ id: user.id, sub: user.email, typ: 'reset_password', cid: this.table.name }),
        }, passwordResetTokenDuration);
    }
    // record should have password, passwordSalt
    async _confirmPasswordReset(token, record) {
        if (!this.mapping.password)
            throw new HTTPException(500, { message: 'Invalid Configuration - password field required' });
        // todo use token after password reset is successful otherwise its annoying
        const { id, sub } = await this._useToken(token, 'reset_password');
        const user = await this.getUser(id);
        if (!user || user.id !== id)
            throw new HTTPException(400, { message: 'Invalid user' });
        // const resetTokenKey = this._kvPasswordResetTokenKey(user)
        // const tokenSaved = await this.table.$db.kv.get(resetTokenKey)
        // if(token !== tokenSaved) throw new HTTPException(400, {message: 'Invalid token (mismatch)'})
        if (sub !== user.email) // todo is there any case for this? (pocketbase also has this check for unverified emails)
            throw new HTTPException(400, { message: 'Invalid email' });
        const data = {};
        if (this.mapping.passwordSalt) {
            const salt = record[this.mapping.passwordSalt];
            if (!salt.l)
                throw new HTTPException(500, { message: 'Invalid salt' });
            data[this.mapping.passwordSalt] = salt;
        }
        data[this.mapping.password] = record[this.mapping.password];
        if (this.mapping.emailVerified && !user.emailVerified)
            data[this.mapping.emailVerified] = { l: true };
        const user2 = await this.updateUser(user.id, data);
        if (!user2)
            throw new HTTPException(500, { message: 'Failed to verify email' });
        // Invalidate all sessions after password reset for security
        const _token = await this._invalidateAllUserSessions(user.id); // todo return token
        // await this.table.$db.kv.remove(resetTokenKey)
        return user2;
    }
    async _useToken(token, typ1) {
        const res = await this.table.$db.kv.pop(this._kvTokenKey(token));
        let parsed = null;
        try {
            parsed = res ? JSON.parse(res) : null;
        }
        catch (e) {
            console.error(e);
        }
        if (!parsed)
            throw new HTTPException(400, { message: 'Invalid token' });
        const { id, sub, typ, cid } = parsed;
        // const {id, sub, typ, cid} = await this.table.$db.jwt.decodeAuth(token, await this.jwtSecret(), false)
        if (!id || !sub)
            throw new HTTPException(400, { message: 'Invalid token' });
        if (cid !== this.table.name)
            throw new HTTPException(400, { message: 'Invalid table' });
        if (typ !== typ1)
            throw new HTTPException(400, { message: 'Invalid token type' });
        return { id, sub };
    }
    async _hashPassword(password, salt) {
        if (!this.data.passwordType)
            return '';
        return passwordProcessors[this.data.passwordType].hash(password, salt);
    }
    _parseUsername(record) {
        if (!this.mapping.username)
            return;
        let val = record[this.mapping.username]?.l;
        if (typeof val !== 'string')
            throw new HTTPException(400, { message: `${this.mapping.username} must be a value, expressions not support in username` });
        val = defaultSchema.username.parse(val, { path: [this.mapping.username] });
        record[this.mapping.username] = { l: val };
        return val;
    }
    _parseEmail(record) {
        if (!this.mapping.email)
            throw new HTTPException(500, { message: 'Invalid Configuration' });
        let val = record[this.mapping.email]?.l;
        if (typeof val !== 'string')
            throw new HTTPException(400, { message: `${this.mapping.email} must be a value, expressions not support in email` });
        val = defaultSchema.email.parse(val, { path: [this.mapping.email] });
        if (this.data.normalizeEmail !== false) {
            try {
                val = normalizeEmail(val);
            }
            catch (e) {
                console.error(e);
                throw new HTTPException(400, { message: `Invalid ${this.mapping.email} format` });
            }
        }
        // todo check domain blocklists
        record[this.mapping.email] = { l: val };
        return val;
    }
    async _parseCurrentPasswordField(record, salt) {
        let current = record[this.mapping.password + this.data.passwordCurrentSuffix]?.l;
        if (!current || typeof current !== 'string')
            throw new HTTPException(400, { message: `${this.mapping.password + this.data.passwordCurrentSuffix} is required` });
        current = defaultSchema.password.parse(current, { path: [this.mapping.password + this.data.passwordCurrentSuffix] });
        delete record[this.mapping.password + this.data.passwordCurrentSuffix];
        return await this._hashPassword(current, salt);
    }
    async _parsePasswordFieldUpdate(record, query, salt) {
        if (!this.data.passwordType)
            return;
        if (!this.mapping.password)
            throw new HTTPException(500, { message: 'Invalid Configuration' });
        let password = record[this.mapping.password]?.l;
        if (!password) {
            if (record[this.mapping.password])
                delete record[this.mapping.password];
            if (this.data.passwordConfirmSuffix && record[this.mapping.password + this.data.passwordConfirmSuffix])
                delete record[this.mapping.password + this.data.passwordConfirmSuffix];
            return;
        }
        const isAdmin = this.jc.globals.auth?.admin;
        if (this.data.passwordCurrentSuffix && !isAdmin) {
            if (!this.mapping.passwordSalt || salt) { // todo get the current salt from the db
                let current = await this._parseCurrentPasswordField(record, salt || '');
                const currentQ = literalToQuery(current);
                const where = {
                    q: `${ident(this.mapping.password, this.jc)} = ${currentQ.q}`,
                    p: currentQ.p
                };
                // const simplify = this.c.req.header('$DB_TEST_PARSE_SIMPLIFY_DISABLE') !== 'true'
                // query.where = query.where ? applyBinaryOperator(query.where, where, 'AND', simplify) : where
                appendWhere(query, where);
            }
            else {
                throw new HTTPException(400, { message: 'Not supported, use change-password route(TBD)' });
            }
        }
        await this._parsePasswordFieldsInsert(record);
    }
    async _parsePasswordFieldsInsert(record) {
        if (this.data.passwordType) {
            const field = this.mapping.password;
            if (!field)
                throw new HTTPException(500, { message: 'Invalid Configuration' });
            let val = record[field]?.l;
            if (!val)
                throw new HTTPException(400, { message: `${field} is required` });
            if (typeof val !== 'string')
                throw new HTTPException(400, { message: `${field} must be a value, expressions not support in password` });
            val = defaultSchema.password.parse(val, { path: [field] });
            if (this.data.passwordConfirmSuffix) {
                const confirmVal = record[field + this.data.passwordConfirmSuffix]?.l;
                // zod match so we get proper errors
                // https://github.com/colinhacks/zod/discussions/1525
                z.literal(val, {
                    // todo better error message
                    errorMap: () => ({ message: `${field} and ${field + this.data.passwordConfirmSuffix} do not match` }),
                }).parse(confirmVal, { path: [field + this.data.passwordConfirmSuffix] });
                if (val !== confirmVal)
                    throw new HTTPException(400, { message: 'unknown zod error' });
                delete record[field + this.data.passwordConfirmSuffix];
            }
            const salt = this.mapping.passwordSalt ? randomString(SALT_LENGTH) : '';
            val = await this._hashPassword(val, salt);
            record[field] = { l: val };
            if (this.mapping.passwordSalt)
                record[this.mapping.passwordSalt] = { l: salt };
            return record[field];
        }
        else if (this.mapping.password) {
            throw new HTTPException(500, { message: 'Invalid Configuration' });
        }
        return null;
    }
    async _parsePasswordFieldsLogin(record, salt) {
        if (!this.data.passwordType)
            return;
        const field = this.mapping.password;
        if (!field)
            throw new HTTPException(500, { message: 'Invalid Configuration' });
        let val = (record[field])?.l;
        if (typeof val !== 'string')
            throw new HTTPException(400, { message: `${field} must be a value, expressions not support in password` });
        val = defaultSchema.password.parse(val, { path: [field] });
        val = await this._hashPassword(val, salt);
        record[field] = { l: val };
        return val;
    }
}
