// teenybase/dist/worker/util/error.js
import { HTTPException } from 'hono/http-exception';
/**
 * parse error with message, input data, and embedded error
 */
export class ProcessError extends HTTPException {
    message;
    input;
    get data() {
        return {
            error: this.cause?.message ? this.cause.message : this.cause,
            ...this.input
        };
    }
    constructor(message, code = 400, input, cause) {
        super(code, { message, cause });
        this.message = message;
        this.input = input;
    }
}
export class D1Error extends Error {
    message;
    errorMessage;
    cause;
    input;
    get data() {
        return {
            error: this.errorMessage,
            // cause: (this.cause as any)?.message ? (this.cause as any).message : this.cause,
            input: this.input,
            ...this._data,
        };
    }
    _data;
    set data(d) {
        this._data = d;
    }
    constructor(message, errorMessage, cause, input) {
        super(message);
        this.message = message;
        this.errorMessage = errorMessage;
        this.cause = cause;
        this.input = input;
    }
}
export class HTTPError extends HTTPException {
    isHTTPException = true;
}
