// teenybase/dist/worker/util/openapi.js
import { OpenApiGeneratorV31, OpenAPIRegistry } from '@asteasolutions/zod-to-openapi';
import { z } from 'zod';
import { swaggerUI } from '@hono/swagger-ui';
import { $DBExtension } from '../$DBExtension';
import { extendZodWithOpenApi } from '@hono/zod-openapi';
export class OpenApiExtension extends $DBExtension {
    constructor(db, swagger = true) {
        super(db);
        // todo move it somewhere else, shouldn't be in the constructor
        extendZodWithOpenApi(z);
        this.routes.push({
            path: '/doc',
            method: 'get',
            handler: { raw: async () => {
                    return db.c.json(this.getDoc());
                } },
            zod: () => ({
                description: 'OpenAPI documentation',
                request: {},
                responses: {
                    '200': {
                        description: 'Success',
                        content: { 'application/json': {
                                schema: z.record(z.unknown()),
                            } },
                    }
                },
            })
        });
        if (swagger) {
            this.routes.push({
                path: '/doc/ui',
                method: 'get',
                handler: { raw: async () => {
                        const res = await swaggerUI({ url: this.db.apiBase + '/doc' })(db.c, async () => { return; });
                        return res ? res : undefined;
                    } },
                zod: () => ({
                    description: 'Swagger UI',
                    request: {},
                    responses: {
                        '200': {
                            description: 'Success',
                            content: { 'text/html': {
                                    schema: z.string(),
                                } },
                        }
                    },
                })
            });
        }
    }
    getDoc() {
        const registry = new OpenAPIRegistry();
        this.getApiRoutes(this.db.getRoutes(), this.db.apiBase).forEach((route) => registry.registerPath(route));
        this.db.allTables().forEach((table) => this.getApiRoutes(table.getRoutes(), this.db.apiTableBase + '/' + table.name).forEach((route) => registry.registerPath(route)));
        const generator = new OpenApiGeneratorV31(registry.definitions);
        const doc = generator.generateDocument({
            openapi: '3.1.0',
            info: {
                version: '1.0.0',
                title: 'Teenybase API',
                // description: '',
                // contact: {},
                // license: {},
                // termsOfService: '',
            },
        });
        return doc;
    }
    getApiRoutes(routes, prefix = '') {
        let apiRoutes = [];
        // const routes = this.getRoutes()
        for (const route of routes) {
            const zod = route.zod();
            if (!zod.request.headers)
                zod.request.headers = z.object({ authorization: z.string().min(1).max(255).optional() });
            apiRoutes.push({
                path: prefix + route.path,
                method: route.method.toLowerCase(),
                ...zod,
            });
        }
        return apiRoutes;
    }
}
